{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "marine-tiffany",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "sys.path.append(\"/home/weber/PycharmProjects/EXOTIC/src\")\n",
    "from utils.utils import load_config_file, mkdir\n",
    "\n",
    "yaml = load_config_file(config_file=\"/home/weber/PycharmProjects/EXOTIC/clean/src/config_clean_clean.yaml\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "sporting-blood",
   "metadata": {},
   "source": [
    "### Fct to compute Odds Ratio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "minute-parade",
   "metadata": {},
   "outputs": [],
   "source": [
    "def stats(cat_const, cat_alt, others_const, others_alt, col, cat, ratio=None):\n",
    "    odd_ratio = (cat_const/others_const) / (cat_alt/others_alt)\n",
    "    se = np.sqrt((1/cat_const) + (1/others_const) + (1/cat_alt) + (1/others_alt))\n",
    "    lower = np.exp(np.log(odd_ratio) - 1.96 *se)\n",
    "    upper = np.exp(np.log(odd_ratio) + 1.96 *se)\n",
    "    print(odd_ratio, se, lower, upper)\n",
    "    d = {\n",
    "            'col' : col,\n",
    "            'cat' : cat,\n",
    "            'ratio' : ratio,\n",
    "            'OR' : odd_ratio,\n",
    "            'SE' : se,\n",
    "            'Lower_CI' : lower,\n",
    "            'Upper_CI' : upper,\n",
    "        }\n",
    "    return d\n",
    "\n",
    "def compute_odds_ratio(df, cols):\n",
    "    l = list()\n",
    "    for col in cols:\n",
    "        print(col)\n",
    "        print(df[['Const_Alt', col]].pivot(columns='Const_Alt', values=col).apply(pd.Series.value_counts))\n",
    "        t = df[['Const_Alt', col]].pivot(columns='Const_Alt', values=col).apply(pd.Series.value_counts).fillna(0)\n",
    "        print(t)\n",
    "        for cat in df[col].unique().tolist():\n",
    "            print(\"Cat : \", cat)\n",
    "            cat_const = t.loc[cat]['Const']\n",
    "            cat_alt = t.loc[cat]['Alt']\n",
    "            others_const = t.loc[~t.index.isin([cat])]['Const'].sum()\n",
    "            others_alt = t.loc[~t.index.isin([cat])]['Alt'].sum()\n",
    "            l.append(stats(cat_const, cat_alt, others_const, others_alt, col, cat, ))\n",
    "    return l\n",
    "\n",
    "\n",
    "def compute_odds_ratio_alt(df, cols):\n",
    "    l = list()\n",
    "    for col in cols:\n",
    "        print(col)\n",
    "        print(df[['Ratio_num_bins_update', col]].pivot(columns='Ratio_num_bins_update', values=col).apply(pd.Series.value_counts))\n",
    "        t = df[['Ratio_num_bins_update', col]].pivot(columns='Ratio_num_bins_update', values=col).apply(pd.Series.value_counts).fillna(0)\n",
    "        print(t)\n",
    "        for ratio in df['Ratio_num_bins_update'].unique().tolist():\n",
    "            print(ratio)\n",
    "            others_ratio = [c for c in df['Ratio_num_bins_update'].unique().tolist() if c != ratio]\n",
    "            for cat in df[col].unique().tolist():\n",
    "                print(\"Cat : \", cat)\n",
    "                cat_ratio = t.loc[cat][ratio]\n",
    "                cat_others_ratio = t.loc[cat][others_ratio]\n",
    "                others_ratio = t.loc[~t.index.isin([cat])][ratio].sum()\n",
    "                others_others_ratio = t.loc[~t.index.isin([cat])][others_ratio].sum()\n",
    "                l.append(stats(cat_const, cat_alt, others_const, others_alt, col, cat, ))\n",
    "    return l\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "differential-difference",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "most_severe_consequence\n",
      "Ratio_num_bins_update    NaN  0 - 0.2  0.2 - 0.4  0.4 - 0.6  0.6 - 0.8  \\\n",
      "missense_variant       180.0    29559      94338     222004     158495   \n",
      "start_lost               NaN      264        799       1198        773   \n",
      "stop_gained              4.0      979       3106       6725       4520   \n",
      "stop_lost                1.0      112        302        514        362   \n",
      "stop_retained_variant    NaN       39        104        163         98   \n",
      "synonymous_variant      96.0    12548      41484     100659      72074   \n",
      "\n",
      "Ratio_num_bins_update  0.8 - 0.99     1.00  \n",
      "missense_variant            35066  1296892  \n",
      "start_lost                    153     3105  \n",
      "stop_gained                  1067    34789  \n",
      "stop_lost                      61     1802  \n",
      "stop_retained_variant          17      606  \n",
      "synonymous_variant          15814   629551  \n",
      "Ratio_num_bins_update    NaN  0 - 0.2  0.2 - 0.4  0.4 - 0.6  0.6 - 0.8  \\\n",
      "missense_variant       180.0    29559      94338     222004     158495   \n",
      "start_lost               0.0      264        799       1198        773   \n",
      "stop_gained              4.0      979       3106       6725       4520   \n",
      "stop_lost                1.0      112        302        514        362   \n",
      "stop_retained_variant    0.0       39        104        163         98   \n",
      "synonymous_variant      96.0    12548      41484     100659      72074   \n",
      "\n",
      "Ratio_num_bins_update  0.8 - 0.99     1.00  \n",
      "missense_variant            35066  1296892  \n",
      "start_lost                    153     3105  \n",
      "stop_gained                  1067    34789  \n",
      "stop_lost                      61     1802  \n",
      "stop_retained_variant          17      606  \n",
      "synonymous_variant          15814   629551  \n",
      "Cat :  start_lost\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "1963640",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m----------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                 Traceback (most recent call last)",
      "\u001b[0;32m~/.conda/envs/ExoCarto/lib/python3.7/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   2894\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2895\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcasted_key\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2896\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 1963640",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-33-fe29af9b7895>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mcompute_odds_ratio_alt\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmerge\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgnomad\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrefseq_exons\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mon\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'MAP'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m'most_severe_consequence'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-32-8e52ccdbc7d7>\u001b[0m in \u001b[0;36mcompute_odds_ratio_alt\u001b[0;34m(df, cols)\u001b[0m\n\u001b[1;32m     46\u001b[0m                 \u001b[0mcat_others_ratio\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcat\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mothers_ratio\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     47\u001b[0m                 \u001b[0mothers_ratio\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m~\u001b[0m\u001b[0mt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0misin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcat\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mratio\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msum\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 48\u001b[0;31m                 \u001b[0mothers_others_ratio\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m~\u001b[0m\u001b[0mt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0misin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcat\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mothers_ratio\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msum\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     49\u001b[0m                 \u001b[0ml\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstats\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcat_const\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcat_alt\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mothers_const\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mothers_alt\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcol\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcat\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     50\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0ml\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.conda/envs/ExoCarto/lib/python3.7/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   2904\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnlevels\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2905\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem_multilevel\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2906\u001b[0;31m             \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2907\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mis_integer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2908\u001b[0m                 \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.conda/envs/ExoCarto/lib/python3.7/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   2895\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcasted_key\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2896\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2897\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2898\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2899\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mtolerance\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 1963640"
     ]
    }
   ],
   "source": [
    "compute_odds_ratio_alt(pd.merge(phylocsf_refseq, refseq_exons, on='MAP'), ['most_severe_consequence'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "outdoor-utilization",
   "metadata": {},
   "source": [
    "## RefSeq Miso exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "english-tracy",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>ranges</th>\n",
       "      <th>mRNA_exons</th>\n",
       "      <th>mRNA_nb</th>\n",
       "      <th>Share</th>\n",
       "      <th>Strand</th>\n",
       "      <th>mRNA_gene</th>\n",
       "      <th>mRNA_nb_total</th>\n",
       "      <th>Ratio</th>\n",
       "      <th>Ratio_num</th>\n",
       "      <th>Const_Alt</th>\n",
       "      <th>Ratio_num_bins</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Length</th>\n",
       "      <th>CDS_count</th>\n",
       "      <th>Ratio_num_bins_update</th>\n",
       "      <th>MAP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A2M</td>\n",
       "      <td>9220419-9220435</td>\n",
       "      <td>[NM_001347423, NM_001347425, NM_000014, NM_001...</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>[NM_001347423, NM_001347424, NM_000014, NM_001...</td>\n",
       "      <td>4</td>\n",
       "      <td>4/4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Const</td>\n",
       "      <td>0.8 - 1</td>\n",
       "      <td>9220419</td>\n",
       "      <td>9220435</td>\n",
       "      <td>16</td>\n",
       "      <td>38</td>\n",
       "      <td>1.00</td>\n",
       "      <td>A2M_9220419-9220435</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A2M</td>\n",
       "      <td>9220779-9220820</td>\n",
       "      <td>[NM_001347423, NM_001347425, NM_000014, NM_001...</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>[NM_001347423, NM_001347424, NM_000014, NM_001...</td>\n",
       "      <td>4</td>\n",
       "      <td>4/4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Const</td>\n",
       "      <td>0.8 - 1</td>\n",
       "      <td>9220779</td>\n",
       "      <td>9220820</td>\n",
       "      <td>41</td>\n",
       "      <td>38</td>\n",
       "      <td>1.00</td>\n",
       "      <td>A2M_9220779-9220820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A2M</td>\n",
       "      <td>9221336-9221438</td>\n",
       "      <td>[NM_001347423, NM_001347425, NM_000014, NM_001...</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>[NM_001347423, NM_001347424, NM_000014, NM_001...</td>\n",
       "      <td>4</td>\n",
       "      <td>4/4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Const</td>\n",
       "      <td>0.8 - 1</td>\n",
       "      <td>9221336</td>\n",
       "      <td>9221438</td>\n",
       "      <td>102</td>\n",
       "      <td>38</td>\n",
       "      <td>1.00</td>\n",
       "      <td>A2M_9221336-9221438</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A2M</td>\n",
       "      <td>9222341-9222409</td>\n",
       "      <td>[NM_001347423, NM_001347425, NM_000014, NM_001...</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>[NM_001347423, NM_001347424, NM_000014, NM_001...</td>\n",
       "      <td>4</td>\n",
       "      <td>4/4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Const</td>\n",
       "      <td>0.8 - 1</td>\n",
       "      <td>9222341</td>\n",
       "      <td>9222409</td>\n",
       "      <td>68</td>\n",
       "      <td>38</td>\n",
       "      <td>1.00</td>\n",
       "      <td>A2M_9222341-9222409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A2M</td>\n",
       "      <td>9223084-9223174</td>\n",
       "      <td>[NM_001347423, NM_001347425, NM_000014, NM_001...</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>[NM_001347423, NM_001347424, NM_000014, NM_001...</td>\n",
       "      <td>4</td>\n",
       "      <td>4/4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Const</td>\n",
       "      <td>0.8 - 1</td>\n",
       "      <td>9223084</td>\n",
       "      <td>9223174</td>\n",
       "      <td>90</td>\n",
       "      <td>38</td>\n",
       "      <td>1.00</td>\n",
       "      <td>A2M_9223084-9223174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178476</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>78047461-78047576</td>\n",
       "      <td>[NM_001308237, NM_015534]</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>[NM_001308237, NM_015534]</td>\n",
       "      <td>2</td>\n",
       "      <td>2/2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Const</td>\n",
       "      <td>0.8 - 1</td>\n",
       "      <td>78047461</td>\n",
       "      <td>78047576</td>\n",
       "      <td>115</td>\n",
       "      <td>12</td>\n",
       "      <td>1.00</td>\n",
       "      <td>ZZZ3_78047461-78047576</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178477</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>78047664-78047811</td>\n",
       "      <td>[NM_001308237, NM_015534]</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>[NM_001308237, NM_015534]</td>\n",
       "      <td>2</td>\n",
       "      <td>2/2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Const</td>\n",
       "      <td>0.8 - 1</td>\n",
       "      <td>78047664</td>\n",
       "      <td>78047811</td>\n",
       "      <td>147</td>\n",
       "      <td>12</td>\n",
       "      <td>1.00</td>\n",
       "      <td>ZZZ3_78047664-78047811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178478</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>78050202-78050340</td>\n",
       "      <td>[NM_001308237, NM_015534]</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>[NM_001308237, NM_015534]</td>\n",
       "      <td>2</td>\n",
       "      <td>2/2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Const</td>\n",
       "      <td>0.8 - 1</td>\n",
       "      <td>78050202</td>\n",
       "      <td>78050340</td>\n",
       "      <td>138</td>\n",
       "      <td>12</td>\n",
       "      <td>1.00</td>\n",
       "      <td>ZZZ3_78050202-78050340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178479</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>78097535-78099039</td>\n",
       "      <td>[NM_015534]</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>[NM_001308237, NM_015534]</td>\n",
       "      <td>2</td>\n",
       "      <td>1/2</td>\n",
       "      <td>0.5</td>\n",
       "      <td>Alt</td>\n",
       "      <td>0.4 - 0.6</td>\n",
       "      <td>78097535</td>\n",
       "      <td>78099039</td>\n",
       "      <td>1504</td>\n",
       "      <td>12</td>\n",
       "      <td>0.4 - 0.6</td>\n",
       "      <td>ZZZ3_78097535-78099039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178480</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>78105134-78105156</td>\n",
       "      <td>[NM_001308237]</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>[NM_001308237, NM_015534]</td>\n",
       "      <td>2</td>\n",
       "      <td>1/2</td>\n",
       "      <td>0.5</td>\n",
       "      <td>Alt</td>\n",
       "      <td>0.4 - 0.6</td>\n",
       "      <td>78105134</td>\n",
       "      <td>78105156</td>\n",
       "      <td>22</td>\n",
       "      <td>12</td>\n",
       "      <td>0.4 - 0.6</td>\n",
       "      <td>ZZZ3_78105134-78105156</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>178481 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Gene             ranges  \\\n",
       "0        A2M    9220419-9220435   \n",
       "1        A2M    9220779-9220820   \n",
       "2        A2M    9221336-9221438   \n",
       "3        A2M    9222341-9222409   \n",
       "4        A2M    9223084-9223174   \n",
       "...      ...                ...   \n",
       "178476  ZZZ3  78047461-78047576   \n",
       "178477  ZZZ3  78047664-78047811   \n",
       "178478  ZZZ3  78050202-78050340   \n",
       "178479  ZZZ3  78097535-78099039   \n",
       "178480  ZZZ3  78105134-78105156   \n",
       "\n",
       "                                               mRNA_exons  mRNA_nb  Share  \\\n",
       "0       [NM_001347423, NM_001347425, NM_000014, NM_001...        4  False   \n",
       "1       [NM_001347423, NM_001347425, NM_000014, NM_001...        4  False   \n",
       "2       [NM_001347423, NM_001347425, NM_000014, NM_001...        4  False   \n",
       "3       [NM_001347423, NM_001347425, NM_000014, NM_001...        4  False   \n",
       "4       [NM_001347423, NM_001347425, NM_000014, NM_001...        4  False   \n",
       "...                                                   ...      ...    ...   \n",
       "178476                          [NM_001308237, NM_015534]        2  False   \n",
       "178477                          [NM_001308237, NM_015534]        2  False   \n",
       "178478                          [NM_001308237, NM_015534]        2  False   \n",
       "178479                                        [NM_015534]        1  False   \n",
       "178480                                     [NM_001308237]        1  False   \n",
       "\n",
       "        Strand                                          mRNA_gene  \\\n",
       "0            0  [NM_001347423, NM_001347424, NM_000014, NM_001...   \n",
       "1            0  [NM_001347423, NM_001347424, NM_000014, NM_001...   \n",
       "2            0  [NM_001347423, NM_001347424, NM_000014, NM_001...   \n",
       "3            0  [NM_001347423, NM_001347424, NM_000014, NM_001...   \n",
       "4            0  [NM_001347423, NM_001347424, NM_000014, NM_001...   \n",
       "...        ...                                                ...   \n",
       "178476       0                          [NM_001308237, NM_015534]   \n",
       "178477       0                          [NM_001308237, NM_015534]   \n",
       "178478       0                          [NM_001308237, NM_015534]   \n",
       "178479       0                          [NM_001308237, NM_015534]   \n",
       "178480       0                          [NM_001308237, NM_015534]   \n",
       "\n",
       "        mRNA_nb_total Ratio  Ratio_num Const_Alt Ratio_num_bins     Start  \\\n",
       "0                   4   4/4        1.0     Const        0.8 - 1   9220419   \n",
       "1                   4   4/4        1.0     Const        0.8 - 1   9220779   \n",
       "2                   4   4/4        1.0     Const        0.8 - 1   9221336   \n",
       "3                   4   4/4        1.0     Const        0.8 - 1   9222341   \n",
       "4                   4   4/4        1.0     Const        0.8 - 1   9223084   \n",
       "...               ...   ...        ...       ...            ...       ...   \n",
       "178476              2   2/2        1.0     Const        0.8 - 1  78047461   \n",
       "178477              2   2/2        1.0     Const        0.8 - 1  78047664   \n",
       "178478              2   2/2        1.0     Const        0.8 - 1  78050202   \n",
       "178479              2   1/2        0.5       Alt      0.4 - 0.6  78097535   \n",
       "178480              2   1/2        0.5       Alt      0.4 - 0.6  78105134   \n",
       "\n",
       "             End  Length  CDS_count Ratio_num_bins_update  \\\n",
       "0        9220435      16         38                  1.00   \n",
       "1        9220820      41         38                  1.00   \n",
       "2        9221438     102         38                  1.00   \n",
       "3        9222409      68         38                  1.00   \n",
       "4        9223174      90         38                  1.00   \n",
       "...          ...     ...        ...                   ...   \n",
       "178476  78047576     115         12                  1.00   \n",
       "178477  78047811     147         12                  1.00   \n",
       "178478  78050340     138         12                  1.00   \n",
       "178479  78099039    1504         12             0.4 - 0.6   \n",
       "178480  78105156      22         12             0.4 - 0.6   \n",
       "\n",
       "                           MAP  \n",
       "0          A2M_9220419-9220435  \n",
       "1          A2M_9220779-9220820  \n",
       "2          A2M_9221336-9221438  \n",
       "3          A2M_9222341-9222409  \n",
       "4          A2M_9223084-9223174  \n",
       "...                        ...  \n",
       "178476  ZZZ3_78047461-78047576  \n",
       "178477  ZZZ3_78047664-78047811  \n",
       "178478  ZZZ3_78050202-78050340  \n",
       "178479  ZZZ3_78097535-78099039  \n",
       "178480  ZZZ3_78105134-78105156  \n",
       "\n",
       "[178481 rows x 18 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "refseq_exons = pd.read_parquet('/gstock/EXOTIC/data/GENOMICS/refseq_37_corrected_recomputed.parquet')\n",
    "refseq_exons.loc[(refseq_exons['Const_Alt'] == 'Alt') & (refseq_exons['Ratio_num_bins'] == '0.8 - 1'), 'Ratio_num_bins_update'] = '0.8 - 0.99'\n",
    "refseq_exons.loc[(refseq_exons['Const_Alt'] == 'Alt') & (refseq_exons['Ratio_num_bins'] != '0.8 - 1'), 'Ratio_num_bins_update'] = refseq_exons.loc[(refseq_exons['Const_Alt'] == 'Alt') & (refseq_exons['Ratio_num_bins'] != '0.8 - 1')]['Ratio_num_bins']\n",
    "refseq_exons.loc[(refseq_exons['Const_Alt'] == 'Const') & (refseq_exons['Ratio_num_bins'] == '0.8 - 1'), 'Ratio_num_bins_update'] = '1.00'\n",
    "refseq_exons['MAP'] = refseq_exons['Gene'] + '_' + refseq_exons['ranges']\n",
    "refseq_exons"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "distributed-outside",
   "metadata": {},
   "source": [
    "## PhyloCSF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "expanded-square",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>interval</th>\n",
       "      <th>phyloCSF_score</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>MAP</th>\n",
       "      <th>Gene</th>\n",
       "      <th>phyloCSF_cat</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>57660</th>\n",
       "      <td>[10:101089145-101090717)</td>\n",
       "      <td>5643.7434</td>\n",
       "      <td>10:101089145</td>\n",
       "      <td>10:101090717</td>\n",
       "      <td>CNNM1_101089145-101090717</td>\n",
       "      <td>CNNM1</td>\n",
       "      <td>Conserved</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57662</th>\n",
       "      <td>[10:101089691-101090717)</td>\n",
       "      <td>3729.0848</td>\n",
       "      <td>10:101089691</td>\n",
       "      <td>10:101090717</td>\n",
       "      <td>CNNM1_101089145-101090717</td>\n",
       "      <td>CNNM1</td>\n",
       "      <td>Conserved</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57663</th>\n",
       "      <td>[10:101090240-101090717)</td>\n",
       "      <td>2077.2201</td>\n",
       "      <td>10:101090240</td>\n",
       "      <td>10:101090717</td>\n",
       "      <td>CNNM1_101089145-101090717</td>\n",
       "      <td>CNNM1</td>\n",
       "      <td>Conserved</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57682</th>\n",
       "      <td>[10:101458291-101458615)</td>\n",
       "      <td>1293.5257</td>\n",
       "      <td>10:101458291</td>\n",
       "      <td>10:101458615</td>\n",
       "      <td>ENTPD7_101458291-101458615</td>\n",
       "      <td>ENTPD7</td>\n",
       "      <td>Conserved</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57694</th>\n",
       "      <td>[10:101639568-101640118)</td>\n",
       "      <td>1125.4920</td>\n",
       "      <td>10:101639568</td>\n",
       "      <td>10:101640118</td>\n",
       "      <td>DNMBP_101639568-101640118</td>\n",
       "      <td>DNMBP</td>\n",
       "      <td>Conserved</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110044</th>\n",
       "      <td>[Y:16734000-16734471)</td>\n",
       "      <td>1757.4552</td>\n",
       "      <td>Y:16734000</td>\n",
       "      <td>Y:16734471</td>\n",
       "      <td>NLGN4Y_16734000-16734471</td>\n",
       "      <td>NLGN4Y</td>\n",
       "      <td>Conserved</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110045</th>\n",
       "      <td>[Y:16734061-16734471)</td>\n",
       "      <td>1652.5963</td>\n",
       "      <td>Y:16734061</td>\n",
       "      <td>Y:16734471</td>\n",
       "      <td>NLGN4Y_16734000-16734471</td>\n",
       "      <td>NLGN4Y</td>\n",
       "      <td>Conserved</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110049</th>\n",
       "      <td>[Y:16941610-16942399)</td>\n",
       "      <td>3028.8020</td>\n",
       "      <td>Y:16941610</td>\n",
       "      <td>Y:16942399</td>\n",
       "      <td>NLGN4Y_16941610-16942399</td>\n",
       "      <td>NLGN4Y</td>\n",
       "      <td>Conserved</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109991</th>\n",
       "      <td>[Y:2829115-2829687)</td>\n",
       "      <td>1279.9484</td>\n",
       "      <td>Y:2829115</td>\n",
       "      <td>Y:2829687</td>\n",
       "      <td>ZFY_2829115-2829687</td>\n",
       "      <td>ZFY</td>\n",
       "      <td>Conserved</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109992</th>\n",
       "      <td>[Y:2829132-2829687)</td>\n",
       "      <td>1290.2635</td>\n",
       "      <td>Y:2829132</td>\n",
       "      <td>Y:2829687</td>\n",
       "      <td>ZFY_2829115-2829687</td>\n",
       "      <td>ZFY</td>\n",
       "      <td>Conserved</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9985 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        interval  phyloCSF_score         start           end  \\\n",
       "57660   [10:101089145-101090717)       5643.7434  10:101089145  10:101090717   \n",
       "57662   [10:101089691-101090717)       3729.0848  10:101089691  10:101090717   \n",
       "57663   [10:101090240-101090717)       2077.2201  10:101090240  10:101090717   \n",
       "57682   [10:101458291-101458615)       1293.5257  10:101458291  10:101458615   \n",
       "57694   [10:101639568-101640118)       1125.4920  10:101639568  10:101640118   \n",
       "...                          ...             ...           ...           ...   \n",
       "110044     [Y:16734000-16734471)       1757.4552    Y:16734000    Y:16734471   \n",
       "110045     [Y:16734061-16734471)       1652.5963    Y:16734061    Y:16734471   \n",
       "110049     [Y:16941610-16942399)       3028.8020    Y:16941610    Y:16942399   \n",
       "109991       [Y:2829115-2829687)       1279.9484     Y:2829115     Y:2829687   \n",
       "109992       [Y:2829132-2829687)       1290.2635     Y:2829132     Y:2829687   \n",
       "\n",
       "                               MAP    Gene phyloCSF_cat  \n",
       "57660    CNNM1_101089145-101090717   CNNM1    Conserved  \n",
       "57662    CNNM1_101089145-101090717   CNNM1    Conserved  \n",
       "57663    CNNM1_101089145-101090717   CNNM1    Conserved  \n",
       "57682   ENTPD7_101458291-101458615  ENTPD7    Conserved  \n",
       "57694    DNMBP_101639568-101640118   DNMBP    Conserved  \n",
       "...                            ...     ...          ...  \n",
       "110044    NLGN4Y_16734000-16734471  NLGN4Y    Conserved  \n",
       "110045    NLGN4Y_16734000-16734471  NLGN4Y    Conserved  \n",
       "110049    NLGN4Y_16941610-16942399  NLGN4Y    Conserved  \n",
       "109991         ZFY_2829115-2829687     ZFY    Conserved  \n",
       "109992         ZFY_2829115-2829687     ZFY    Conserved  \n",
       "\n",
       "[9985 rows x 7 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phylocsf_refseq = pd.read_parquet(\"/gstock/EXOTIC/data/EXON_PROPERTIES/refseq_phylocsf.parquet\").rename({'target' : 'phyloCSF_score'}, axis=1)\n",
    "phylocsf_refseq.loc[(phylocsf_refseq['phyloCSF_score'] <= -100), 'phyloCSF_cat'] = 'Unconserved'\n",
    "phylocsf_refseq.loc[(phylocsf_refseq['phyloCSF_score'] > 1000), 'phyloCSF_cat'] = 'Conserved'\n",
    "phylocsf_refseq.loc[(phylocsf_refseq['phyloCSF_score'] >= -100) & (phylocsf_refseq['phyloCSF_score'] < 1000), 'phyloCSF_cat'] = 'Standard'\n",
    "phylocsf_refseq = phylocsf_refseq.fillna(False)\n",
    "phylocsf_refseq = phylocsf_refseq.loc[phylocsf_refseq['phyloCSF_cat'] != 'Standard']\n",
    "phylocsf_refseq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "wrong-roller",
   "metadata": {},
   "source": [
    "## gnomAD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "approved-spare",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>locus</th>\n",
       "      <th>alleles</th>\n",
       "      <th>AF</th>\n",
       "      <th>rsid</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>end</th>\n",
       "      <th>id</th>\n",
       "      <th>most_severe_consequence</th>\n",
       "      <th>MAP</th>\n",
       "      <th>Gene</th>\n",
       "      <th>AF_bins</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:901913</td>\n",
       "      <td>[\"T\",\"G\"]</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>rs759148200</td>\n",
       "      <td>snv</td>\n",
       "      <td>901913</td>\n",
       "      <td>1_901913_T_G</td>\n",
       "      <td>start_lost</td>\n",
       "      <td>PLEKHN1_901912-901994</td>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>1e-05 - 0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:901916</td>\n",
       "      <td>[\"G\",\"A\"]</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>rs764917347</td>\n",
       "      <td>snv</td>\n",
       "      <td>901916</td>\n",
       "      <td>1_901916_G_A</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>PLEKHN1_901912-901994</td>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>1e-06 - 1e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:901917</td>\n",
       "      <td>[\"G\",\"A\"]</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>rs774961129</td>\n",
       "      <td>snv</td>\n",
       "      <td>901917</td>\n",
       "      <td>1_901917_G_A</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>PLEKHN1_901912-901994</td>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>1e-06 - 1e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:901919</td>\n",
       "      <td>[\"A\",\"C\"]</td>\n",
       "      <td>0.000226</td>\n",
       "      <td>rs763342668</td>\n",
       "      <td>snv</td>\n",
       "      <td>901919</td>\n",
       "      <td>1_901919_A_C</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>PLEKHN1_901912-901994</td>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>0.0001 - 0.001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:901920</td>\n",
       "      <td>[\"C\",\"G\"]</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>rs1408404600</td>\n",
       "      <td>snv</td>\n",
       "      <td>901920</td>\n",
       "      <td>1_901920_C_G</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>PLEKHN1_901912-901994</td>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>1e-06 - 1e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2810848</th>\n",
       "      <td>Y:22749975</td>\n",
       "      <td>[\"C\",\"T\"]</td>\n",
       "      <td>0.000017</td>\n",
       "      <td>rs746588234</td>\n",
       "      <td>snv</td>\n",
       "      <td>22749975</td>\n",
       "      <td>Y_22749975_C_T</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>EIF1AY_22749910-22749991</td>\n",
       "      <td>EIF1AY</td>\n",
       "      <td>1e-05 - 0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2810849</th>\n",
       "      <td>Y:22751395</td>\n",
       "      <td>[\"T\",\"C\"]</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>rs1382594808</td>\n",
       "      <td>snv</td>\n",
       "      <td>22751395</td>\n",
       "      <td>Y_22751395_T_C</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>EIF1AY_22751370-22751461</td>\n",
       "      <td>EIF1AY</td>\n",
       "      <td>1e-05 - 0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2810850</th>\n",
       "      <td>Y:22751399</td>\n",
       "      <td>[\"C\",\"G\"]</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>rs762602059</td>\n",
       "      <td>snv</td>\n",
       "      <td>22751399</td>\n",
       "      <td>Y_22751399_C_G</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>EIF1AY_22751370-22751461</td>\n",
       "      <td>EIF1AY</td>\n",
       "      <td>1e-05 - 0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2810851</th>\n",
       "      <td>Y:22751428</td>\n",
       "      <td>[\"C\",\"T\"]</td>\n",
       "      <td>0.000030</td>\n",
       "      <td>rs1158365540</td>\n",
       "      <td>snv</td>\n",
       "      <td>22751428</td>\n",
       "      <td>Y_22751428_C_T</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>EIF1AY_22751370-22751461</td>\n",
       "      <td>EIF1AY</td>\n",
       "      <td>1e-05 - 0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2810852</th>\n",
       "      <td>Y:22751443</td>\n",
       "      <td>[\"T\",\"C\"]</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>rs772483555</td>\n",
       "      <td>snv</td>\n",
       "      <td>22751443</td>\n",
       "      <td>Y_22751443_T_C</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>EIF1AY_22751370-22751461</td>\n",
       "      <td>EIF1AY</td>\n",
       "      <td>1e-05 - 0.0001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2770423 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              locus    alleles        AF          rsid variant_type       end  \\\n",
       "0          1:901913  [\"T\",\"G\"]  0.000012   rs759148200          snv    901913   \n",
       "1          1:901916  [\"G\",\"A\"]  0.000004   rs764917347          snv    901916   \n",
       "2          1:901917  [\"G\",\"A\"]  0.000004   rs774961129          snv    901917   \n",
       "3          1:901919  [\"A\",\"C\"]  0.000226   rs763342668          snv    901919   \n",
       "4          1:901920  [\"C\",\"G\"]  0.000004  rs1408404600          snv    901920   \n",
       "...             ...        ...       ...           ...          ...       ...   \n",
       "2810848  Y:22749975  [\"C\",\"T\"]  0.000017   rs746588234          snv  22749975   \n",
       "2810849  Y:22751395  [\"T\",\"C\"]  0.000015  rs1382594808          snv  22751395   \n",
       "2810850  Y:22751399  [\"C\",\"G\"]  0.000015   rs762602059          snv  22751399   \n",
       "2810851  Y:22751428  [\"C\",\"T\"]  0.000030  rs1158365540          snv  22751428   \n",
       "2810852  Y:22751443  [\"T\",\"C\"]  0.000015   rs772483555          snv  22751443   \n",
       "\n",
       "                     id most_severe_consequence                       MAP  \\\n",
       "0          1_901913_T_G              start_lost     PLEKHN1_901912-901994   \n",
       "1          1_901916_G_A        missense_variant     PLEKHN1_901912-901994   \n",
       "2          1_901917_G_A      synonymous_variant     PLEKHN1_901912-901994   \n",
       "3          1_901919_A_C        missense_variant     PLEKHN1_901912-901994   \n",
       "4          1_901920_C_G        missense_variant     PLEKHN1_901912-901994   \n",
       "...                 ...                     ...                       ...   \n",
       "2810848  Y_22749975_C_T      synonymous_variant  EIF1AY_22749910-22749991   \n",
       "2810849  Y_22751395_T_C      synonymous_variant  EIF1AY_22751370-22751461   \n",
       "2810850  Y_22751399_C_G        missense_variant  EIF1AY_22751370-22751461   \n",
       "2810851  Y_22751428_C_T      synonymous_variant  EIF1AY_22751370-22751461   \n",
       "2810852  Y_22751443_T_C      synonymous_variant  EIF1AY_22751370-22751461   \n",
       "\n",
       "            Gene         AF_bins  \n",
       "0        PLEKHN1  1e-05 - 0.0001  \n",
       "1        PLEKHN1   1e-06 - 1e-05  \n",
       "2        PLEKHN1   1e-06 - 1e-05  \n",
       "3        PLEKHN1  0.0001 - 0.001  \n",
       "4        PLEKHN1   1e-06 - 1e-05  \n",
       "...          ...             ...  \n",
       "2810848   EIF1AY  1e-05 - 0.0001  \n",
       "2810849   EIF1AY  1e-05 - 0.0001  \n",
       "2810850   EIF1AY  1e-05 - 0.0001  \n",
       "2810851   EIF1AY  1e-05 - 0.0001  \n",
       "2810852   EIF1AY  1e-05 - 0.0001  \n",
       "\n",
       "[2770423 rows x 11 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gnomad = pd.read_parquet('/gstock/EXOTIC/data/EXON_PROPERTIES/refseq_miso_gnomad_variations.parquet')\n",
    "bins_af = list(reversed([1/(10 ** e) for e in list(range(0,7,1))]))\n",
    "\n",
    "def convert_bins_into_labels(bins):\n",
    "    return [\" - \".join([str(bins[j]), str(bins[j+1])]) for j, e in enumerate(bins) if j <len(bins)-1]\n",
    "\n",
    "gnomad['AF'] = gnomad['AF'].str.replace(',', '.').astype(float)\n",
    "gnomad['AF_bins'] = pd.cut(gnomad['AF'], bins=bins_af, labels=convert_bins_into_labels(bins_af))\n",
    "mc = [\"missense_variant\", \"start_lost\", \"stop_gained\", \"stop_lost\", \"stop_retained_variant\", \"synonymous_variant\", ]\n",
    "gnomad = gnomad.loc[gnomad['most_severe_consequence'].isin(mc)]\n",
    "gnomad"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "muslim-antarctica",
   "metadata": {},
   "source": [
    "## ClinVar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "hungarian-delay",
   "metadata": {},
   "outputs": [],
   "source": [
    "clinvar = pd.read_parquet('/gstock/EXOTIC/data/EXON_PROPERTIES/refseq_clinvar_pathogenic.parquet')\n",
    "clinvar['id'] = clinvar['locus.contig'] + '_' + clinvar['locus.position'].astype(str) + '_' + clinvar['alleles'].apply(lambda r: \"_\".join(r))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "clinical-night",
   "metadata": {},
   "source": [
    "## CCR data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "prospective-exhibit",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>interval</th>\n",
       "      <th>target</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>MAP</th>\n",
       "      <th>Gene</th>\n",
       "      <th>CCR_bins</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1953387</th>\n",
       "      <td>[10:100003848-100003856)</td>\n",
       "      <td>63.055177</td>\n",
       "      <td>10:100003848</td>\n",
       "      <td>10:100003856</td>\n",
       "      <td>R3HCC1L_100003848-100003915</td>\n",
       "      <td>R3HCC1L</td>\n",
       "      <td>20 - 80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1953388</th>\n",
       "      <td>[10:100003856-100003857)</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>10:100003856</td>\n",
       "      <td>10:100003857</td>\n",
       "      <td>R3HCC1L_100003848-100003915</td>\n",
       "      <td>R3HCC1L</td>\n",
       "      <td>0 - 20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1907621</th>\n",
       "      <td>[10:100003857-100003858)</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>10:100003857</td>\n",
       "      <td>10:100003858</td>\n",
       "      <td>R3HCC1L_100003848-100003915</td>\n",
       "      <td>R3HCC1L</td>\n",
       "      <td>0 - 20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1953390</th>\n",
       "      <td>[10:100003858-100003874)</td>\n",
       "      <td>71.921144</td>\n",
       "      <td>10:100003858</td>\n",
       "      <td>10:100003874</td>\n",
       "      <td>R3HCC1L_100003848-100003915</td>\n",
       "      <td>R3HCC1L</td>\n",
       "      <td>20 - 80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1907623</th>\n",
       "      <td>[10:100003874-100003875)</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>10:100003874</td>\n",
       "      <td>10:100003875</td>\n",
       "      <td>R3HCC1L_100003848-100003915</td>\n",
       "      <td>R3HCC1L</td>\n",
       "      <td>0 - 20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3526305</th>\n",
       "      <td>[X:99956953-99956958)</td>\n",
       "      <td>8.932019</td>\n",
       "      <td>X:99956953</td>\n",
       "      <td>X:99956958</td>\n",
       "      <td>SYTL4_99956924-99957033</td>\n",
       "      <td>SYTL4</td>\n",
       "      <td>0 - 20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3526306</th>\n",
       "      <td>[X:99956958-99956959)</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>X:99956958</td>\n",
       "      <td>X:99956959</td>\n",
       "      <td>SYTL4_99956924-99957033</td>\n",
       "      <td>SYTL4</td>\n",
       "      <td>0 - 20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3526307</th>\n",
       "      <td>[X:99956959-99957026)</td>\n",
       "      <td>90.007142</td>\n",
       "      <td>X:99956959</td>\n",
       "      <td>X:99957026</td>\n",
       "      <td>SYTL4_99956924-99957033</td>\n",
       "      <td>SYTL4</td>\n",
       "      <td>90 - 95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3611490</th>\n",
       "      <td>[X:99957026-99957027)</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>X:99957026</td>\n",
       "      <td>X:99957027</td>\n",
       "      <td>SYTL4_99956924-99957033</td>\n",
       "      <td>SYTL4</td>\n",
       "      <td>0 - 20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3611491</th>\n",
       "      <td>[X:99957027-99957034)</td>\n",
       "      <td>0.514571</td>\n",
       "      <td>X:99957027</td>\n",
       "      <td>X:99957034</td>\n",
       "      <td>SYTL4_99956924-99957033</td>\n",
       "      <td>SYTL4</td>\n",
       "      <td>0 - 20</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3643388 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         interval     target         start           end  \\\n",
       "1953387  [10:100003848-100003856)  63.055177  10:100003848  10:100003856   \n",
       "1953388  [10:100003856-100003857)   0.000000  10:100003856  10:100003857   \n",
       "1907621  [10:100003857-100003858)   0.000000  10:100003857  10:100003858   \n",
       "1953390  [10:100003858-100003874)  71.921144  10:100003858  10:100003874   \n",
       "1907623  [10:100003874-100003875)   0.000000  10:100003874  10:100003875   \n",
       "...                           ...        ...           ...           ...   \n",
       "3526305     [X:99956953-99956958)   8.932019    X:99956953    X:99956958   \n",
       "3526306     [X:99956958-99956959)   0.000000    X:99956958    X:99956959   \n",
       "3526307     [X:99956959-99957026)  90.007142    X:99956959    X:99957026   \n",
       "3611490     [X:99957026-99957027)   0.000000    X:99957026    X:99957027   \n",
       "3611491     [X:99957027-99957034)   0.514571    X:99957027    X:99957034   \n",
       "\n",
       "                                 MAP     Gene CCR_bins  \n",
       "1953387  R3HCC1L_100003848-100003915  R3HCC1L  20 - 80  \n",
       "1953388  R3HCC1L_100003848-100003915  R3HCC1L   0 - 20  \n",
       "1907621  R3HCC1L_100003848-100003915  R3HCC1L   0 - 20  \n",
       "1953390  R3HCC1L_100003848-100003915  R3HCC1L  20 - 80  \n",
       "1907623  R3HCC1L_100003848-100003915  R3HCC1L   0 - 20  \n",
       "...                              ...      ...      ...  \n",
       "3526305      SYTL4_99956924-99957033    SYTL4   0 - 20  \n",
       "3526306      SYTL4_99956924-99957033    SYTL4   0 - 20  \n",
       "3526307      SYTL4_99956924-99957033    SYTL4  90 - 95  \n",
       "3611490      SYTL4_99956924-99957033    SYTL4   0 - 20  \n",
       "3611491      SYTL4_99956924-99957033    SYTL4   0 - 20  \n",
       "\n",
       "[3643388 rows x 7 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ccr = pd.read_parquet('/gstock/EXOTIC/data/EXON_PROPERTIES/refseq_ccr_start_end.parquet')\n",
    "bins = [0,20,80,90,95,100]\n",
    "ccr['CCR_bins'] = pd.cut(ccr['target'], bins, labels=[\" - \".join([str(bins[j]), str(bins[j+1])]) for j, e in enumerate(bins) if j <len(bins)-1], include_lowest=True)\n",
    "ccr"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "deluxe-movie",
   "metadata": {},
   "source": [
    "## Merge gnomAD & ClinVar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "interracial-isaac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>locus</th>\n",
       "      <th>alleles</th>\n",
       "      <th>AF</th>\n",
       "      <th>rsid</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>end</th>\n",
       "      <th>id</th>\n",
       "      <th>most_severe_consequence</th>\n",
       "      <th>MAP</th>\n",
       "      <th>Gene</th>\n",
       "      <th>AF_bins</th>\n",
       "      <th>locus.contig</th>\n",
       "      <th>locus.position</th>\n",
       "      <th>ALLELEID</th>\n",
       "      <th>CLNREVSTAT</th>\n",
       "      <th>CLNSIG</th>\n",
       "      <th>GENEINFO</th>\n",
       "      <th>RS</th>\n",
       "      <th>RS_STARS</th>\n",
       "      <th>Status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:901913</td>\n",
       "      <td>[\"T\",\"G\"]</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>rs759148200</td>\n",
       "      <td>snv</td>\n",
       "      <td>901913.0</td>\n",
       "      <td>1_901913_T_G</td>\n",
       "      <td>start_lost</td>\n",
       "      <td>PLEKHN1_901912-901994</td>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>1e-05 - 0.0001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:901916</td>\n",
       "      <td>[\"G\",\"A\"]</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>rs764917347</td>\n",
       "      <td>snv</td>\n",
       "      <td>901916.0</td>\n",
       "      <td>1_901916_G_A</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>PLEKHN1_901912-901994</td>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>1e-06 - 1e-05</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:901917</td>\n",
       "      <td>[\"G\",\"A\"]</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>rs774961129</td>\n",
       "      <td>snv</td>\n",
       "      <td>901917.0</td>\n",
       "      <td>1_901917_G_A</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>PLEKHN1_901912-901994</td>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>1e-06 - 1e-05</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:901919</td>\n",
       "      <td>[\"A\",\"C\"]</td>\n",
       "      <td>0.000226</td>\n",
       "      <td>rs763342668</td>\n",
       "      <td>snv</td>\n",
       "      <td>901919.0</td>\n",
       "      <td>1_901919_A_C</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>PLEKHN1_901912-901994</td>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>0.0001 - 0.001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:901920</td>\n",
       "      <td>[\"C\",\"G\"]</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>rs1408404600</td>\n",
       "      <td>snv</td>\n",
       "      <td>901920.0</td>\n",
       "      <td>1_901920_C_G</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>PLEKHN1_901912-901994</td>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>1e-06 - 1e-05</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256896</th>\n",
       "      <td>NaN</td>\n",
       "      <td>[C, A]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10156</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>X_154250707_C_A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F8_154250685-154250827</td>\n",
       "      <td>F8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>X</td>\n",
       "      <td>154250707.0</td>\n",
       "      <td>25195.0</td>\n",
       "      <td>no_assertion_criteria_provided</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>F8:2157</td>\n",
       "      <td>[137852379]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256897</th>\n",
       "      <td>NaN</td>\n",
       "      <td>[T, C]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10352</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>X_154250724_T_C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F8_154250685-154250827</td>\n",
       "      <td>F8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>X</td>\n",
       "      <td>154250724.0</td>\n",
       "      <td>25391.0</td>\n",
       "      <td>no_assertion_criteria_provided</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>F8:2157</td>\n",
       "      <td>[137852476]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256899</th>\n",
       "      <td>NaN</td>\n",
       "      <td>[T, A]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10154</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>X_154250739_T_A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F8_154250685-154250827</td>\n",
       "      <td>F8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>X</td>\n",
       "      <td>154250739.0</td>\n",
       "      <td>25193.0</td>\n",
       "      <td>no_assertion_criteria_provided</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>F8:2157</td>\n",
       "      <td>[137852378]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256900</th>\n",
       "      <td>NaN</td>\n",
       "      <td>[A, C]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10153</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>X_154250751_A_C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F8_154250685-154250827</td>\n",
       "      <td>F8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>X</td>\n",
       "      <td>154250751.0</td>\n",
       "      <td>25192.0</td>\n",
       "      <td>no_assertion_criteria_provided</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>F8:2157</td>\n",
       "      <td>[137852377]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256902</th>\n",
       "      <td>NaN</td>\n",
       "      <td>[G, A]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10152</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>X_154250785_G_A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F8_154250685-154250827</td>\n",
       "      <td>F8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>X</td>\n",
       "      <td>154250785.0</td>\n",
       "      <td>25191.0</td>\n",
       "      <td>criteria_provided,_single_submitter</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>F8:2157</td>\n",
       "      <td>[387906432]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2789760 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           locus    alleles        AF          rsid variant_type       end  \\\n",
       "0       1:901913  [\"T\",\"G\"]  0.000012   rs759148200          snv  901913.0   \n",
       "1       1:901916  [\"G\",\"A\"]  0.000004   rs764917347          snv  901916.0   \n",
       "2       1:901917  [\"G\",\"A\"]  0.000004   rs774961129          snv  901917.0   \n",
       "3       1:901919  [\"A\",\"C\"]  0.000226   rs763342668          snv  901919.0   \n",
       "4       1:901920  [\"C\",\"G\"]  0.000004  rs1408404600          snv  901920.0   \n",
       "...          ...        ...       ...           ...          ...       ...   \n",
       "256896       NaN     [C, A]       NaN         10156          NaN       NaN   \n",
       "256897       NaN     [T, C]       NaN         10352          NaN       NaN   \n",
       "256899       NaN     [T, A]       NaN         10154          NaN       NaN   \n",
       "256900       NaN     [A, C]       NaN         10153          NaN       NaN   \n",
       "256902       NaN     [G, A]       NaN         10152          NaN       NaN   \n",
       "\n",
       "                     id most_severe_consequence                     MAP  \\\n",
       "0          1_901913_T_G              start_lost   PLEKHN1_901912-901994   \n",
       "1          1_901916_G_A        missense_variant   PLEKHN1_901912-901994   \n",
       "2          1_901917_G_A      synonymous_variant   PLEKHN1_901912-901994   \n",
       "3          1_901919_A_C        missense_variant   PLEKHN1_901912-901994   \n",
       "4          1_901920_C_G        missense_variant   PLEKHN1_901912-901994   \n",
       "...                 ...                     ...                     ...   \n",
       "256896  X_154250707_C_A                     NaN  F8_154250685-154250827   \n",
       "256897  X_154250724_T_C                     NaN  F8_154250685-154250827   \n",
       "256899  X_154250739_T_A                     NaN  F8_154250685-154250827   \n",
       "256900  X_154250751_A_C                     NaN  F8_154250685-154250827   \n",
       "256902  X_154250785_G_A                     NaN  F8_154250685-154250827   \n",
       "\n",
       "           Gene         AF_bins locus.contig  locus.position  ALLELEID  \\\n",
       "0       PLEKHN1  1e-05 - 0.0001          NaN             NaN       NaN   \n",
       "1       PLEKHN1   1e-06 - 1e-05          NaN             NaN       NaN   \n",
       "2       PLEKHN1   1e-06 - 1e-05          NaN             NaN       NaN   \n",
       "3       PLEKHN1  0.0001 - 0.001          NaN             NaN       NaN   \n",
       "4       PLEKHN1   1e-06 - 1e-05          NaN             NaN       NaN   \n",
       "...         ...             ...          ...             ...       ...   \n",
       "256896       F8             NaN            X     154250707.0   25195.0   \n",
       "256897       F8             NaN            X     154250724.0   25391.0   \n",
       "256899       F8             NaN            X     154250739.0   25193.0   \n",
       "256900       F8             NaN            X     154250751.0   25192.0   \n",
       "256902       F8             NaN            X     154250785.0   25191.0   \n",
       "\n",
       "                                 CLNREVSTAT      CLNSIG GENEINFO           RS  \\\n",
       "0                                       NaN         NaN      NaN          NaN   \n",
       "1                                       NaN         NaN      NaN          NaN   \n",
       "2                                       NaN         NaN      NaN          NaN   \n",
       "3                                       NaN         NaN      NaN          NaN   \n",
       "4                                       NaN         NaN      NaN          NaN   \n",
       "...                                     ...         ...      ...          ...   \n",
       "256896       no_assertion_criteria_provided  Pathogenic  F8:2157  [137852379]   \n",
       "256897       no_assertion_criteria_provided  Pathogenic  F8:2157  [137852476]   \n",
       "256899       no_assertion_criteria_provided  Pathogenic  F8:2157  [137852378]   \n",
       "256900       no_assertion_criteria_provided  Pathogenic  F8:2157  [137852377]   \n",
       "256902  criteria_provided,_single_submitter  Pathogenic  F8:2157  [387906432]   \n",
       "\n",
       "        RS_STARS      Status  \n",
       "0            NaN      Benign  \n",
       "1            NaN      Benign  \n",
       "2            NaN      Benign  \n",
       "3            NaN      Benign  \n",
       "4            NaN      Benign  \n",
       "...          ...         ...  \n",
       "256896       0.0  Pathogenic  \n",
       "256897       0.0  Pathogenic  \n",
       "256899       0.0  Pathogenic  \n",
       "256900       0.0  Pathogenic  \n",
       "256902       1.0  Pathogenic  \n",
       "\n",
       "[2789760 rows x 20 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merge_variants = pd.concat(\n",
    "    [\n",
    "        gnomad.loc[~gnomad['id'].isin(clinvar.id.values.tolist())],\n",
    "        clinvar.loc[~clinvar['id'].isin(gnomad.id.values.tolist())]\n",
    "    ]\n",
    ")\n",
    "merge_variants.loc[merge_variants['CLNSIG'].isna() == True, 'Status'] = 'Benign'\n",
    "merge_variants['Status'] = merge_variants['Status'].fillna('Pathogenic')\n",
    "merge_variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "spread-twist",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "phyloCSF_cat\n",
      "Const_Alt    NaN   Alt  Const\n",
      "Conserved      2  1711   4733\n",
      "Unconserved    1  1797   1741\n",
      "Const_Alt    NaN   Alt  Const\n",
      "Conserved      2  1711   4733\n",
      "Unconserved    1  1797   1741\n",
      "Cat :  Unconserved\n",
      "0.35023875391069537 0.043893069916154215 0.32136728494935574 0.3817040205578104\n",
      "Cat :  Conserved\n",
      "2.855195174246714 0.043893069916154215 2.619830932193564 3.1117044168250967\n",
      "most_severe_consequence\n",
      "Const_Alt                NaN     Alt    Const\n",
      "missense_variant       180.0  539462  1296892\n",
      "start_lost               NaN    3187     3105\n",
      "stop_gained              4.0   16397    34789\n",
      "stop_lost                1.0    1351     1802\n",
      "stop_retained_variant    NaN     421      606\n",
      "synonymous_variant      96.0  242579   629551\n",
      "Const_Alt                NaN     Alt    Const\n",
      "missense_variant       180.0  539462  1296892\n",
      "start_lost               0.0    3187     3105\n",
      "stop_gained              4.0   16397    34789\n",
      "stop_lost                1.0    1351     1802\n",
      "stop_retained_variant    0.0     421      606\n",
      "synonymous_variant      96.0  242579   629551\n",
      "Cat :  synonymous_variant\n",
      "1.0884418749234832 0.0028708441896528854 1.0823345693142068 1.0945836423182949\n",
      "Cat :  missense_variant\n",
      "0.9472408776510061 0.002811840928364767 0.9420347950200562 0.9524757312960455\n",
      "Cat :  stop_gained\n",
      "0.8642811671479665 0.009566595120117063 0.8482264280579159 0.8806397810510664\n",
      "Cat :  stop_lost\n",
      "0.5444384114937965 0.036012319262259715 0.5073345662886622 0.5842558414228686\n",
      "Cat :  start_lost\n",
      "0.397028465421028 0.025250640222816152 0.3778573428779077 0.4171722617694635\n",
      "Cat :  stop_retained_variant\n",
      "0.5878667207459939 0.06346034909208075 0.5191111243047623 0.6657289069339214\n",
      "AF_bins\n",
      "Const_Alt       NaN     Alt    Const\n",
      "1e-06 - 1e-05   183  520700  1294505\n",
      "1e-05 - 0.0001   75  229079   547997\n",
      "0.0001 - 0.001   17   37660    87940\n",
      "0.001 - 0.01      4   10054    23381\n",
      "0.01 - 0.1        1    3003     6725\n",
      "0.1 - 1.0         1    2901     6197\n",
      "Const_Alt       NaN     Alt    Const\n",
      "1e-06 - 1e-05   183  520700  1294505\n",
      "1e-05 - 0.0001   75  229079   547997\n",
      "0.0001 - 0.001   17   37660    87940\n",
      "0.001 - 0.01      4   10054    23381\n",
      "0.01 - 0.1        1    3003     6725\n",
      "0.1 - 1.0         1    2901     6197\n",
      "Cat :  1e-06 - 1e-05\n",
      "1.0454734390868754 0.0027781122929808203 1.0397962021539355 1.0511816734586654\n",
      "Cat :  1e-05 - 0.0001\n",
      "0.9683671935999355 0.0029387371867377566 0.9628055041360526 0.9739610104141087\n",
      "Cat :  0.001 - 0.01\n",
      "0.9493602457732052 0.012000322126800723 0.9272912491716346 0.9719544717581385\n",
      "Cat :  0.0001 - 0.001\n",
      "0.9517087687698033 0.006305786746005693 0.9400186643983359 0.9635442516804332\n",
      "Cat :  0.1 - 1.0\n",
      "0.8721987379089559 0.02253522602837476 0.8345129534347524 0.9115863753569097\n",
      "Cat :  0.01 - 0.1\n",
      "0.9144927734789063 0.021987696589508075 0.8759190520481056 0.9547652043755439\n",
      "Status\n",
      "Const_Alt   NaN     Alt    Const\n",
      "Benign      281  802132  1962926\n",
      "Pathogenic    2    6654    17765\n",
      "Const_Alt   NaN     Alt    Const\n",
      "Benign      281  802132  1962926\n",
      "Pathogenic    2    6654    17765\n",
      "Cat :  Benign\n",
      "0.9165911716651426 0.014433714783707441 0.8910240858112835 0.9428918806493612\n",
      "Cat :  Pathogenic\n",
      "1.0909989436002654 0.014433714783707441 1.0605669860167943 1.1223041171659163\n",
      "CCR_bins\n",
      "Const_Alt  NaN     Alt    Const\n",
      "0 - 20     343  825820  1980838\n",
      "20 - 80     73  208281   530896\n",
      "80 - 90      1   14777    42923\n",
      "90 - 95      1    5457    17548\n",
      "95 - 100     2    3629    12799\n",
      "Const_Alt  NaN     Alt    Const\n",
      "0 - 20     343  825820  1980838\n",
      "20 - 80     73  208281   530896\n",
      "80 - 90      1   14777    42923\n",
      "90 - 95      1    5457    17548\n",
      "95 - 100     2    3629    12799\n",
      "Cat :  90 - 95\n",
      "1.3182435507075838 0.015542748298482092 1.2786903888190908 1.3590201929859016\n",
      "Cat :  0 - 20\n",
      "0.9216472770607997 0.002771028656663815 0.9166551803190102 0.9266665607213068\n",
      "Cat :  20 - 80\n",
      "1.0543710046285748 0.0028893862631653488 1.0484167697546907 1.0603590551700026\n",
      "Cat :  95 - 100\n",
      "1.4456465245939014 0.018842142808086198 1.3932317429632792 1.5000332031090595\n",
      "Cat :  80 - 90\n",
      "1.1920008817965122 0.00960844664059193 1.1697625189950471 1.214662018256783\n"
     ]
    }
   ],
   "source": [
    "complete_list_stats = list()\n",
    "list_df = [phylocsf_refseq, gnomad, merge_variants, ccr]\n",
    "list_cols = [['phyloCSF_cat'], ['most_severe_consequence', 'AF_bins'], ['Status'], ['CCR_bins']]\n",
    "for df, cols in zip(list_df, list_cols):\n",
    "    tmp_m = pd.merge(refseq_exons, df, on='MAP')\n",
    "    complete_list_stats.extend(compute_odds_ratio(tmp_m, cols))\n",
    "df_or_stat = pd.DataFrame(complete_list_stats)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "split-model",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>col</th>\n",
       "      <th>cat</th>\n",
       "      <th>OR</th>\n",
       "      <th>SE</th>\n",
       "      <th>Lower_CI</th>\n",
       "      <th>Upper_CI</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>phyloCSF_cat</td>\n",
       "      <td>Unconserved</td>\n",
       "      <td>0.350239</td>\n",
       "      <td>0.043893</td>\n",
       "      <td>0.321367</td>\n",
       "      <td>0.381704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>phyloCSF_cat</td>\n",
       "      <td>Conserved</td>\n",
       "      <td>2.855195</td>\n",
       "      <td>0.043893</td>\n",
       "      <td>2.619831</td>\n",
       "      <td>3.111704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>most_severe_consequence</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>1.088442</td>\n",
       "      <td>0.002871</td>\n",
       "      <td>1.082335</td>\n",
       "      <td>1.094584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>most_severe_consequence</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>0.947241</td>\n",
       "      <td>0.002812</td>\n",
       "      <td>0.942035</td>\n",
       "      <td>0.952476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>most_severe_consequence</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>0.864281</td>\n",
       "      <td>0.009567</td>\n",
       "      <td>0.848226</td>\n",
       "      <td>0.880640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>most_severe_consequence</td>\n",
       "      <td>stop_lost</td>\n",
       "      <td>0.544438</td>\n",
       "      <td>0.036012</td>\n",
       "      <td>0.507335</td>\n",
       "      <td>0.584256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>most_severe_consequence</td>\n",
       "      <td>start_lost</td>\n",
       "      <td>0.397028</td>\n",
       "      <td>0.025251</td>\n",
       "      <td>0.377857</td>\n",
       "      <td>0.417172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>most_severe_consequence</td>\n",
       "      <td>stop_retained_variant</td>\n",
       "      <td>0.587867</td>\n",
       "      <td>0.063460</td>\n",
       "      <td>0.519111</td>\n",
       "      <td>0.665729</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>AF_bins</td>\n",
       "      <td>1e-06 - 1e-05</td>\n",
       "      <td>1.045473</td>\n",
       "      <td>0.002778</td>\n",
       "      <td>1.039796</td>\n",
       "      <td>1.051182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>AF_bins</td>\n",
       "      <td>1e-05 - 0.0001</td>\n",
       "      <td>0.968367</td>\n",
       "      <td>0.002939</td>\n",
       "      <td>0.962806</td>\n",
       "      <td>0.973961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>AF_bins</td>\n",
       "      <td>0.001 - 0.01</td>\n",
       "      <td>0.949360</td>\n",
       "      <td>0.012000</td>\n",
       "      <td>0.927291</td>\n",
       "      <td>0.971954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>AF_bins</td>\n",
       "      <td>0.0001 - 0.001</td>\n",
       "      <td>0.951709</td>\n",
       "      <td>0.006306</td>\n",
       "      <td>0.940019</td>\n",
       "      <td>0.963544</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>AF_bins</td>\n",
       "      <td>0.1 - 1.0</td>\n",
       "      <td>0.872199</td>\n",
       "      <td>0.022535</td>\n",
       "      <td>0.834513</td>\n",
       "      <td>0.911586</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>AF_bins</td>\n",
       "      <td>0.01 - 0.1</td>\n",
       "      <td>0.914493</td>\n",
       "      <td>0.021988</td>\n",
       "      <td>0.875919</td>\n",
       "      <td>0.954765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Status</td>\n",
       "      <td>Benign</td>\n",
       "      <td>0.916591</td>\n",
       "      <td>0.014434</td>\n",
       "      <td>0.891024</td>\n",
       "      <td>0.942892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Status</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1.090999</td>\n",
       "      <td>0.014434</td>\n",
       "      <td>1.060567</td>\n",
       "      <td>1.122304</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>CCR_bins</td>\n",
       "      <td>90 - 95</td>\n",
       "      <td>1.318244</td>\n",
       "      <td>0.015543</td>\n",
       "      <td>1.278690</td>\n",
       "      <td>1.359020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>CCR_bins</td>\n",
       "      <td>0 - 20</td>\n",
       "      <td>0.921647</td>\n",
       "      <td>0.002771</td>\n",
       "      <td>0.916655</td>\n",
       "      <td>0.926667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>CCR_bins</td>\n",
       "      <td>20 - 80</td>\n",
       "      <td>1.054371</td>\n",
       "      <td>0.002889</td>\n",
       "      <td>1.048417</td>\n",
       "      <td>1.060359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>CCR_bins</td>\n",
       "      <td>95 - 100</td>\n",
       "      <td>1.445647</td>\n",
       "      <td>0.018842</td>\n",
       "      <td>1.393232</td>\n",
       "      <td>1.500033</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>CCR_bins</td>\n",
       "      <td>80 - 90</td>\n",
       "      <td>1.192001</td>\n",
       "      <td>0.009608</td>\n",
       "      <td>1.169763</td>\n",
       "      <td>1.214662</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        col                    cat        OR        SE  \\\n",
       "0              phyloCSF_cat            Unconserved  0.350239  0.043893   \n",
       "1              phyloCSF_cat              Conserved  2.855195  0.043893   \n",
       "2   most_severe_consequence     synonymous_variant  1.088442  0.002871   \n",
       "3   most_severe_consequence       missense_variant  0.947241  0.002812   \n",
       "4   most_severe_consequence            stop_gained  0.864281  0.009567   \n",
       "5   most_severe_consequence              stop_lost  0.544438  0.036012   \n",
       "6   most_severe_consequence             start_lost  0.397028  0.025251   \n",
       "7   most_severe_consequence  stop_retained_variant  0.587867  0.063460   \n",
       "8                   AF_bins          1e-06 - 1e-05  1.045473  0.002778   \n",
       "9                   AF_bins         1e-05 - 0.0001  0.968367  0.002939   \n",
       "10                  AF_bins           0.001 - 0.01  0.949360  0.012000   \n",
       "11                  AF_bins         0.0001 - 0.001  0.951709  0.006306   \n",
       "12                  AF_bins              0.1 - 1.0  0.872199  0.022535   \n",
       "13                  AF_bins             0.01 - 0.1  0.914493  0.021988   \n",
       "14                   Status                 Benign  0.916591  0.014434   \n",
       "15                   Status             Pathogenic  1.090999  0.014434   \n",
       "16                 CCR_bins                90 - 95  1.318244  0.015543   \n",
       "17                 CCR_bins                 0 - 20  0.921647  0.002771   \n",
       "18                 CCR_bins                20 - 80  1.054371  0.002889   \n",
       "19                 CCR_bins               95 - 100  1.445647  0.018842   \n",
       "20                 CCR_bins                80 - 90  1.192001  0.009608   \n",
       "\n",
       "    Lower_CI  Upper_CI  \n",
       "0   0.321367  0.381704  \n",
       "1   2.619831  3.111704  \n",
       "2   1.082335  1.094584  \n",
       "3   0.942035  0.952476  \n",
       "4   0.848226  0.880640  \n",
       "5   0.507335  0.584256  \n",
       "6   0.377857  0.417172  \n",
       "7   0.519111  0.665729  \n",
       "8   1.039796  1.051182  \n",
       "9   0.962806  0.973961  \n",
       "10  0.927291  0.971954  \n",
       "11  0.940019  0.963544  \n",
       "12  0.834513  0.911586  \n",
       "13  0.875919  0.954765  \n",
       "14  0.891024  0.942892  \n",
       "15  1.060567  1.122304  \n",
       "16  1.278690  1.359020  \n",
       "17  0.916655  0.926667  \n",
       "18  1.048417  1.060359  \n",
       "19  1.393232  1.500033  \n",
       "20  1.169763  1.214662  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_or_stat"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.7.6 64-bit ('ExoCarto': conda)",
   "language": "python",
   "name": "python37664bitexocartoconda47cfe11889aa4008ba405d9d7b40c893"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
