{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "novel-prisoner",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "valuable-discrimination",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/weber/.conda/envs/ExoCarto/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3166: DtypeWarning: Columns (0) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>861332</td>\n",
       "      <td>1019397</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=1003021;CLNDISDB=MedGen:CN517202;CLND...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>865519</td>\n",
       "      <td>1125147</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=1110865;CLNDISDB=MedGen:CN517202;CLND...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>865545</td>\n",
       "      <td>1164676</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=1153701;CLNDISDB=MedGen:CN517202;CLND...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>865567</td>\n",
       "      <td>1144630</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=1131738;CLNDISDB=MedGen:CN517202;CLND...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>865568</td>\n",
       "      <td>846933</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=824438;CLNDISDB=MedGen:CN517202;CLNDN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>970178</th>\n",
       "      <td>MT</td>\n",
       "      <td>16011</td>\n",
       "      <td>690273</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=677665;CLNDISDB=MONDO:MONDO:0010789,M...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>970179</th>\n",
       "      <td>MT</td>\n",
       "      <td>16013</td>\n",
       "      <td>690274</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=677666;CLNDISDB=MONDO:MONDO:0010789,M...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>970180</th>\n",
       "      <td>MT</td>\n",
       "      <td>16015</td>\n",
       "      <td>690276</td>\n",
       "      <td>T</td>\n",
       "      <td>TATTCTCTGTTCTTTC</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=677667;CLNDISDB=MONDO:MONDO:0010789,M...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>970181</th>\n",
       "      <td>MT</td>\n",
       "      <td>16017</td>\n",
       "      <td>690275</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=677668;CLNDISDB=MONDO:MONDO:0010789,M...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>970182</th>\n",
       "      <td>MT</td>\n",
       "      <td>16021</td>\n",
       "      <td>690277</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=677669;CLNDISDB=MONDO:MONDO:0010789,M...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>970183 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       #CHROM     POS       ID REF               ALT QUAL FILTER  \\\n",
       "0           1  861332  1019397   G                 A    .      .   \n",
       "1           1  865519  1125147   C                 T    .      .   \n",
       "2           1  865545  1164676   G                 A    .      .   \n",
       "3           1  865567  1144630   C                 T    .      .   \n",
       "4           1  865568   846933   G                 A    .      .   \n",
       "...       ...     ...      ...  ..               ...  ...    ...   \n",
       "970178     MT   16011   690273   A                 G    .      .   \n",
       "970179     MT   16013   690274   A                 G    .      .   \n",
       "970180     MT   16015   690276   T  TATTCTCTGTTCTTTC    .      .   \n",
       "970181     MT   16017   690275   T                 C    .      .   \n",
       "970182     MT   16021   690277   C                 T    .      .   \n",
       "\n",
       "                                                     INFO  \n",
       "0       ALLELEID=1003021;CLNDISDB=MedGen:CN517202;CLND...  \n",
       "1       ALLELEID=1110865;CLNDISDB=MedGen:CN517202;CLND...  \n",
       "2       ALLELEID=1153701;CLNDISDB=MedGen:CN517202;CLND...  \n",
       "3       ALLELEID=1131738;CLNDISDB=MedGen:CN517202;CLND...  \n",
       "4       ALLELEID=824438;CLNDISDB=MedGen:CN517202;CLNDN...  \n",
       "...                                                   ...  \n",
       "970178  ALLELEID=677665;CLNDISDB=MONDO:MONDO:0010789,M...  \n",
       "970179  ALLELEID=677666;CLNDISDB=MONDO:MONDO:0010789,M...  \n",
       "970180  ALLELEID=677667;CLNDISDB=MONDO:MONDO:0010789,M...  \n",
       "970181  ALLELEID=677668;CLNDISDB=MONDO:MONDO:0010789,M...  \n",
       "970182  ALLELEID=677669;CLNDISDB=MONDO:MONDO:0010789,M...  \n",
       "\n",
       "[970183 rows x 8 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cv = pd.read_csv('/gstock/biolo_datasets/variation/variation_sets/clinvar/vcf_GRCh37/v2/clinvar_20210731.vcf.gz', compression='gzip', sep='\\t', skiprows=27)\n",
    "cv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "judicial-consolidation",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>MC</th>\n",
       "      <th>CLNVC</th>\n",
       "      <th>CLNSIG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>861332</td>\n",
       "      <td>1019397</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=1003021;CLNDISDB=MedGen:CN517202;CLND...</td>\n",
       "      <td>SO:0001583|missense_variant</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>865519</td>\n",
       "      <td>1125147</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=1110865;CLNDISDB=MedGen:CN517202;CLND...</td>\n",
       "      <td>SO:0001627|intron_variant</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>Likely_benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>865545</td>\n",
       "      <td>1164676</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=1153701;CLNDISDB=MedGen:CN517202;CLND...</td>\n",
       "      <td>SO:0001583|missense_variant</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>Benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>865567</td>\n",
       "      <td>1144630</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=1131738;CLNDISDB=MedGen:CN517202;CLND...</td>\n",
       "      <td>SO:0001819|synonymous_variant</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>Likely_benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>865568</td>\n",
       "      <td>846933</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=824438;CLNDISDB=MedGen:CN517202;CLNDN...</td>\n",
       "      <td>SO:0001583|missense_variant</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>970178</th>\n",
       "      <td>MT</td>\n",
       "      <td>16011</td>\n",
       "      <td>690273</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=677665;CLNDISDB=MONDO:MONDO:0010789,M...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>Likely_benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>970179</th>\n",
       "      <td>MT</td>\n",
       "      <td>16013</td>\n",
       "      <td>690274</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=677666;CLNDISDB=MONDO:MONDO:0010789,M...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>970180</th>\n",
       "      <td>MT</td>\n",
       "      <td>16015</td>\n",
       "      <td>690276</td>\n",
       "      <td>T</td>\n",
       "      <td>TATTCTCTGTTCTTTC</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=677667;CLNDISDB=MONDO:MONDO:0010789,M...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Duplication</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>970181</th>\n",
       "      <td>MT</td>\n",
       "      <td>16017</td>\n",
       "      <td>690275</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=677668;CLNDISDB=MONDO:MONDO:0010789,M...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>Benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>970182</th>\n",
       "      <td>MT</td>\n",
       "      <td>16021</td>\n",
       "      <td>690277</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=677669;CLNDISDB=MONDO:MONDO:0010789,M...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>970183 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       #CHROM     POS       ID REF               ALT QUAL FILTER  \\\n",
       "0           1  861332  1019397   G                 A    .      .   \n",
       "1           1  865519  1125147   C                 T    .      .   \n",
       "2           1  865545  1164676   G                 A    .      .   \n",
       "3           1  865567  1144630   C                 T    .      .   \n",
       "4           1  865568   846933   G                 A    .      .   \n",
       "...       ...     ...      ...  ..               ...  ...    ...   \n",
       "970178     MT   16011   690273   A                 G    .      .   \n",
       "970179     MT   16013   690274   A                 G    .      .   \n",
       "970180     MT   16015   690276   T  TATTCTCTGTTCTTTC    .      .   \n",
       "970181     MT   16017   690275   T                 C    .      .   \n",
       "970182     MT   16021   690277   C                 T    .      .   \n",
       "\n",
       "                                                     INFO  \\\n",
       "0       ALLELEID=1003021;CLNDISDB=MedGen:CN517202;CLND...   \n",
       "1       ALLELEID=1110865;CLNDISDB=MedGen:CN517202;CLND...   \n",
       "2       ALLELEID=1153701;CLNDISDB=MedGen:CN517202;CLND...   \n",
       "3       ALLELEID=1131738;CLNDISDB=MedGen:CN517202;CLND...   \n",
       "4       ALLELEID=824438;CLNDISDB=MedGen:CN517202;CLNDN...   \n",
       "...                                                   ...   \n",
       "970178  ALLELEID=677665;CLNDISDB=MONDO:MONDO:0010789,M...   \n",
       "970179  ALLELEID=677666;CLNDISDB=MONDO:MONDO:0010789,M...   \n",
       "970180  ALLELEID=677667;CLNDISDB=MONDO:MONDO:0010789,M...   \n",
       "970181  ALLELEID=677668;CLNDISDB=MONDO:MONDO:0010789,M...   \n",
       "970182  ALLELEID=677669;CLNDISDB=MONDO:MONDO:0010789,M...   \n",
       "\n",
       "                                   MC                      CLNVC  \\\n",
       "0         SO:0001583|missense_variant  single_nucleotide_variant   \n",
       "1           SO:0001627|intron_variant  single_nucleotide_variant   \n",
       "2         SO:0001583|missense_variant  single_nucleotide_variant   \n",
       "3       SO:0001819|synonymous_variant  single_nucleotide_variant   \n",
       "4         SO:0001583|missense_variant  single_nucleotide_variant   \n",
       "...                               ...                        ...   \n",
       "970178                            NaN  single_nucleotide_variant   \n",
       "970179                            NaN  single_nucleotide_variant   \n",
       "970180                            NaN                Duplication   \n",
       "970181                            NaN  single_nucleotide_variant   \n",
       "970182                            NaN  single_nucleotide_variant   \n",
       "\n",
       "                        CLNSIG  \n",
       "0       Uncertain_significance  \n",
       "1                Likely_benign  \n",
       "2                       Benign  \n",
       "3                Likely_benign  \n",
       "4       Uncertain_significance  \n",
       "...                        ...  \n",
       "970178           Likely_benign  \n",
       "970179  Uncertain_significance  \n",
       "970180  Uncertain_significance  \n",
       "970181                  Benign  \n",
       "970182  Uncertain_significance  \n",
       "\n",
       "[970183 rows x 11 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "cv['MC'] = cv['INFO'].apply(lambda r: [e.split('=')[1] for e in r.split(';') if 'MC' in e])\n",
    "cv['MC'] = cv['MC'].apply(lambda r: r[0] if r else np.nan)\n",
    "cv['CLNVC'] = cv['INFO'].apply(lambda r: [e.split('=')[1] for e in r.split(';') if 'CLNVC=' in e][0])\n",
    "cv['CLNSIG'] = cv['INFO'].apply(lambda r: [e.split('=')[1] for e in r.split(';') if 'CLNSIG' in e][0])\n",
    "cv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "sudden-tomato",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3_prime_UTR_variant                     44227\n",
       "5_prime_UTR_variant                     10711\n",
       "frameshift_variant                      51923\n",
       "genic_downstream_transcript_variant        97\n",
       "genic_upstream_transcript_variant         768\n",
       "inframe_deletion                         7038\n",
       "inframe_indel                             583\n",
       "inframe_insertion                        2759\n",
       "initiatior_codon_variant                 1878\n",
       "intron_variant                         112127\n",
       "missense_variant                       407152\n",
       "no_sequence_alteration                    435\n",
       "non-coding_transcript_variant           41565\n",
       "nonsense                                37245\n",
       "splice_acceptor_variant                  9727\n",
       "splice_donor_variant                    11923\n",
       "stop_lost                                 625\n",
       "synonymous_variant                     206699\n",
       "Name: MC, dtype: int64"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_cv_series = cv.dropna(subset=['MC']).MC.apply(lambda r: [e.split('|')[1] for e in r.split(',') if 'SO' in e]).dropna().apply(lambda r : r[0] if r else np.nan).value_counts().sort_index()\n",
    "raw_cv_series"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "conscious-treaty",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3_prime_UTR_variant                     39641\n",
       "5_prime_UTR_variant                      9777\n",
       "genic_downstream_transcript_variant        61\n",
       "genic_upstream_transcript_variant         661\n",
       "initiatior_codon_variant                 1623\n",
       "intron_variant                          99847\n",
       "missense_variant                       404723\n",
       "no_sequence_alteration                    237\n",
       "non-coding_transcript_variant           39041\n",
       "nonsense                                33423\n",
       "splice_acceptor_variant                  7681\n",
       "splice_donor_variant                     9389\n",
       "stop_lost                                 502\n",
       "synonymous_variant                     206627\n",
       "Name: SNV, dtype: int64"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_cv_snv = cv.loc[cv['CLNVC'] == 'single_nucleotide_variant'].dropna(subset=['MC']).MC.apply(lambda r: [e.split('|')[1] for e in r.split(',') if 'SO' in e]).dropna().apply(lambda r : r[0] if r else np.nan).value_counts().sort_index().rename('SNV')\n",
    "raw_cv_snv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "possible-programmer",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3_prime_UTR_variant                    843\n",
       "5_prime_UTR_variant                    445\n",
       "genic_upstream_transcript_variant       19\n",
       "initiatior_codon_variant               870\n",
       "intron_variant                        7065\n",
       "missense_variant                     63832\n",
       "no_sequence_alteration                   9\n",
       "non-coding_transcript_variant         2135\n",
       "nonsense                             29315\n",
       "splice_acceptor_variant               6542\n",
       "splice_donor_variant                  8006\n",
       "stop_lost                              164\n",
       "synonymous_variant                   14312\n",
       "Name: Pathogenic, dtype: int64"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_cv_snv_path = cv.loc[(cv['CLNVC'] == 'single_nucleotide_variant') & (cv['CLNSIG'].str.contains('athogenic') == True)].dropna(subset=['MC']).MC.apply(lambda r: [e.split('|')[1] for e in r.split(',') if 'SO' in e]).dropna().apply(lambda r : r[0] if r else np.nan).value_counts().sort_index().rename('Pathogenic')\n",
    "raw_cv_snv_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "alpha-separate",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3_prime_UTR_variant                     39\n",
       "5_prime_UTR_variant                    102\n",
       "genic_upstream_transcript_variant       11\n",
       "initiatior_codon_variant               811\n",
       "intron_variant                        1953\n",
       "missense_variant                     40349\n",
       "no_sequence_alteration                   7\n",
       "non-coding_transcript_variant          186\n",
       "nonsense                             28781\n",
       "splice_acceptor_variant               6405\n",
       "splice_donor_variant                  7819\n",
       "stop_lost                              150\n",
       "synonymous_variant                     476\n",
       "Name: Pathogenic_wt_conflicts, dtype: int64"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_cv_snv_path_conflict = cv.loc[(cv['CLNVC'] == 'single_nucleotide_variant') & (cv['CLNSIG'].str.contains('athogenic') == True) & (~cv['CLNSIG'].str.contains('onflicting') == True)].dropna(subset=['MC']).MC.apply(lambda r: [e.split('|')[1] for e in r.split(',') if 'SO' in e]).dropna().apply(lambda r : r[0] if r else np.nan).value_counts().sort_index().rename('Pathogenic_wt_conflicts')\n",
    "raw_cv_snv_path_conflict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "violent-behavior",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3_prime_UTR_variant                     25709\n",
       "5_prime_UTR_variant                      5521\n",
       "genic_downstream_transcript_variant        31\n",
       "genic_upstream_transcript_variant         348\n",
       "initiatior_codon_variant                  619\n",
       "intron_variant                          18257\n",
       "missense_variant                       284790\n",
       "no_sequence_alteration                     14\n",
       "non-coding_transcript_variant            7479\n",
       "nonsense                                 3233\n",
       "splice_acceptor_variant                   858\n",
       "splice_donor_variant                     1091\n",
       "stop_lost                                 272\n",
       "synonymous_variant                      16297\n",
       "Name: Uncertain, dtype: int64"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_cv_snv_uncertain = cv.loc[(cv['CLNVC'] == 'single_nucleotide_variant') & (cv['CLNSIG'].str.contains('Uncertain') == True)].dropna(subset=['MC']).MC.apply(lambda r: [e.split('|')[1] for e in r.split(',') if 'SO' in e]).dropna().apply(lambda r : r[0] if r else np.nan).value_counts().sort_index().rename('Uncertain')\n",
    "raw_cv_snv_uncertain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "partial-parliament",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNV</th>\n",
       "      <th>Pathogenic</th>\n",
       "      <th>Pathogenic_wt_conflicts</th>\n",
       "      <th>Uncertain</th>\n",
       "      <th>Ratio_uncertain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3_prime_UTR_variant</th>\n",
       "      <td>39641.0</td>\n",
       "      <td>843.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>25709.0</td>\n",
       "      <td>64.854570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5_prime_UTR_variant</th>\n",
       "      <td>9777.0</td>\n",
       "      <td>445.0</td>\n",
       "      <td>102.0</td>\n",
       "      <td>5521.0</td>\n",
       "      <td>56.469265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genic_downstream_transcript_variant</th>\n",
       "      <td>61.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>50.819672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genic_upstream_transcript_variant</th>\n",
       "      <td>661.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>348.0</td>\n",
       "      <td>52.647504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>initiatior_codon_variant</th>\n",
       "      <td>1623.0</td>\n",
       "      <td>870.0</td>\n",
       "      <td>811.0</td>\n",
       "      <td>619.0</td>\n",
       "      <td>38.139248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>intron_variant</th>\n",
       "      <td>99847.0</td>\n",
       "      <td>7065.0</td>\n",
       "      <td>1953.0</td>\n",
       "      <td>18257.0</td>\n",
       "      <td>18.284976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>missense_variant</th>\n",
       "      <td>404723.0</td>\n",
       "      <td>63832.0</td>\n",
       "      <td>40349.0</td>\n",
       "      <td>284790.0</td>\n",
       "      <td>70.366646</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>no_sequence_alteration</th>\n",
       "      <td>237.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>5.907173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>non-coding_transcript_variant</th>\n",
       "      <td>39041.0</td>\n",
       "      <td>2135.0</td>\n",
       "      <td>186.0</td>\n",
       "      <td>7479.0</td>\n",
       "      <td>19.156784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nonsense</th>\n",
       "      <td>33423.0</td>\n",
       "      <td>29315.0</td>\n",
       "      <td>28781.0</td>\n",
       "      <td>3233.0</td>\n",
       "      <td>9.672980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>splice_acceptor_variant</th>\n",
       "      <td>7681.0</td>\n",
       "      <td>6542.0</td>\n",
       "      <td>6405.0</td>\n",
       "      <td>858.0</td>\n",
       "      <td>11.170421</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>splice_donor_variant</th>\n",
       "      <td>9389.0</td>\n",
       "      <td>8006.0</td>\n",
       "      <td>7819.0</td>\n",
       "      <td>1091.0</td>\n",
       "      <td>11.619981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>stop_lost</th>\n",
       "      <td>502.0</td>\n",
       "      <td>164.0</td>\n",
       "      <td>150.0</td>\n",
       "      <td>272.0</td>\n",
       "      <td>54.183267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>synonymous_variant</th>\n",
       "      <td>206627.0</td>\n",
       "      <td>14312.0</td>\n",
       "      <td>476.0</td>\n",
       "      <td>16297.0</td>\n",
       "      <td>7.887159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sum</th>\n",
       "      <td>853233.0</td>\n",
       "      <td>133557.0</td>\n",
       "      <td>87089.0</td>\n",
       "      <td>364519.0</td>\n",
       "      <td>471.179644</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          SNV  Pathogenic  \\\n",
       "3_prime_UTR_variant                   39641.0       843.0   \n",
       "5_prime_UTR_variant                    9777.0       445.0   \n",
       "genic_downstream_transcript_variant      61.0         0.0   \n",
       "genic_upstream_transcript_variant       661.0        19.0   \n",
       "initiatior_codon_variant               1623.0       870.0   \n",
       "intron_variant                        99847.0      7065.0   \n",
       "missense_variant                     404723.0     63832.0   \n",
       "no_sequence_alteration                  237.0         9.0   \n",
       "non-coding_transcript_variant         39041.0      2135.0   \n",
       "nonsense                              33423.0     29315.0   \n",
       "splice_acceptor_variant                7681.0      6542.0   \n",
       "splice_donor_variant                   9389.0      8006.0   \n",
       "stop_lost                               502.0       164.0   \n",
       "synonymous_variant                   206627.0     14312.0   \n",
       "Sum                                  853233.0    133557.0   \n",
       "\n",
       "                                     Pathogenic_wt_conflicts  Uncertain  \\\n",
       "3_prime_UTR_variant                                     39.0    25709.0   \n",
       "5_prime_UTR_variant                                    102.0     5521.0   \n",
       "genic_downstream_transcript_variant                      0.0       31.0   \n",
       "genic_upstream_transcript_variant                       11.0      348.0   \n",
       "initiatior_codon_variant                               811.0      619.0   \n",
       "intron_variant                                        1953.0    18257.0   \n",
       "missense_variant                                     40349.0   284790.0   \n",
       "no_sequence_alteration                                   7.0       14.0   \n",
       "non-coding_transcript_variant                          186.0     7479.0   \n",
       "nonsense                                             28781.0     3233.0   \n",
       "splice_acceptor_variant                               6405.0      858.0   \n",
       "splice_donor_variant                                  7819.0     1091.0   \n",
       "stop_lost                                              150.0      272.0   \n",
       "synonymous_variant                                     476.0    16297.0   \n",
       "Sum                                                  87089.0   364519.0   \n",
       "\n",
       "                                     Ratio_uncertain  \n",
       "3_prime_UTR_variant                        64.854570  \n",
       "5_prime_UTR_variant                        56.469265  \n",
       "genic_downstream_transcript_variant        50.819672  \n",
       "genic_upstream_transcript_variant          52.647504  \n",
       "initiatior_codon_variant                   38.139248  \n",
       "intron_variant                             18.284976  \n",
       "missense_variant                           70.366646  \n",
       "no_sequence_alteration                      5.907173  \n",
       "non-coding_transcript_variant              19.156784  \n",
       "nonsense                                    9.672980  \n",
       "splice_acceptor_variant                    11.170421  \n",
       "splice_donor_variant                       11.619981  \n",
       "stop_lost                                  54.183267  \n",
       "synonymous_variant                          7.887159  \n",
       "Sum                                       471.179644  "
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "concat = pd.concat([raw_cv_snv, raw_cv_snv_path, raw_cv_snv_path_conflict, raw_cv_snv_uncertain], axis=1).fillna(0).astype(int)\n",
    "concat['Ratio_uncertain'] = 100 * (concat['Uncertain'] / concat['SNV'])\n",
    "concat.loc[\"Sum\"] = concat.sum(axis=0)\n",
    "\n",
    "concat"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.7.6 64-bit ('ExoCarto': conda)",
   "language": "python",
   "name": "python37664bitexocartoconda47cfe11889aa4008ba405d9d7b40c893"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
