{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "acquired-allocation",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy.stats import fisher_exact\n",
    "sys.path.append(\"/home/weber/PycharmProjects/EXOTIC/src\")\n",
    "from utils.utils import load_config_file, mkdir\n",
    "\n",
    "yaml = load_config_file(config_file=\"/home/weber/PycharmProjects/EXOTIC/clean/src/config_clean_clean.yaml\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "annoying-solomon",
   "metadata": {},
   "source": [
    "### Fct to compute Odds Ratio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "global-concentration",
   "metadata": {},
   "outputs": [],
   "source": [
    "def stats(cat_const, cat_alt, others_const, others_alt, col, cat, ratio=None):\n",
    "    \n",
    "#     print(others_alt)\n",
    "#     odd_ratio = (cat_const/others_const) / (cat_alt/others_alt)\n",
    "    odd_ratio, pvalue = fisher_exact([[cat_const, cat_alt], [others_const, others_alt]])\n",
    "\n",
    "    se = np.sqrt((1/cat_const) + (1/others_const) + (1/cat_alt) + (1/others_alt))\n",
    "    lower = np.exp(np.log(odd_ratio) - 1.96 *se)\n",
    "    upper = np.exp(np.log(odd_ratio) + 1.96 *se)\n",
    "#     print(odd_ratio, se, lower, upper)\n",
    "    print(col, ratio, cat, cat_const, cat_alt, others_const, others_alt, odd_ratio, se, lower, upper)\n",
    "\n",
    "    convert_col = {\n",
    "        'AF_bins' : 'MAF',\n",
    "        'most_severe_consequence' : 'Consequence',\n",
    "        'CCR_bins' : 'Constraint',\n",
    "        'phyloCSF_cat' : 'Conservation',\n",
    "        'Status' : 'Status',\n",
    "    }\n",
    "    \n",
    "    d = {\n",
    "            'group' : convert_col[col],\n",
    "            'Cat' : cat,\n",
    "            'ratio' : ratio,\n",
    "            'OR' : odd_ratio,\n",
    "            'SE' : se,\n",
    "            'IC_low' : lower,\n",
    "            'IC_up' : upper,\n",
    "            'pvalue' : pvalue,\n",
    "            'Const OK' : cat_const,\n",
    "            'Alt OK' : cat_alt,\n",
    "            'Const_NOT OK' : others_const,\n",
    "            'Alt_NOT OK' : others_alt,\n",
    "        }\n",
    "    return d\n",
    "\n",
    "def compute_odds_ratio(df, cols):\n",
    "    l = list()\n",
    "    for col in cols:\n",
    "#         print(col)\n",
    "#         print(df[['Const_Alt', col]].pivot(columns='Const_Alt', values=col).apply(pd.Series.value_counts))\n",
    "        t = df[['Const_Alt', col]].pivot(columns='Const_Alt', values=col).apply(pd.Series.value_counts).fillna(0)\n",
    "#         print(t)\n",
    "        for i, cat in enumerate(df[col].unique().tolist()):\n",
    "#             print(\"Cat : \", cat)\n",
    "            cat_const = t.loc[cat]['Const']\n",
    "            cat_alt = t.loc[cat]['Alt']\n",
    "            others_const = t.loc[~t.index.isin([cat])]['Const'].sum()\n",
    "            others_alt = t.loc[~t.index.isin([cat])]['Alt'].sum()\n",
    "            d_stats = stats(cat_const, cat_alt, others_const, others_alt, col, cat, )\n",
    "            d_stats['id'] = i+1\n",
    "            l.append(d_stats)\n",
    "    return l\n",
    "\n",
    "\n",
    "def compute_odds_ratio_alt(df, cols):\n",
    "    l = list()\n",
    "    for col in cols:\n",
    "        t = df[['Ratio_num_bins_update', col]].dropna(subset=['Ratio_num_bins_update']).pivot(columns='Ratio_num_bins_update', values=col).apply(pd.Series.value_counts).fillna(0)\n",
    "        t = t[t.columns.dropna()]\n",
    "#         print(df['Ratio_num_bins_update'].unique().tolist())\n",
    "        for ratio in sorted(df['Ratio_num_bins_update'].unique().tolist()):\n",
    "            others_ratio = [c for c in df['Ratio_num_bins_update'].unique().tolist() if c != ratio]\n",
    "            for i, cat in enumerate(df[col].unique().tolist()):\n",
    "                freq_ratio_cat_value = t.loc[cat][ratio]\n",
    "                others_freq_cat_sum = t.loc[cat][others_ratio].sum()\n",
    "                freq_ratio_others_cat = t.loc[~t.index.isin([cat])][ratio].sum()\n",
    "                others_freq_others_cat = t.loc[~t.index.isin([cat])][others_ratio].sum(axis=1).values[0]\n",
    "\n",
    "                d_stats = stats(freq_ratio_cat_value, others_freq_cat_sum, freq_ratio_others_cat, others_freq_others_cat, col, cat, ratio)\n",
    "                d_stats['id'] = i+1\n",
    "                l.append(d_stats)\n",
    "    return l\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fatty-covering",
   "metadata": {},
   "source": [
    "## RefSeq Miso exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "threaded-pipeline",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>ranges</th>\n",
       "      <th>mRNA_exons</th>\n",
       "      <th>mRNA_nb</th>\n",
       "      <th>Share</th>\n",
       "      <th>Strand</th>\n",
       "      <th>mRNA_gene</th>\n",
       "      <th>mRNA_nb_total</th>\n",
       "      <th>Ratio</th>\n",
       "      <th>Ratio_num</th>\n",
       "      <th>Const_Alt</th>\n",
       "      <th>Ratio_num_bins</th>\n",
       "      <th>Ratio_num_bins_update</th>\n",
       "      <th>MAP</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Length</th>\n",
       "      <th>CDS_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A2M</td>\n",
       "      <td>9220419-9220435</td>\n",
       "      <td>[NM_001347425, NM_001347423, NM_000014, NM_001...</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>[NM_001347425, NM_001347423, NM_000014, NM_001...</td>\n",
       "      <td>4</td>\n",
       "      <td>4/4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Const</td>\n",
       "      <td>0.8 - 1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>A2M_9220419-9220435</td>\n",
       "      <td>9220419</td>\n",
       "      <td>9220435</td>\n",
       "      <td>16</td>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A2M</td>\n",
       "      <td>9220779-9220820</td>\n",
       "      <td>[NM_001347425, NM_001347423, NM_000014, NM_001...</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>[NM_001347425, NM_001347423, NM_000014, NM_001...</td>\n",
       "      <td>4</td>\n",
       "      <td>4/4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Const</td>\n",
       "      <td>0.8 - 1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>A2M_9220779-9220820</td>\n",
       "      <td>9220779</td>\n",
       "      <td>9220820</td>\n",
       "      <td>41</td>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A2M</td>\n",
       "      <td>9221336-9221438</td>\n",
       "      <td>[NM_001347425, NM_001347423, NM_000014, NM_001...</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>[NM_001347425, NM_001347423, NM_000014, NM_001...</td>\n",
       "      <td>4</td>\n",
       "      <td>4/4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Const</td>\n",
       "      <td>0.8 - 1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>A2M_9221336-9221438</td>\n",
       "      <td>9221336</td>\n",
       "      <td>9221438</td>\n",
       "      <td>102</td>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A2M</td>\n",
       "      <td>9222341-9222409</td>\n",
       "      <td>[NM_001347425, NM_001347423, NM_000014, NM_001...</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>[NM_001347425, NM_001347423, NM_000014, NM_001...</td>\n",
       "      <td>4</td>\n",
       "      <td>4/4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Const</td>\n",
       "      <td>0.8 - 1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>A2M_9222341-9222409</td>\n",
       "      <td>9222341</td>\n",
       "      <td>9222409</td>\n",
       "      <td>68</td>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A2M</td>\n",
       "      <td>9223084-9223174</td>\n",
       "      <td>[NM_001347425, NM_001347423, NM_000014, NM_001...</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>[NM_001347425, NM_001347423, NM_000014, NM_001...</td>\n",
       "      <td>4</td>\n",
       "      <td>4/4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Const</td>\n",
       "      <td>0.8 - 1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>A2M_9223084-9223174</td>\n",
       "      <td>9223084</td>\n",
       "      <td>9223174</td>\n",
       "      <td>90</td>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178476</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>78047461-78047576</td>\n",
       "      <td>[NM_001308237, NM_015534]</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>[NM_001308237, NM_015534]</td>\n",
       "      <td>2</td>\n",
       "      <td>2/2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Const</td>\n",
       "      <td>0.8 - 1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>ZZZ3_78047461-78047576</td>\n",
       "      <td>78047461</td>\n",
       "      <td>78047576</td>\n",
       "      <td>115</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178477</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>78047664-78047811</td>\n",
       "      <td>[NM_001308237, NM_015534]</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>[NM_001308237, NM_015534]</td>\n",
       "      <td>2</td>\n",
       "      <td>2/2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Const</td>\n",
       "      <td>0.8 - 1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>ZZZ3_78047664-78047811</td>\n",
       "      <td>78047664</td>\n",
       "      <td>78047811</td>\n",
       "      <td>147</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178478</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>78050202-78050340</td>\n",
       "      <td>[NM_001308237, NM_015534]</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>[NM_001308237, NM_015534]</td>\n",
       "      <td>2</td>\n",
       "      <td>2/2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Const</td>\n",
       "      <td>0.8 - 1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>ZZZ3_78050202-78050340</td>\n",
       "      <td>78050202</td>\n",
       "      <td>78050340</td>\n",
       "      <td>138</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178479</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>78097535-78099039</td>\n",
       "      <td>[NM_015534]</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>[NM_001308237, NM_015534]</td>\n",
       "      <td>2</td>\n",
       "      <td>1/2</td>\n",
       "      <td>0.5</td>\n",
       "      <td>Alt</td>\n",
       "      <td>0.4 - 0.6</td>\n",
       "      <td>0.4 - 0.6</td>\n",
       "      <td>ZZZ3_78097535-78099039</td>\n",
       "      <td>78097535</td>\n",
       "      <td>78099039</td>\n",
       "      <td>1504</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178480</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>78105134-78105156</td>\n",
       "      <td>[NM_001308237]</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>[NM_001308237, NM_015534]</td>\n",
       "      <td>2</td>\n",
       "      <td>1/2</td>\n",
       "      <td>0.5</td>\n",
       "      <td>Alt</td>\n",
       "      <td>0.4 - 0.6</td>\n",
       "      <td>0.4 - 0.6</td>\n",
       "      <td>ZZZ3_78105134-78105156</td>\n",
       "      <td>78105134</td>\n",
       "      <td>78105156</td>\n",
       "      <td>22</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>178481 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Gene             ranges  \\\n",
       "0        A2M    9220419-9220435   \n",
       "1        A2M    9220779-9220820   \n",
       "2        A2M    9221336-9221438   \n",
       "3        A2M    9222341-9222409   \n",
       "4        A2M    9223084-9223174   \n",
       "...      ...                ...   \n",
       "178476  ZZZ3  78047461-78047576   \n",
       "178477  ZZZ3  78047664-78047811   \n",
       "178478  ZZZ3  78050202-78050340   \n",
       "178479  ZZZ3  78097535-78099039   \n",
       "178480  ZZZ3  78105134-78105156   \n",
       "\n",
       "                                               mRNA_exons  mRNA_nb  Share  \\\n",
       "0       [NM_001347425, NM_001347423, NM_000014, NM_001...        4  False   \n",
       "1       [NM_001347425, NM_001347423, NM_000014, NM_001...        4  False   \n",
       "2       [NM_001347425, NM_001347423, NM_000014, NM_001...        4  False   \n",
       "3       [NM_001347425, NM_001347423, NM_000014, NM_001...        4  False   \n",
       "4       [NM_001347425, NM_001347423, NM_000014, NM_001...        4  False   \n",
       "...                                                   ...      ...    ...   \n",
       "178476                          [NM_001308237, NM_015534]        2  False   \n",
       "178477                          [NM_001308237, NM_015534]        2  False   \n",
       "178478                          [NM_001308237, NM_015534]        2  False   \n",
       "178479                                        [NM_015534]        1  False   \n",
       "178480                                     [NM_001308237]        1  False   \n",
       "\n",
       "        Strand                                          mRNA_gene  \\\n",
       "0            0  [NM_001347425, NM_001347423, NM_000014, NM_001...   \n",
       "1            0  [NM_001347425, NM_001347423, NM_000014, NM_001...   \n",
       "2            0  [NM_001347425, NM_001347423, NM_000014, NM_001...   \n",
       "3            0  [NM_001347425, NM_001347423, NM_000014, NM_001...   \n",
       "4            0  [NM_001347425, NM_001347423, NM_000014, NM_001...   \n",
       "...        ...                                                ...   \n",
       "178476       0                          [NM_001308237, NM_015534]   \n",
       "178477       0                          [NM_001308237, NM_015534]   \n",
       "178478       0                          [NM_001308237, NM_015534]   \n",
       "178479       0                          [NM_001308237, NM_015534]   \n",
       "178480       0                          [NM_001308237, NM_015534]   \n",
       "\n",
       "        mRNA_nb_total Ratio  Ratio_num Const_Alt Ratio_num_bins  \\\n",
       "0                   4   4/4        1.0     Const        0.8 - 1   \n",
       "1                   4   4/4        1.0     Const        0.8 - 1   \n",
       "2                   4   4/4        1.0     Const        0.8 - 1   \n",
       "3                   4   4/4        1.0     Const        0.8 - 1   \n",
       "4                   4   4/4        1.0     Const        0.8 - 1   \n",
       "...               ...   ...        ...       ...            ...   \n",
       "178476              2   2/2        1.0     Const        0.8 - 1   \n",
       "178477              2   2/2        1.0     Const        0.8 - 1   \n",
       "178478              2   2/2        1.0     Const        0.8 - 1   \n",
       "178479              2   1/2        0.5       Alt      0.4 - 0.6   \n",
       "178480              2   1/2        0.5       Alt      0.4 - 0.6   \n",
       "\n",
       "       Ratio_num_bins_update                     MAP     Start       End  \\\n",
       "0                       1.00     A2M_9220419-9220435   9220419   9220435   \n",
       "1                       1.00     A2M_9220779-9220820   9220779   9220820   \n",
       "2                       1.00     A2M_9221336-9221438   9221336   9221438   \n",
       "3                       1.00     A2M_9222341-9222409   9222341   9222409   \n",
       "4                       1.00     A2M_9223084-9223174   9223084   9223174   \n",
       "...                      ...                     ...       ...       ...   \n",
       "178476                  1.00  ZZZ3_78047461-78047576  78047461  78047576   \n",
       "178477                  1.00  ZZZ3_78047664-78047811  78047664  78047811   \n",
       "178478                  1.00  ZZZ3_78050202-78050340  78050202  78050340   \n",
       "178479             0.4 - 0.6  ZZZ3_78097535-78099039  78097535  78099039   \n",
       "178480             0.4 - 0.6  ZZZ3_78105134-78105156  78105134  78105156   \n",
       "\n",
       "        Length  CDS_count  \n",
       "0           16         38  \n",
       "1           41         38  \n",
       "2          102         38  \n",
       "3           68         38  \n",
       "4           90         38  \n",
       "...        ...        ...  \n",
       "178476     115         12  \n",
       "178477     147         12  \n",
       "178478     138         12  \n",
       "178479    1504         12  \n",
       "178480      22         12  \n",
       "\n",
       "[178481 rows x 18 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "refseq_exons = pd.read_parquet(yaml['2_EXPRESSION']['Final']['refseq_corrected_cds_recomputed'])\n",
    "refseq_exons.loc[(refseq_exons['Const_Alt'] == 'Alt') & (refseq_exons['Ratio_num_bins'] == '0.8 - 1'), 'Ratio_num_bins_update'] = '0.8 - 0.99'\n",
    "refseq_exons.loc[(refseq_exons['Const_Alt'] == 'Alt') & (refseq_exons['Ratio_num_bins'] != '0.8 - 1'), 'Ratio_num_bins_update'] = refseq_exons.loc[(refseq_exons['Const_Alt'] == 'Alt') & (refseq_exons['Ratio_num_bins'] != '0.8 - 1')]['Ratio_num_bins']\n",
    "refseq_exons.loc[(refseq_exons['Const_Alt'] == 'Const') & (refseq_exons['Ratio_num_bins'] == '0.8 - 1'), 'Ratio_num_bins_update'] = '1.00'\n",
    "refseq_exons['MAP'] = refseq_exons['Gene'] + '_' + refseq_exons['ranges']\n",
    "refseq_exons"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "competent-print",
   "metadata": {},
   "source": [
    "## PhyloCSF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "careful-bishop",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>interval</th>\n",
       "      <th>phyloCSF_score</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>MAP</th>\n",
       "      <th>Gene</th>\n",
       "      <th>phyloCSF_cat</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>57660</th>\n",
       "      <td>[10:101089145-101090717)</td>\n",
       "      <td>5643.7434</td>\n",
       "      <td>10:101089145</td>\n",
       "      <td>10:101090717</td>\n",
       "      <td>CNNM1_101089145-101090717</td>\n",
       "      <td>CNNM1</td>\n",
       "      <td>Conserved</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57662</th>\n",
       "      <td>[10:101089691-101090717)</td>\n",
       "      <td>3729.0848</td>\n",
       "      <td>10:101089691</td>\n",
       "      <td>10:101090717</td>\n",
       "      <td>CNNM1_101089145-101090717</td>\n",
       "      <td>CNNM1</td>\n",
       "      <td>Conserved</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57663</th>\n",
       "      <td>[10:101090240-101090717)</td>\n",
       "      <td>2077.2201</td>\n",
       "      <td>10:101090240</td>\n",
       "      <td>10:101090717</td>\n",
       "      <td>CNNM1_101089145-101090717</td>\n",
       "      <td>CNNM1</td>\n",
       "      <td>Conserved</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57682</th>\n",
       "      <td>[10:101458291-101458615)</td>\n",
       "      <td>1293.5257</td>\n",
       "      <td>10:101458291</td>\n",
       "      <td>10:101458615</td>\n",
       "      <td>ENTPD7_101458291-101458615</td>\n",
       "      <td>ENTPD7</td>\n",
       "      <td>Conserved</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57694</th>\n",
       "      <td>[10:101639568-101640118)</td>\n",
       "      <td>1125.4920</td>\n",
       "      <td>10:101639568</td>\n",
       "      <td>10:101640118</td>\n",
       "      <td>DNMBP_101639568-101640118</td>\n",
       "      <td>DNMBP</td>\n",
       "      <td>Conserved</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110044</th>\n",
       "      <td>[Y:16734000-16734471)</td>\n",
       "      <td>1757.4552</td>\n",
       "      <td>Y:16734000</td>\n",
       "      <td>Y:16734471</td>\n",
       "      <td>NLGN4Y_16734000-16734471</td>\n",
       "      <td>NLGN4Y</td>\n",
       "      <td>Conserved</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110045</th>\n",
       "      <td>[Y:16734061-16734471)</td>\n",
       "      <td>1652.5963</td>\n",
       "      <td>Y:16734061</td>\n",
       "      <td>Y:16734471</td>\n",
       "      <td>NLGN4Y_16734000-16734471</td>\n",
       "      <td>NLGN4Y</td>\n",
       "      <td>Conserved</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110049</th>\n",
       "      <td>[Y:16941610-16942399)</td>\n",
       "      <td>3028.8020</td>\n",
       "      <td>Y:16941610</td>\n",
       "      <td>Y:16942399</td>\n",
       "      <td>NLGN4Y_16941610-16942399</td>\n",
       "      <td>NLGN4Y</td>\n",
       "      <td>Conserved</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109991</th>\n",
       "      <td>[Y:2829115-2829687)</td>\n",
       "      <td>1279.9484</td>\n",
       "      <td>Y:2829115</td>\n",
       "      <td>Y:2829687</td>\n",
       "      <td>ZFY_2829115-2829687</td>\n",
       "      <td>ZFY</td>\n",
       "      <td>Conserved</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109992</th>\n",
       "      <td>[Y:2829132-2829687)</td>\n",
       "      <td>1290.2635</td>\n",
       "      <td>Y:2829132</td>\n",
       "      <td>Y:2829687</td>\n",
       "      <td>ZFY_2829115-2829687</td>\n",
       "      <td>ZFY</td>\n",
       "      <td>Conserved</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9985 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        interval  phyloCSF_score         start           end  \\\n",
       "57660   [10:101089145-101090717)       5643.7434  10:101089145  10:101090717   \n",
       "57662   [10:101089691-101090717)       3729.0848  10:101089691  10:101090717   \n",
       "57663   [10:101090240-101090717)       2077.2201  10:101090240  10:101090717   \n",
       "57682   [10:101458291-101458615)       1293.5257  10:101458291  10:101458615   \n",
       "57694   [10:101639568-101640118)       1125.4920  10:101639568  10:101640118   \n",
       "...                          ...             ...           ...           ...   \n",
       "110044     [Y:16734000-16734471)       1757.4552    Y:16734000    Y:16734471   \n",
       "110045     [Y:16734061-16734471)       1652.5963    Y:16734061    Y:16734471   \n",
       "110049     [Y:16941610-16942399)       3028.8020    Y:16941610    Y:16942399   \n",
       "109991       [Y:2829115-2829687)       1279.9484     Y:2829115     Y:2829687   \n",
       "109992       [Y:2829132-2829687)       1290.2635     Y:2829132     Y:2829687   \n",
       "\n",
       "                               MAP    Gene phyloCSF_cat  \n",
       "57660    CNNM1_101089145-101090717   CNNM1    Conserved  \n",
       "57662    CNNM1_101089145-101090717   CNNM1    Conserved  \n",
       "57663    CNNM1_101089145-101090717   CNNM1    Conserved  \n",
       "57682   ENTPD7_101458291-101458615  ENTPD7    Conserved  \n",
       "57694    DNMBP_101639568-101640118   DNMBP    Conserved  \n",
       "...                            ...     ...          ...  \n",
       "110044    NLGN4Y_16734000-16734471  NLGN4Y    Conserved  \n",
       "110045    NLGN4Y_16734000-16734471  NLGN4Y    Conserved  \n",
       "110049    NLGN4Y_16941610-16942399  NLGN4Y    Conserved  \n",
       "109991         ZFY_2829115-2829687     ZFY    Conserved  \n",
       "109992         ZFY_2829115-2829687     ZFY    Conserved  \n",
       "\n",
       "[9985 rows x 7 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phylocsf_refseq = pd.read_parquet(yaml['3_EXONS_PROPERTIES']['Final']['refseq_phylocsf'])\n",
    "phylocsf_refseq = phylocsf_refseq.rename({'target' : 'phyloCSF_score'}, axis=1)\n",
    "phylocsf_refseq.loc[(phylocsf_refseq['phyloCSF_score'] <= -100), 'phyloCSF_cat'] = 'Unconserved'\n",
    "phylocsf_refseq.loc[(phylocsf_refseq['phyloCSF_score'] > 1000), 'phyloCSF_cat'] = 'Conserved'\n",
    "phylocsf_refseq.loc[(phylocsf_refseq['phyloCSF_score'] >= -100) & (phylocsf_refseq['phyloCSF_score'] < 1000), 'phyloCSF_cat'] = 'Standard'\n",
    "phylocsf_refseq = phylocsf_refseq.fillna(False)\n",
    "phylocsf_refseq = phylocsf_refseq.loc[phylocsf_refseq['phyloCSF_cat'] != 'Standard']\n",
    "phylocsf_refseq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "sensitive-wedding",
   "metadata": {},
   "source": [
    "## gnomAD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "acquired-kazakhstan",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>locus</th>\n",
       "      <th>alleles</th>\n",
       "      <th>AF</th>\n",
       "      <th>rsid</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>end</th>\n",
       "      <th>id</th>\n",
       "      <th>most_severe_consequence</th>\n",
       "      <th>MAP</th>\n",
       "      <th>Gene</th>\n",
       "      <th>AF_bins</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:901913</td>\n",
       "      <td>[\"T\",\"G\"]</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>rs759148200</td>\n",
       "      <td>snv</td>\n",
       "      <td>901913</td>\n",
       "      <td>1_901913_T_G</td>\n",
       "      <td>start_lost</td>\n",
       "      <td>PLEKHN1_901912-901994</td>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>1e-05 - 1e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:901916</td>\n",
       "      <td>[\"G\",\"A\"]</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>rs764917347</td>\n",
       "      <td>snv</td>\n",
       "      <td>901916</td>\n",
       "      <td>1_901916_G_A</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>PLEKHN1_901912-901994</td>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>1e-06 - 1e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:901917</td>\n",
       "      <td>[\"G\",\"A\"]</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>rs774961129</td>\n",
       "      <td>snv</td>\n",
       "      <td>901917</td>\n",
       "      <td>1_901917_G_A</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>PLEKHN1_901912-901994</td>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>1e-06 - 1e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:901919</td>\n",
       "      <td>[\"A\",\"C\"]</td>\n",
       "      <td>0.000226</td>\n",
       "      <td>rs763342668</td>\n",
       "      <td>snv</td>\n",
       "      <td>901919</td>\n",
       "      <td>1_901919_A_C</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>PLEKHN1_901912-901994</td>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>1e-04 - 1e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:901920</td>\n",
       "      <td>[\"C\",\"G\"]</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>rs1408404600</td>\n",
       "      <td>snv</td>\n",
       "      <td>901920</td>\n",
       "      <td>1_901920_C_G</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>PLEKHN1_901912-901994</td>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>1e-06 - 1e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2810848</th>\n",
       "      <td>Y:22749975</td>\n",
       "      <td>[\"C\",\"T\"]</td>\n",
       "      <td>0.000017</td>\n",
       "      <td>rs746588234</td>\n",
       "      <td>snv</td>\n",
       "      <td>22749975</td>\n",
       "      <td>Y_22749975_C_T</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>EIF1AY_22749910-22749991</td>\n",
       "      <td>EIF1AY</td>\n",
       "      <td>1e-05 - 1e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2810849</th>\n",
       "      <td>Y:22751395</td>\n",
       "      <td>[\"T\",\"C\"]</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>rs1382594808</td>\n",
       "      <td>snv</td>\n",
       "      <td>22751395</td>\n",
       "      <td>Y_22751395_T_C</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>EIF1AY_22751370-22751461</td>\n",
       "      <td>EIF1AY</td>\n",
       "      <td>1e-05 - 1e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2810850</th>\n",
       "      <td>Y:22751399</td>\n",
       "      <td>[\"C\",\"G\"]</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>rs762602059</td>\n",
       "      <td>snv</td>\n",
       "      <td>22751399</td>\n",
       "      <td>Y_22751399_C_G</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>EIF1AY_22751370-22751461</td>\n",
       "      <td>EIF1AY</td>\n",
       "      <td>1e-05 - 1e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2810851</th>\n",
       "      <td>Y:22751428</td>\n",
       "      <td>[\"C\",\"T\"]</td>\n",
       "      <td>0.000030</td>\n",
       "      <td>rs1158365540</td>\n",
       "      <td>snv</td>\n",
       "      <td>22751428</td>\n",
       "      <td>Y_22751428_C_T</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>EIF1AY_22751370-22751461</td>\n",
       "      <td>EIF1AY</td>\n",
       "      <td>1e-05 - 1e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2810852</th>\n",
       "      <td>Y:22751443</td>\n",
       "      <td>[\"T\",\"C\"]</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>rs772483555</td>\n",
       "      <td>snv</td>\n",
       "      <td>22751443</td>\n",
       "      <td>Y_22751443_T_C</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>EIF1AY_22751370-22751461</td>\n",
       "      <td>EIF1AY</td>\n",
       "      <td>1e-05 - 1e-04</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2770423 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              locus    alleles        AF          rsid variant_type       end  \\\n",
       "0          1:901913  [\"T\",\"G\"]  0.000012   rs759148200          snv    901913   \n",
       "1          1:901916  [\"G\",\"A\"]  0.000004   rs764917347          snv    901916   \n",
       "2          1:901917  [\"G\",\"A\"]  0.000004   rs774961129          snv    901917   \n",
       "3          1:901919  [\"A\",\"C\"]  0.000226   rs763342668          snv    901919   \n",
       "4          1:901920  [\"C\",\"G\"]  0.000004  rs1408404600          snv    901920   \n",
       "...             ...        ...       ...           ...          ...       ...   \n",
       "2810848  Y:22749975  [\"C\",\"T\"]  0.000017   rs746588234          snv  22749975   \n",
       "2810849  Y:22751395  [\"T\",\"C\"]  0.000015  rs1382594808          snv  22751395   \n",
       "2810850  Y:22751399  [\"C\",\"G\"]  0.000015   rs762602059          snv  22751399   \n",
       "2810851  Y:22751428  [\"C\",\"T\"]  0.000030  rs1158365540          snv  22751428   \n",
       "2810852  Y:22751443  [\"T\",\"C\"]  0.000015   rs772483555          snv  22751443   \n",
       "\n",
       "                     id most_severe_consequence                       MAP  \\\n",
       "0          1_901913_T_G              start_lost     PLEKHN1_901912-901994   \n",
       "1          1_901916_G_A        missense_variant     PLEKHN1_901912-901994   \n",
       "2          1_901917_G_A      synonymous_variant     PLEKHN1_901912-901994   \n",
       "3          1_901919_A_C        missense_variant     PLEKHN1_901912-901994   \n",
       "4          1_901920_C_G        missense_variant     PLEKHN1_901912-901994   \n",
       "...                 ...                     ...                       ...   \n",
       "2810848  Y_22749975_C_T      synonymous_variant  EIF1AY_22749910-22749991   \n",
       "2810849  Y_22751395_T_C      synonymous_variant  EIF1AY_22751370-22751461   \n",
       "2810850  Y_22751399_C_G        missense_variant  EIF1AY_22751370-22751461   \n",
       "2810851  Y_22751428_C_T      synonymous_variant  EIF1AY_22751370-22751461   \n",
       "2810852  Y_22751443_T_C      synonymous_variant  EIF1AY_22751370-22751461   \n",
       "\n",
       "            Gene        AF_bins  \n",
       "0        PLEKHN1  1e-05 - 1e-04  \n",
       "1        PLEKHN1  1e-06 - 1e-05  \n",
       "2        PLEKHN1  1e-06 - 1e-05  \n",
       "3        PLEKHN1  1e-04 - 1e-03  \n",
       "4        PLEKHN1  1e-06 - 1e-05  \n",
       "...          ...            ...  \n",
       "2810848   EIF1AY  1e-05 - 1e-04  \n",
       "2810849   EIF1AY  1e-05 - 1e-04  \n",
       "2810850   EIF1AY  1e-05 - 1e-04  \n",
       "2810851   EIF1AY  1e-05 - 1e-04  \n",
       "2810852   EIF1AY  1e-05 - 1e-04  \n",
       "\n",
       "[2770423 rows x 11 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gnomad = pd.read_parquet(yaml['3_EXONS_PROPERTIES']['Final']['refseq_gnomad_hail_retrieved'])\n",
    "\"{:e}\".format(0.001)\n",
    "bins_af = list(reversed([1/(10 ** e) for e in list(range(0,7,1))]))\n",
    "\n",
    "def convert_bins_into_labels(bins):\n",
    "#     return [\" - \".join([str(bins[j]), str(bins[j+1])]) for j, e in enumerate(bins) if j <len(bins)-1]\n",
    "    return [\" - \".join([\"{:.0e}\".format(bins[j]), \"{:.0e}\".format(bins[j+1])]) for j, e in enumerate(bins) if j <len(bins)-1]\n",
    "\n",
    "\n",
    "\n",
    "gnomad['AF'] = gnomad['AF'].str.replace(',', '.').astype(float)\n",
    "gnomad['AF_bins'] = pd.cut(gnomad['AF'], bins=bins_af, labels=convert_bins_into_labels(bins_af))\n",
    "mc = [\"missense_variant\", \"start_lost\", \"stop_gained\", \"stop_lost\", \"stop_retained_variant\", \"synonymous_variant\", ]\n",
    "gnomad = gnomad.loc[gnomad['most_severe_consequence'].isin(mc)]\n",
    "gnomad"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "understood-triangle",
   "metadata": {},
   "source": [
    "## ClinVar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "retained-therapist",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>locus.contig</th>\n",
       "      <th>locus.position</th>\n",
       "      <th>alleles</th>\n",
       "      <th>rsid</th>\n",
       "      <th>ALLELEID</th>\n",
       "      <th>CLNREVSTAT</th>\n",
       "      <th>CLNSIG</th>\n",
       "      <th>GENEINFO</th>\n",
       "      <th>RS</th>\n",
       "      <th>MAP</th>\n",
       "      <th>Gene</th>\n",
       "      <th>RS_STARS</th>\n",
       "      <th>snpId</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>1</td>\n",
       "      <td>1289293</td>\n",
       "      <td>[A, T]</td>\n",
       "      <td>402162</td>\n",
       "      <td>389106</td>\n",
       "      <td>criteria_provided,_single_submitter</td>\n",
       "      <td>Likely_pathogenic</td>\n",
       "      <td>MXRA8:54587</td>\n",
       "      <td>[374879755]</td>\n",
       "      <td>MXRA8_1289228-1289308</td>\n",
       "      <td>MXRA8</td>\n",
       "      <td>1</td>\n",
       "      <td>1_1289293_A_T</td>\n",
       "      <td>1_1289293_A_T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>1</td>\n",
       "      <td>1372327</td>\n",
       "      <td>[C, T]</td>\n",
       "      <td>830326</td>\n",
       "      <td>818588</td>\n",
       "      <td>no_assertion_criteria_provided</td>\n",
       "      <td>Likely_pathogenic</td>\n",
       "      <td>VWA1:64856</td>\n",
       "      <td>None</td>\n",
       "      <td>VWA1_1372307-1372864</td>\n",
       "      <td>VWA1</td>\n",
       "      <td>0</td>\n",
       "      <td>1_1372327_C_T</td>\n",
       "      <td>1_1372327_C_T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>1</td>\n",
       "      <td>1451416</td>\n",
       "      <td>[T, G]</td>\n",
       "      <td>828051</td>\n",
       "      <td>816386</td>\n",
       "      <td>no_assertion_criteria_provided</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>ATAD3A:55210</td>\n",
       "      <td>[1570319915]</td>\n",
       "      <td>ATAD3A_1451392-1451468</td>\n",
       "      <td>ATAD3A</td>\n",
       "      <td>0</td>\n",
       "      <td>1_1451416_T_G</td>\n",
       "      <td>1_1451416_T_G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>1</td>\n",
       "      <td>1454346</td>\n",
       "      <td>[C, T]</td>\n",
       "      <td>828050</td>\n",
       "      <td>816385</td>\n",
       "      <td>no_assertion_criteria_provided</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>ATAD3A:55210</td>\n",
       "      <td>[760826883]</td>\n",
       "      <td>ATAD3A_1454301-1454370</td>\n",
       "      <td>ATAD3A</td>\n",
       "      <td>0</td>\n",
       "      <td>1_1454346_C_T</td>\n",
       "      <td>1_1454346_C_T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>1</td>\n",
       "      <td>1460622</td>\n",
       "      <td>[T, G]</td>\n",
       "      <td>637018</td>\n",
       "      <td>624841</td>\n",
       "      <td>criteria_provided,_single_submitter</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>ATAD3A:55210</td>\n",
       "      <td>[1570345942]</td>\n",
       "      <td>ATAD3A_1460620-1460671</td>\n",
       "      <td>ATAD3A</td>\n",
       "      <td>1</td>\n",
       "      <td>1_1460622_T_G</td>\n",
       "      <td>1_1460622_T_G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256896</th>\n",
       "      <td>X</td>\n",
       "      <td>154250707</td>\n",
       "      <td>[C, A]</td>\n",
       "      <td>10156</td>\n",
       "      <td>25195</td>\n",
       "      <td>no_assertion_criteria_provided</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>F8:2157</td>\n",
       "      <td>[137852379]</td>\n",
       "      <td>F8_154250685-154250827</td>\n",
       "      <td>F8</td>\n",
       "      <td>0</td>\n",
       "      <td>X_154250707_C_A</td>\n",
       "      <td>X_154250707_C_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256897</th>\n",
       "      <td>X</td>\n",
       "      <td>154250724</td>\n",
       "      <td>[T, C]</td>\n",
       "      <td>10352</td>\n",
       "      <td>25391</td>\n",
       "      <td>no_assertion_criteria_provided</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>F8:2157</td>\n",
       "      <td>[137852476]</td>\n",
       "      <td>F8_154250685-154250827</td>\n",
       "      <td>F8</td>\n",
       "      <td>0</td>\n",
       "      <td>X_154250724_T_C</td>\n",
       "      <td>X_154250724_T_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256899</th>\n",
       "      <td>X</td>\n",
       "      <td>154250739</td>\n",
       "      <td>[T, A]</td>\n",
       "      <td>10154</td>\n",
       "      <td>25193</td>\n",
       "      <td>no_assertion_criteria_provided</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>F8:2157</td>\n",
       "      <td>[137852378]</td>\n",
       "      <td>F8_154250685-154250827</td>\n",
       "      <td>F8</td>\n",
       "      <td>0</td>\n",
       "      <td>X_154250739_T_A</td>\n",
       "      <td>X_154250739_T_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256900</th>\n",
       "      <td>X</td>\n",
       "      <td>154250751</td>\n",
       "      <td>[A, C]</td>\n",
       "      <td>10153</td>\n",
       "      <td>25192</td>\n",
       "      <td>no_assertion_criteria_provided</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>F8:2157</td>\n",
       "      <td>[137852377]</td>\n",
       "      <td>F8_154250685-154250827</td>\n",
       "      <td>F8</td>\n",
       "      <td>0</td>\n",
       "      <td>X_154250751_A_C</td>\n",
       "      <td>X_154250751_A_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256902</th>\n",
       "      <td>X</td>\n",
       "      <td>154250785</td>\n",
       "      <td>[G, A]</td>\n",
       "      <td>10152</td>\n",
       "      <td>25191</td>\n",
       "      <td>criteria_provided,_single_submitter</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>F8:2157</td>\n",
       "      <td>[387906432]</td>\n",
       "      <td>F8_154250685-154250827</td>\n",
       "      <td>F8</td>\n",
       "      <td>1</td>\n",
       "      <td>X_154250785_G_A</td>\n",
       "      <td>X_154250785_G_A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>24402 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       locus.contig  locus.position alleles    rsid  ALLELEID  \\\n",
       "71                1         1289293  [A, T]  402162    389106   \n",
       "72                1         1372327  [C, T]  830326    818588   \n",
       "82                1         1451416  [T, G]  828051    816386   \n",
       "89                1         1454346  [C, T]  828050    816385   \n",
       "101               1         1460622  [T, G]  637018    624841   \n",
       "...             ...             ...     ...     ...       ...   \n",
       "256896            X       154250707  [C, A]   10156     25195   \n",
       "256897            X       154250724  [T, C]   10352     25391   \n",
       "256899            X       154250739  [T, A]   10154     25193   \n",
       "256900            X       154250751  [A, C]   10153     25192   \n",
       "256902            X       154250785  [G, A]   10152     25191   \n",
       "\n",
       "                                 CLNREVSTAT             CLNSIG      GENEINFO  \\\n",
       "71      criteria_provided,_single_submitter  Likely_pathogenic   MXRA8:54587   \n",
       "72           no_assertion_criteria_provided  Likely_pathogenic    VWA1:64856   \n",
       "82           no_assertion_criteria_provided         Pathogenic  ATAD3A:55210   \n",
       "89           no_assertion_criteria_provided         Pathogenic  ATAD3A:55210   \n",
       "101     criteria_provided,_single_submitter         Pathogenic  ATAD3A:55210   \n",
       "...                                     ...                ...           ...   \n",
       "256896       no_assertion_criteria_provided         Pathogenic       F8:2157   \n",
       "256897       no_assertion_criteria_provided         Pathogenic       F8:2157   \n",
       "256899       no_assertion_criteria_provided         Pathogenic       F8:2157   \n",
       "256900       no_assertion_criteria_provided         Pathogenic       F8:2157   \n",
       "256902  criteria_provided,_single_submitter         Pathogenic       F8:2157   \n",
       "\n",
       "                  RS                     MAP    Gene  RS_STARS  \\\n",
       "71       [374879755]   MXRA8_1289228-1289308   MXRA8         1   \n",
       "72              None    VWA1_1372307-1372864    VWA1         0   \n",
       "82      [1570319915]  ATAD3A_1451392-1451468  ATAD3A         0   \n",
       "89       [760826883]  ATAD3A_1454301-1454370  ATAD3A         0   \n",
       "101     [1570345942]  ATAD3A_1460620-1460671  ATAD3A         1   \n",
       "...              ...                     ...     ...       ...   \n",
       "256896   [137852379]  F8_154250685-154250827      F8         0   \n",
       "256897   [137852476]  F8_154250685-154250827      F8         0   \n",
       "256899   [137852378]  F8_154250685-154250827      F8         0   \n",
       "256900   [137852377]  F8_154250685-154250827      F8         0   \n",
       "256902   [387906432]  F8_154250685-154250827      F8         1   \n",
       "\n",
       "                  snpId               id  \n",
       "71        1_1289293_A_T    1_1289293_A_T  \n",
       "72        1_1372327_C_T    1_1372327_C_T  \n",
       "82        1_1451416_T_G    1_1451416_T_G  \n",
       "89        1_1454346_C_T    1_1454346_C_T  \n",
       "101       1_1460622_T_G    1_1460622_T_G  \n",
       "...                 ...              ...  \n",
       "256896  X_154250707_C_A  X_154250707_C_A  \n",
       "256897  X_154250724_T_C  X_154250724_T_C  \n",
       "256899  X_154250739_T_A  X_154250739_T_A  \n",
       "256900  X_154250751_A_C  X_154250751_A_C  \n",
       "256902  X_154250785_G_A  X_154250785_G_A  \n",
       "\n",
       "[24402 rows x 14 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar = pd.read_parquet(yaml['3_EXONS_PROPERTIES']['Final']['refseq_clinvar_hail_retrieved'])\n",
    "clinvar['id'] = clinvar['locus.contig'] + '_' + clinvar['locus.position'].astype(str) + '_' + clinvar['alleles'].apply(lambda r: \"_\".join(r))\n",
    "clinvar"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "detailed-logic",
   "metadata": {},
   "source": [
    "## CCR data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "friendly-avenue",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>interval</th>\n",
       "      <th>target</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>MAP</th>\n",
       "      <th>Gene</th>\n",
       "      <th>CCR_bins</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1953387</th>\n",
       "      <td>[10:100003848-100003856)</td>\n",
       "      <td>63.055177</td>\n",
       "      <td>10:100003848</td>\n",
       "      <td>10:100003856</td>\n",
       "      <td>R3HCC1L_100003848-100003915</td>\n",
       "      <td>R3HCC1L</td>\n",
       "      <td>20 - 80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1953388</th>\n",
       "      <td>[10:100003856-100003857)</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>10:100003856</td>\n",
       "      <td>10:100003857</td>\n",
       "      <td>R3HCC1L_100003848-100003915</td>\n",
       "      <td>R3HCC1L</td>\n",
       "      <td>0 - 20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1907621</th>\n",
       "      <td>[10:100003857-100003858)</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>10:100003857</td>\n",
       "      <td>10:100003858</td>\n",
       "      <td>R3HCC1L_100003848-100003915</td>\n",
       "      <td>R3HCC1L</td>\n",
       "      <td>0 - 20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1953390</th>\n",
       "      <td>[10:100003858-100003874)</td>\n",
       "      <td>71.921144</td>\n",
       "      <td>10:100003858</td>\n",
       "      <td>10:100003874</td>\n",
       "      <td>R3HCC1L_100003848-100003915</td>\n",
       "      <td>R3HCC1L</td>\n",
       "      <td>20 - 80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1907623</th>\n",
       "      <td>[10:100003874-100003875)</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>10:100003874</td>\n",
       "      <td>10:100003875</td>\n",
       "      <td>R3HCC1L_100003848-100003915</td>\n",
       "      <td>R3HCC1L</td>\n",
       "      <td>0 - 20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3526305</th>\n",
       "      <td>[X:99956953-99956958)</td>\n",
       "      <td>8.932019</td>\n",
       "      <td>X:99956953</td>\n",
       "      <td>X:99956958</td>\n",
       "      <td>SYTL4_99956924-99957033</td>\n",
       "      <td>SYTL4</td>\n",
       "      <td>0 - 20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3526306</th>\n",
       "      <td>[X:99956958-99956959)</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>X:99956958</td>\n",
       "      <td>X:99956959</td>\n",
       "      <td>SYTL4_99956924-99957033</td>\n",
       "      <td>SYTL4</td>\n",
       "      <td>0 - 20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3526307</th>\n",
       "      <td>[X:99956959-99957026)</td>\n",
       "      <td>90.007142</td>\n",
       "      <td>X:99956959</td>\n",
       "      <td>X:99957026</td>\n",
       "      <td>SYTL4_99956924-99957033</td>\n",
       "      <td>SYTL4</td>\n",
       "      <td>90 - 95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3611490</th>\n",
       "      <td>[X:99957026-99957027)</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>X:99957026</td>\n",
       "      <td>X:99957027</td>\n",
       "      <td>SYTL4_99956924-99957033</td>\n",
       "      <td>SYTL4</td>\n",
       "      <td>0 - 20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3611491</th>\n",
       "      <td>[X:99957027-99957034)</td>\n",
       "      <td>0.514571</td>\n",
       "      <td>X:99957027</td>\n",
       "      <td>X:99957034</td>\n",
       "      <td>SYTL4_99956924-99957033</td>\n",
       "      <td>SYTL4</td>\n",
       "      <td>0 - 20</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3643388 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         interval     target         start           end  \\\n",
       "1953387  [10:100003848-100003856)  63.055177  10:100003848  10:100003856   \n",
       "1953388  [10:100003856-100003857)   0.000000  10:100003856  10:100003857   \n",
       "1907621  [10:100003857-100003858)   0.000000  10:100003857  10:100003858   \n",
       "1953390  [10:100003858-100003874)  71.921144  10:100003858  10:100003874   \n",
       "1907623  [10:100003874-100003875)   0.000000  10:100003874  10:100003875   \n",
       "...                           ...        ...           ...           ...   \n",
       "3526305     [X:99956953-99956958)   8.932019    X:99956953    X:99956958   \n",
       "3526306     [X:99956958-99956959)   0.000000    X:99956958    X:99956959   \n",
       "3526307     [X:99956959-99957026)  90.007142    X:99956959    X:99957026   \n",
       "3611490     [X:99957026-99957027)   0.000000    X:99957026    X:99957027   \n",
       "3611491     [X:99957027-99957034)   0.514571    X:99957027    X:99957034   \n",
       "\n",
       "                                 MAP     Gene CCR_bins  \n",
       "1953387  R3HCC1L_100003848-100003915  R3HCC1L  20 - 80  \n",
       "1953388  R3HCC1L_100003848-100003915  R3HCC1L   0 - 20  \n",
       "1907621  R3HCC1L_100003848-100003915  R3HCC1L   0 - 20  \n",
       "1953390  R3HCC1L_100003848-100003915  R3HCC1L  20 - 80  \n",
       "1907623  R3HCC1L_100003848-100003915  R3HCC1L   0 - 20  \n",
       "...                              ...      ...      ...  \n",
       "3526305      SYTL4_99956924-99957033    SYTL4   0 - 20  \n",
       "3526306      SYTL4_99956924-99957033    SYTL4   0 - 20  \n",
       "3526307      SYTL4_99956924-99957033    SYTL4  90 - 95  \n",
       "3611490      SYTL4_99956924-99957033    SYTL4   0 - 20  \n",
       "3611491      SYTL4_99956924-99957033    SYTL4   0 - 20  \n",
       "\n",
       "[3643388 rows x 7 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ccr = pd.read_parquet(yaml['3_EXONS_PROPERTIES']['Final']['refseq_ccr_start_end'])\n",
    "bins = [0,20,80,90,95,100]\n",
    "ccr['CCR_bins'] = pd.cut(ccr['target'], bins, labels=[\" - \".join([str(bins[j]), str(bins[j+1])]) for j, e in enumerate(bins) if j <len(bins)-1], include_lowest=True)\n",
    "ccr"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "departmental-merit",
   "metadata": {},
   "source": [
    "## Merge gnomAD & ClinVar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "official-ferry",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>locus</th>\n",
       "      <th>alleles</th>\n",
       "      <th>AF</th>\n",
       "      <th>rsid</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>end</th>\n",
       "      <th>id</th>\n",
       "      <th>most_severe_consequence</th>\n",
       "      <th>MAP</th>\n",
       "      <th>Gene</th>\n",
       "      <th>...</th>\n",
       "      <th>locus.contig</th>\n",
       "      <th>locus.position</th>\n",
       "      <th>ALLELEID</th>\n",
       "      <th>CLNREVSTAT</th>\n",
       "      <th>CLNSIG</th>\n",
       "      <th>GENEINFO</th>\n",
       "      <th>RS</th>\n",
       "      <th>RS_STARS</th>\n",
       "      <th>snpId</th>\n",
       "      <th>Status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:901913</td>\n",
       "      <td>[\"T\",\"G\"]</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>rs759148200</td>\n",
       "      <td>snv</td>\n",
       "      <td>901913.0</td>\n",
       "      <td>1_901913_T_G</td>\n",
       "      <td>start_lost</td>\n",
       "      <td>PLEKHN1_901912-901994</td>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:901916</td>\n",
       "      <td>[\"G\",\"A\"]</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>rs764917347</td>\n",
       "      <td>snv</td>\n",
       "      <td>901916.0</td>\n",
       "      <td>1_901916_G_A</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>PLEKHN1_901912-901994</td>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:901917</td>\n",
       "      <td>[\"G\",\"A\"]</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>rs774961129</td>\n",
       "      <td>snv</td>\n",
       "      <td>901917.0</td>\n",
       "      <td>1_901917_G_A</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>PLEKHN1_901912-901994</td>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:901919</td>\n",
       "      <td>[\"A\",\"C\"]</td>\n",
       "      <td>0.000226</td>\n",
       "      <td>rs763342668</td>\n",
       "      <td>snv</td>\n",
       "      <td>901919.0</td>\n",
       "      <td>1_901919_A_C</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>PLEKHN1_901912-901994</td>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:901920</td>\n",
       "      <td>[\"C\",\"G\"]</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>rs1408404600</td>\n",
       "      <td>snv</td>\n",
       "      <td>901920.0</td>\n",
       "      <td>1_901920_C_G</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>PLEKHN1_901912-901994</td>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256896</th>\n",
       "      <td>NaN</td>\n",
       "      <td>[C, A]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10156</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>X_154250707_C_A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F8_154250685-154250827</td>\n",
       "      <td>F8</td>\n",
       "      <td>...</td>\n",
       "      <td>X</td>\n",
       "      <td>154250707.0</td>\n",
       "      <td>25195.0</td>\n",
       "      <td>no_assertion_criteria_provided</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>F8:2157</td>\n",
       "      <td>[137852379]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>X_154250707_C_A</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256897</th>\n",
       "      <td>NaN</td>\n",
       "      <td>[T, C]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10352</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>X_154250724_T_C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F8_154250685-154250827</td>\n",
       "      <td>F8</td>\n",
       "      <td>...</td>\n",
       "      <td>X</td>\n",
       "      <td>154250724.0</td>\n",
       "      <td>25391.0</td>\n",
       "      <td>no_assertion_criteria_provided</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>F8:2157</td>\n",
       "      <td>[137852476]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>X_154250724_T_C</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256899</th>\n",
       "      <td>NaN</td>\n",
       "      <td>[T, A]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10154</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>X_154250739_T_A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F8_154250685-154250827</td>\n",
       "      <td>F8</td>\n",
       "      <td>...</td>\n",
       "      <td>X</td>\n",
       "      <td>154250739.0</td>\n",
       "      <td>25193.0</td>\n",
       "      <td>no_assertion_criteria_provided</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>F8:2157</td>\n",
       "      <td>[137852378]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>X_154250739_T_A</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256900</th>\n",
       "      <td>NaN</td>\n",
       "      <td>[A, C]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10153</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>X_154250751_A_C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F8_154250685-154250827</td>\n",
       "      <td>F8</td>\n",
       "      <td>...</td>\n",
       "      <td>X</td>\n",
       "      <td>154250751.0</td>\n",
       "      <td>25192.0</td>\n",
       "      <td>no_assertion_criteria_provided</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>F8:2157</td>\n",
       "      <td>[137852377]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>X_154250751_A_C</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256902</th>\n",
       "      <td>NaN</td>\n",
       "      <td>[G, A]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10152</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>X_154250785_G_A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F8_154250685-154250827</td>\n",
       "      <td>F8</td>\n",
       "      <td>...</td>\n",
       "      <td>X</td>\n",
       "      <td>154250785.0</td>\n",
       "      <td>25191.0</td>\n",
       "      <td>criteria_provided,_single_submitter</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>F8:2157</td>\n",
       "      <td>[387906432]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>X_154250785_G_A</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2794825 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           locus    alleles        AF          rsid variant_type       end  \\\n",
       "0       1:901913  [\"T\",\"G\"]  0.000012   rs759148200          snv  901913.0   \n",
       "1       1:901916  [\"G\",\"A\"]  0.000004   rs764917347          snv  901916.0   \n",
       "2       1:901917  [\"G\",\"A\"]  0.000004   rs774961129          snv  901917.0   \n",
       "3       1:901919  [\"A\",\"C\"]  0.000226   rs763342668          snv  901919.0   \n",
       "4       1:901920  [\"C\",\"G\"]  0.000004  rs1408404600          snv  901920.0   \n",
       "...          ...        ...       ...           ...          ...       ...   \n",
       "256896       NaN     [C, A]       NaN         10156          NaN       NaN   \n",
       "256897       NaN     [T, C]       NaN         10352          NaN       NaN   \n",
       "256899       NaN     [T, A]       NaN         10154          NaN       NaN   \n",
       "256900       NaN     [A, C]       NaN         10153          NaN       NaN   \n",
       "256902       NaN     [G, A]       NaN         10152          NaN       NaN   \n",
       "\n",
       "                     id most_severe_consequence                     MAP  \\\n",
       "0          1_901913_T_G              start_lost   PLEKHN1_901912-901994   \n",
       "1          1_901916_G_A        missense_variant   PLEKHN1_901912-901994   \n",
       "2          1_901917_G_A      synonymous_variant   PLEKHN1_901912-901994   \n",
       "3          1_901919_A_C        missense_variant   PLEKHN1_901912-901994   \n",
       "4          1_901920_C_G        missense_variant   PLEKHN1_901912-901994   \n",
       "...                 ...                     ...                     ...   \n",
       "256896  X_154250707_C_A                     NaN  F8_154250685-154250827   \n",
       "256897  X_154250724_T_C                     NaN  F8_154250685-154250827   \n",
       "256899  X_154250739_T_A                     NaN  F8_154250685-154250827   \n",
       "256900  X_154250751_A_C                     NaN  F8_154250685-154250827   \n",
       "256902  X_154250785_G_A                     NaN  F8_154250685-154250827   \n",
       "\n",
       "           Gene  ... locus.contig locus.position  ALLELEID  \\\n",
       "0       PLEKHN1  ...          NaN            NaN       NaN   \n",
       "1       PLEKHN1  ...          NaN            NaN       NaN   \n",
       "2       PLEKHN1  ...          NaN            NaN       NaN   \n",
       "3       PLEKHN1  ...          NaN            NaN       NaN   \n",
       "4       PLEKHN1  ...          NaN            NaN       NaN   \n",
       "...         ...  ...          ...            ...       ...   \n",
       "256896       F8  ...            X    154250707.0   25195.0   \n",
       "256897       F8  ...            X    154250724.0   25391.0   \n",
       "256899       F8  ...            X    154250739.0   25193.0   \n",
       "256900       F8  ...            X    154250751.0   25192.0   \n",
       "256902       F8  ...            X    154250785.0   25191.0   \n",
       "\n",
       "                                 CLNREVSTAT      CLNSIG GENEINFO           RS  \\\n",
       "0                                       NaN         NaN      NaN          NaN   \n",
       "1                                       NaN         NaN      NaN          NaN   \n",
       "2                                       NaN         NaN      NaN          NaN   \n",
       "3                                       NaN         NaN      NaN          NaN   \n",
       "4                                       NaN         NaN      NaN          NaN   \n",
       "...                                     ...         ...      ...          ...   \n",
       "256896       no_assertion_criteria_provided  Pathogenic  F8:2157  [137852379]   \n",
       "256897       no_assertion_criteria_provided  Pathogenic  F8:2157  [137852476]   \n",
       "256899       no_assertion_criteria_provided  Pathogenic  F8:2157  [137852378]   \n",
       "256900       no_assertion_criteria_provided  Pathogenic  F8:2157  [137852377]   \n",
       "256902  criteria_provided,_single_submitter  Pathogenic  F8:2157  [387906432]   \n",
       "\n",
       "       RS_STARS            snpId      Status  \n",
       "0           NaN              NaN      Benign  \n",
       "1           NaN              NaN      Benign  \n",
       "2           NaN              NaN      Benign  \n",
       "3           NaN              NaN      Benign  \n",
       "4           NaN              NaN      Benign  \n",
       "...         ...              ...         ...  \n",
       "256896      0.0  X_154250707_C_A  Pathogenic  \n",
       "256897      0.0  X_154250724_T_C  Pathogenic  \n",
       "256899      0.0  X_154250739_T_A  Pathogenic  \n",
       "256900      0.0  X_154250751_A_C  Pathogenic  \n",
       "256902      1.0  X_154250785_G_A  Pathogenic  \n",
       "\n",
       "[2794825 rows x 21 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merge_variants = pd.concat(\n",
    "    [\n",
    "        gnomad.loc[~gnomad['id'].isin(clinvar.id.values.tolist())],\n",
    "        clinvar.loc[~clinvar['id'].isin(gnomad.id.values.tolist())]\n",
    "    ]\n",
    ")\n",
    "merge_variants.loc[merge_variants['CLNSIG'].isna() == True, 'Status'] = 'Benign'\n",
    "merge_variants['Status'] = merge_variants['Status'].fillna('Pathogenic')\n",
    "merge_variants"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "smaller-poetry",
   "metadata": {},
   "source": [
    "# Compute OR for Const VS Alt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "sound-publication",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>group</th>\n",
       "      <th>Cat</th>\n",
       "      <th>ratio</th>\n",
       "      <th>OR</th>\n",
       "      <th>SE</th>\n",
       "      <th>IC_low</th>\n",
       "      <th>IC_up</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>Const OK</th>\n",
       "      <th>Alt OK</th>\n",
       "      <th>Const_NOT OK</th>\n",
       "      <th>Alt_NOT OK</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Conservation</td>\n",
       "      <td>Unconserved</td>\n",
       "      <td>None</td>\n",
       "      <td>0.350292</td>\n",
       "      <td>0.043888</td>\n",
       "      <td>0.321419</td>\n",
       "      <td>0.381758</td>\n",
       "      <td>4.604263e-128</td>\n",
       "      <td>1742</td>\n",
       "      <td>1797</td>\n",
       "      <td>4735</td>\n",
       "      <td>1711</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Conservation</td>\n",
       "      <td>Conserved</td>\n",
       "      <td>None</td>\n",
       "      <td>2.854762</td>\n",
       "      <td>0.043888</td>\n",
       "      <td>2.619458</td>\n",
       "      <td>3.111203</td>\n",
       "      <td>4.604263e-128</td>\n",
       "      <td>4735</td>\n",
       "      <td>1711</td>\n",
       "      <td>1742</td>\n",
       "      <td>1797</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          group          Cat ratio        OR        SE    IC_low     IC_up  \\\n",
       "0  Conservation  Unconserved  None  0.350292  0.043888  0.321419  0.381758   \n",
       "1  Conservation    Conserved  None  2.854762  0.043888  2.619458  3.111203   \n",
       "\n",
       "          pvalue  Const OK  Alt OK  Const_NOT OK  Alt_NOT OK  id  \n",
       "0  4.604263e-128      1742    1797          4735        1711   1  \n",
       "1  4.604263e-128      4735    1711          1742        1797   2  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TMP TO LOAD CONFIG\n",
    "yaml = load_config_file(config_file=\"/home/weber/PycharmProjects/EXOTIC/clean/src/config_clean_clean.yaml\")\n",
    "\n",
    "complete_list_stats = list()\n",
    "\n",
    "list_df = [phylocsf_refseq]\n",
    "list_cols = [['phyloCSF_cat']]\n",
    "\n",
    "for df, cols in zip(list_df, list_cols):\n",
    "    tmp_m = pd.merge(refseq_exons.dropna(subset=['Ratio_num_bins_update']), df, on='MAP')\n",
    "    complete_list_stats.extend(compute_odds_ratio(tmp_m, cols))\n",
    "    \n",
    "df_or_stat = pd.DataFrame(complete_list_stats)\n",
    "df_or_stat['Cat'] = df_or_stat['Cat'].str.replace('_' , ' ')\n",
    "# df_or_stat.to_excel(yaml['3_EXONS_PROPERTIES']['Figures_data']['odds_ratio_const_alt'])\n",
    "df_or_stat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "authorized-chair",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TMP TO LOAD CONFIG\n",
    "yaml = load_config_file(config_file=\"/home/weber/PycharmProjects/EXOTIC/clean/src/config_clean_clean.yaml\")\n",
    "\n",
    "complete_list_stats = list()\n",
    "\n",
    "list_df = [phylocsf_refseq, gnomad, merge_variants, ccr]\n",
    "list_cols = [['phyloCSF_cat'], ['most_severe_consequence', 'AF_bins'], ['Status'], ['CCR_bins']]\n",
    "\n",
    "for df, cols in zip(list_df, list_cols):\n",
    "    tmp_m = pd.merge(refseq_exons.dropna(subset=['Ratio_num_bins_update']), df, on='MAP')\n",
    "    complete_list_stats.extend(compute_odds_ratio(tmp_m, cols))\n",
    "    \n",
    "df_or_stat = pd.DataFrame(complete_list_stats)\n",
    "df_or_stat['Cat'] = df_or_stat['Cat'].str.replace('_' , ' ')\n",
    "# df_or_stat.to_excel(yaml['3_EXONS_PROPERTIES']['Figures_data']['odds_ratio_const_alt'])\n",
    "df_or_stat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "sufficient-brazilian",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>group</th>\n",
       "      <th>Cat</th>\n",
       "      <th>ratio</th>\n",
       "      <th>OR</th>\n",
       "      <th>SE</th>\n",
       "      <th>IC_low</th>\n",
       "      <th>IC_up</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>Const OK</th>\n",
       "      <th>Alt OK</th>\n",
       "      <th>Const_NOT OK</th>\n",
       "      <th>Alt_NOT OK</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>Consequence</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.95</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.94</td>\n",
       "      <td>0.95</td>\n",
       "      <td>4.840154e-83</td>\n",
       "      <td>1297022</td>\n",
       "      <td>539512</td>\n",
       "      <td>669929</td>\n",
       "      <td>263960</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>Consequence</td>\n",
       "      <td>start lost</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.40</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.38</td>\n",
       "      <td>0.42</td>\n",
       "      <td>2.826548e-284</td>\n",
       "      <td>3105</td>\n",
       "      <td>3187</td>\n",
       "      <td>1963846</td>\n",
       "      <td>800285</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>Consequence</td>\n",
       "      <td>stop gained</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.86</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.85</td>\n",
       "      <td>0.88</td>\n",
       "      <td>1.155900e-51</td>\n",
       "      <td>34791</td>\n",
       "      <td>16399</td>\n",
       "      <td>1932160</td>\n",
       "      <td>787073</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>Consequence</td>\n",
       "      <td>stop lost</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.54</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.51</td>\n",
       "      <td>0.58</td>\n",
       "      <td>1.626188e-61</td>\n",
       "      <td>1802</td>\n",
       "      <td>1352</td>\n",
       "      <td>1965149</td>\n",
       "      <td>802120</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>Consequence</td>\n",
       "      <td>stop retained variant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.59</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.67</td>\n",
       "      <td>2.420526e-16</td>\n",
       "      <td>606</td>\n",
       "      <td>421</td>\n",
       "      <td>1966345</td>\n",
       "      <td>803051</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>Consequence</td>\n",
       "      <td>synonymous variant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.09</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.08</td>\n",
       "      <td>1.09</td>\n",
       "      <td>1.520289e-192</td>\n",
       "      <td>629625</td>\n",
       "      <td>242601</td>\n",
       "      <td>1337326</td>\n",
       "      <td>560871</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>Conservation</td>\n",
       "      <td>Conserved</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.85</td>\n",
       "      <td>0.04</td>\n",
       "      <td>2.62</td>\n",
       "      <td>3.11</td>\n",
       "      <td>4.604263e-128</td>\n",
       "      <td>4735</td>\n",
       "      <td>1711</td>\n",
       "      <td>1742</td>\n",
       "      <td>1797</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>Conservation</td>\n",
       "      <td>Unconserved</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.35</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.32</td>\n",
       "      <td>0.38</td>\n",
       "      <td>4.604263e-128</td>\n",
       "      <td>1742</td>\n",
       "      <td>1797</td>\n",
       "      <td>4735</td>\n",
       "      <td>1711</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>17</td>\n",
       "      <td>Constraint</td>\n",
       "      <td>0 - 20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.92</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.92</td>\n",
       "      <td>0.93</td>\n",
       "      <td>5.981530e-192</td>\n",
       "      <td>1981071</td>\n",
       "      <td>825930</td>\n",
       "      <td>604224</td>\n",
       "      <td>232163</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>18</td>\n",
       "      <td>Constraint</td>\n",
       "      <td>20 - 80</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.05</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.05</td>\n",
       "      <td>1.06</td>\n",
       "      <td>2.126345e-75</td>\n",
       "      <td>530950</td>\n",
       "      <td>208300</td>\n",
       "      <td>2054345</td>\n",
       "      <td>849793</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>20</td>\n",
       "      <td>Constraint</td>\n",
       "      <td>80 - 90</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.19</td>\n",
       "      <td>0.01</td>\n",
       "      <td>1.17</td>\n",
       "      <td>1.21</td>\n",
       "      <td>1.271965e-76</td>\n",
       "      <td>42924</td>\n",
       "      <td>14777</td>\n",
       "      <td>2542371</td>\n",
       "      <td>1043316</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>16</td>\n",
       "      <td>Constraint</td>\n",
       "      <td>90 - 95</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.32</td>\n",
       "      <td>0.02</td>\n",
       "      <td>1.28</td>\n",
       "      <td>1.36</td>\n",
       "      <td>6.719486e-74</td>\n",
       "      <td>17549</td>\n",
       "      <td>5457</td>\n",
       "      <td>2567746</td>\n",
       "      <td>1052636</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>19</td>\n",
       "      <td>Constraint</td>\n",
       "      <td>95 - 100</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.45</td>\n",
       "      <td>0.02</td>\n",
       "      <td>1.39</td>\n",
       "      <td>1.50</td>\n",
       "      <td>8.182799e-91</td>\n",
       "      <td>12801</td>\n",
       "      <td>3629</td>\n",
       "      <td>2572494</td>\n",
       "      <td>1054464</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>12</td>\n",
       "      <td>MAF</td>\n",
       "      <td>1e-01 - 1e+00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.87</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.83</td>\n",
       "      <td>0.91</td>\n",
       "      <td>1.886569e-09</td>\n",
       "      <td>6198</td>\n",
       "      <td>2901</td>\n",
       "      <td>1960753</td>\n",
       "      <td>800571</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>13</td>\n",
       "      <td>MAF</td>\n",
       "      <td>1e-02 - 1e-01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.91</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.88</td>\n",
       "      <td>0.95</td>\n",
       "      <td>5.083885e-05</td>\n",
       "      <td>6725</td>\n",
       "      <td>3004</td>\n",
       "      <td>1960226</td>\n",
       "      <td>800468</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>10</td>\n",
       "      <td>MAF</td>\n",
       "      <td>1e-03 - 1e-02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.95</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.93</td>\n",
       "      <td>0.97</td>\n",
       "      <td>1.584670e-05</td>\n",
       "      <td>23384</td>\n",
       "      <td>10055</td>\n",
       "      <td>1943567</td>\n",
       "      <td>793417</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>11</td>\n",
       "      <td>MAF</td>\n",
       "      <td>1e-04 - 1e-03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.95</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.94</td>\n",
       "      <td>0.96</td>\n",
       "      <td>5.223220e-15</td>\n",
       "      <td>87953</td>\n",
       "      <td>37664</td>\n",
       "      <td>1878998</td>\n",
       "      <td>765808</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>MAF</td>\n",
       "      <td>1e-05 - 1e-04</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.97</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.96</td>\n",
       "      <td>0.97</td>\n",
       "      <td>8.322589e-28</td>\n",
       "      <td>548051</td>\n",
       "      <td>229100</td>\n",
       "      <td>1418900</td>\n",
       "      <td>574372</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>MAF</td>\n",
       "      <td>1e-06 - 1e-05</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.05</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.04</td>\n",
       "      <td>1.05</td>\n",
       "      <td>1.423029e-57</td>\n",
       "      <td>1294640</td>\n",
       "      <td>520748</td>\n",
       "      <td>672311</td>\n",
       "      <td>282724</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>14</td>\n",
       "      <td>Status</td>\n",
       "      <td>Benign</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.92</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.89</td>\n",
       "      <td>0.94</td>\n",
       "      <td>1.273585e-09</td>\n",
       "      <td>1963132</td>\n",
       "      <td>802207</td>\n",
       "      <td>17767</td>\n",
       "      <td>6654</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>15</td>\n",
       "      <td>Status</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.09</td>\n",
       "      <td>0.01</td>\n",
       "      <td>1.06</td>\n",
       "      <td>1.12</td>\n",
       "      <td>1.273585e-09</td>\n",
       "      <td>17767</td>\n",
       "      <td>6654</td>\n",
       "      <td>1963132</td>\n",
       "      <td>802207</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Unnamed: 0         group                    Cat  ratio    OR    SE  \\\n",
       "3            3   Consequence       missense variant    NaN  0.95  0.00   \n",
       "6            6   Consequence             start lost    NaN  0.40  0.03   \n",
       "4            4   Consequence            stop gained    NaN  0.86  0.01   \n",
       "5            5   Consequence              stop lost    NaN  0.54  0.04   \n",
       "7            7   Consequence  stop retained variant    NaN  0.59  0.06   \n",
       "2            2   Consequence     synonymous variant    NaN  1.09  0.00   \n",
       "1            1  Conservation              Conserved    NaN  2.85  0.04   \n",
       "0            0  Conservation            Unconserved    NaN  0.35  0.04   \n",
       "17          17    Constraint                 0 - 20    NaN  0.92  0.00   \n",
       "18          18    Constraint                20 - 80    NaN  1.05  0.00   \n",
       "20          20    Constraint                80 - 90    NaN  1.19  0.01   \n",
       "16          16    Constraint                90 - 95    NaN  1.32  0.02   \n",
       "19          19    Constraint               95 - 100    NaN  1.45  0.02   \n",
       "12          12           MAF          1e-01 - 1e+00    NaN  0.87  0.02   \n",
       "13          13           MAF          1e-02 - 1e-01    NaN  0.91  0.02   \n",
       "10          10           MAF          1e-03 - 1e-02    NaN  0.95  0.01   \n",
       "11          11           MAF          1e-04 - 1e-03    NaN  0.95  0.01   \n",
       "9            9           MAF          1e-05 - 1e-04    NaN  0.97  0.00   \n",
       "8            8           MAF          1e-06 - 1e-05    NaN  1.05  0.00   \n",
       "14          14        Status                 Benign    NaN  0.92  0.01   \n",
       "15          15        Status             Pathogenic    NaN  1.09  0.01   \n",
       "\n",
       "    IC_low  IC_up         pvalue  Const OK  Alt OK  Const_NOT OK  Alt_NOT OK  \\\n",
       "3     0.94   0.95   4.840154e-83   1297022  539512        669929      263960   \n",
       "6     0.38   0.42  2.826548e-284      3105    3187       1963846      800285   \n",
       "4     0.85   0.88   1.155900e-51     34791   16399       1932160      787073   \n",
       "5     0.51   0.58   1.626188e-61      1802    1352       1965149      802120   \n",
       "7     0.52   0.67   2.420526e-16       606     421       1966345      803051   \n",
       "2     1.08   1.09  1.520289e-192    629625  242601       1337326      560871   \n",
       "1     2.62   3.11  4.604263e-128      4735    1711          1742        1797   \n",
       "0     0.32   0.38  4.604263e-128      1742    1797          4735        1711   \n",
       "17    0.92   0.93  5.981530e-192   1981071  825930        604224      232163   \n",
       "18    1.05   1.06   2.126345e-75    530950  208300       2054345      849793   \n",
       "20    1.17   1.21   1.271965e-76     42924   14777       2542371     1043316   \n",
       "16    1.28   1.36   6.719486e-74     17549    5457       2567746     1052636   \n",
       "19    1.39   1.50   8.182799e-91     12801    3629       2572494     1054464   \n",
       "12    0.83   0.91   1.886569e-09      6198    2901       1960753      800571   \n",
       "13    0.88   0.95   5.083885e-05      6725    3004       1960226      800468   \n",
       "10    0.93   0.97   1.584670e-05     23384   10055       1943567      793417   \n",
       "11    0.94   0.96   5.223220e-15     87953   37664       1878998      765808   \n",
       "9     0.96   0.97   8.322589e-28    548051  229100       1418900      574372   \n",
       "8     1.04   1.05   1.423029e-57   1294640  520748        672311      282724   \n",
       "14    0.89   0.94   1.273585e-09   1963132  802207         17767        6654   \n",
       "15    1.06   1.12   1.273585e-09     17767    6654       1963132      802207   \n",
       "\n",
       "    id  \n",
       "3    2  \n",
       "6    5  \n",
       "4    3  \n",
       "5    4  \n",
       "7    6  \n",
       "2    1  \n",
       "1    2  \n",
       "0    1  \n",
       "17   2  \n",
       "18   3  \n",
       "20   5  \n",
       "16   1  \n",
       "19   4  \n",
       "12   5  \n",
       "13   6  \n",
       "10   3  \n",
       "11   4  \n",
       "9    2  \n",
       "8    1  \n",
       "14   1  \n",
       "15   2  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp = pd.read_excel(yaml['3_EXONS_PROPERTIES']['Figures_data']['odds_ratio_const_alt'])\n",
    "tmp = tmp.sort_values(by=['group', 'Cat'])\n",
    "tmp[['OR', 'SE', 'IC_low', 'IC_up']] = tmp[['OR', 'SE', 'IC_low', 'IC_up']].round(2)\n",
    "tmp"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "capital-equation",
   "metadata": {},
   "source": [
    "# Compute OR for exonic gradient frequency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "likely-africa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>group</th>\n",
       "      <th>Cat</th>\n",
       "      <th>ratio</th>\n",
       "      <th>OR</th>\n",
       "      <th>SE</th>\n",
       "      <th>IC_low</th>\n",
       "      <th>IC_up</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>Const OK</th>\n",
       "      <th>Alt OK</th>\n",
       "      <th>Const_NOT OK</th>\n",
       "      <th>Alt_NOT OK</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Conservation</td>\n",
       "      <td>Unconserved</td>\n",
       "      <td>0 - 0.2</td>\n",
       "      <td>5.646330</td>\n",
       "      <td>0.124058</td>\n",
       "      <td>4.427575</td>\n",
       "      <td>7.200565</td>\n",
       "      <td>6.157649e-52</td>\n",
       "      <td>262</td>\n",
       "      <td>3277</td>\n",
       "      <td>90</td>\n",
       "      <td>6356</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Conservation</td>\n",
       "      <td>Conserved</td>\n",
       "      <td>0 - 0.2</td>\n",
       "      <td>0.177106</td>\n",
       "      <td>0.124058</td>\n",
       "      <td>0.138878</td>\n",
       "      <td>0.225857</td>\n",
       "      <td>6.157649e-52</td>\n",
       "      <td>90</td>\n",
       "      <td>6356</td>\n",
       "      <td>262</td>\n",
       "      <td>3277</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Conservation</td>\n",
       "      <td>Unconserved</td>\n",
       "      <td>0.2 - 0.4</td>\n",
       "      <td>3.205327</td>\n",
       "      <td>0.077339</td>\n",
       "      <td>2.754484</td>\n",
       "      <td>3.729963</td>\n",
       "      <td>9.444457e-53</td>\n",
       "      <td>473</td>\n",
       "      <td>3066</td>\n",
       "      <td>296</td>\n",
       "      <td>6150</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Conservation</td>\n",
       "      <td>Conserved</td>\n",
       "      <td>0.2 - 0.4</td>\n",
       "      <td>0.311981</td>\n",
       "      <td>0.077339</td>\n",
       "      <td>0.268099</td>\n",
       "      <td>0.363044</td>\n",
       "      <td>9.444457e-53</td>\n",
       "      <td>296</td>\n",
       "      <td>6150</td>\n",
       "      <td>473</td>\n",
       "      <td>3066</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Conservation</td>\n",
       "      <td>Unconserved</td>\n",
       "      <td>0.4 - 0.6</td>\n",
       "      <td>1.865827</td>\n",
       "      <td>0.059558</td>\n",
       "      <td>1.660253</td>\n",
       "      <td>2.096856</td>\n",
       "      <td>2.556626e-25</td>\n",
       "      <td>640</td>\n",
       "      <td>2899</td>\n",
       "      <td>682</td>\n",
       "      <td>5764</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>Constraint</td>\n",
       "      <td>90 - 95</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.034403</td>\n",
       "      <td>0.015551</td>\n",
       "      <td>1.003350</td>\n",
       "      <td>1.066417</td>\n",
       "      <td>2.972012e-02</td>\n",
       "      <td>17549</td>\n",
       "      <td>5457</td>\n",
       "      <td>2567746</td>\n",
       "      <td>825930</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>Constraint</td>\n",
       "      <td>0 - 20</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.826891</td>\n",
       "      <td>0.002859</td>\n",
       "      <td>0.822271</td>\n",
       "      <td>0.831537</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1981071</td>\n",
       "      <td>825930</td>\n",
       "      <td>604224</td>\n",
       "      <td>208300</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>Constraint</td>\n",
       "      <td>20 - 80</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.024788</td>\n",
       "      <td>0.002895</td>\n",
       "      <td>1.018990</td>\n",
       "      <td>1.030620</td>\n",
       "      <td>2.534699e-17</td>\n",
       "      <td>530950</td>\n",
       "      <td>208300</td>\n",
       "      <td>2054345</td>\n",
       "      <td>825930</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>Constraint</td>\n",
       "      <td>95 - 100</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.132520</td>\n",
       "      <td>0.018849</td>\n",
       "      <td>1.091444</td>\n",
       "      <td>1.175142</td>\n",
       "      <td>2.611431e-11</td>\n",
       "      <td>12801</td>\n",
       "      <td>3629</td>\n",
       "      <td>2572494</td>\n",
       "      <td>825930</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>Constraint</td>\n",
       "      <td>80 - 90</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.943666</td>\n",
       "      <td>0.009622</td>\n",
       "      <td>0.926037</td>\n",
       "      <td>0.961631</td>\n",
       "      <td>1.964136e-09</td>\n",
       "      <td>42924</td>\n",
       "      <td>14777</td>\n",
       "      <td>2542371</td>\n",
       "      <td>825930</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>126 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            group          Cat      ratio        OR        SE    IC_low  \\\n",
       "0    Conservation  Unconserved    0 - 0.2  5.646330  0.124058  4.427575   \n",
       "1    Conservation    Conserved    0 - 0.2  0.177106  0.124058  0.138878   \n",
       "2    Conservation  Unconserved  0.2 - 0.4  3.205327  0.077339  2.754484   \n",
       "3    Conservation    Conserved  0.2 - 0.4  0.311981  0.077339  0.268099   \n",
       "4    Conservation  Unconserved  0.4 - 0.6  1.865827  0.059558  1.660253   \n",
       "..            ...          ...        ...       ...       ...       ...   \n",
       "121    Constraint      90 - 95       1.00  1.034403  0.015551  1.003350   \n",
       "122    Constraint       0 - 20       1.00  0.826891  0.002859  0.822271   \n",
       "123    Constraint      20 - 80       1.00  1.024788  0.002895  1.018990   \n",
       "124    Constraint     95 - 100       1.00  1.132520  0.018849  1.091444   \n",
       "125    Constraint      80 - 90       1.00  0.943666  0.009622  0.926037   \n",
       "\n",
       "        IC_up        pvalue  Const OK  Alt OK  Const_NOT OK  Alt_NOT OK  id  \n",
       "0    7.200565  6.157649e-52       262    3277            90        6356   1  \n",
       "1    0.225857  6.157649e-52        90    6356           262        3277   2  \n",
       "2    3.729963  9.444457e-53       473    3066           296        6150   1  \n",
       "3    0.363044  9.444457e-53       296    6150           473        3066   2  \n",
       "4    2.096856  2.556626e-25       640    2899           682        5764   1  \n",
       "..        ...           ...       ...     ...           ...         ...  ..  \n",
       "121  1.066417  2.972012e-02     17549    5457       2567746      825930   1  \n",
       "122  0.831537  0.000000e+00   1981071  825930        604224      208300   2  \n",
       "123  1.030620  2.534699e-17    530950  208300       2054345      825930   3  \n",
       "124  1.175142  2.611431e-11     12801    3629       2572494      825930   4  \n",
       "125  0.961631  1.964136e-09     42924   14777       2542371      825930   5  \n",
       "\n",
       "[126 rows x 13 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TMP TO LOAD CONFIG\n",
    "yaml = load_config_file(config_file=\"/home/weber/PycharmProjects/EXOTIC/clean/src/config_clean_clean.yaml\")\n",
    "\n",
    "complete_list_stats = list()\n",
    "\n",
    "list_df = [phylocsf_refseq, gnomad, merge_variants, ccr]\n",
    "list_cols = [['phyloCSF_cat'], ['most_severe_consequence', 'AF_bins'], ['Status'], ['CCR_bins']]\n",
    "\n",
    "for df, cols in zip(list_df, list_cols):\n",
    "    tmp_m = pd.merge(refseq_exons.dropna(subset=['Ratio_num_bins_update']), df, on='MAP')\n",
    "    complete_list_stats.extend(compute_odds_ratio_alt(tmp_m, cols))\n",
    "    \n",
    "df_or_stat = pd.DataFrame(complete_list_stats)\n",
    "df_or_stat['Cat'] = df_or_stat['Cat'].str.replace('_' , ' ')\n",
    "df_or_stat.to_excel(yaml['3_EXONS_PROPERTIES']['Figures_data']['odds_ratio_const_alt_gradient'])\n",
    "df_or_stat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "through-chart",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>group</th>\n",
       "      <th>Cat</th>\n",
       "      <th>ratio</th>\n",
       "      <th>OR</th>\n",
       "      <th>SE</th>\n",
       "      <th>IC_low</th>\n",
       "      <th>IC_up</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>Const OK</th>\n",
       "      <th>Alt OK</th>\n",
       "      <th>Const_NOT OK</th>\n",
       "      <th>Alt_NOT OK</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>13</td>\n",
       "      <td>Consequence</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>0 - 0.2</td>\n",
       "      <td>1.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.99</td>\n",
       "      <td>1.03</td>\n",
       "      <td>4.080406e-01</td>\n",
       "      <td>29559</td>\n",
       "      <td>1806975</td>\n",
       "      <td>13942</td>\n",
       "      <td>859678</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>19</td>\n",
       "      <td>Consequence</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>0.2 - 0.4</td>\n",
       "      <td>0.98</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.97</td>\n",
       "      <td>0.99</td>\n",
       "      <td>2.264742e-03</td>\n",
       "      <td>94338</td>\n",
       "      <td>1742196</td>\n",
       "      <td>45795</td>\n",
       "      <td>830742</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>25</td>\n",
       "      <td>Consequence</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>0.4 - 0.6</td>\n",
       "      <td>0.97</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.96</td>\n",
       "      <td>0.98</td>\n",
       "      <td>9.953412e-14</td>\n",
       "      <td>222004</td>\n",
       "      <td>1614530</td>\n",
       "      <td>109259</td>\n",
       "      <td>771567</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>31</td>\n",
       "      <td>Consequence</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>0.6 - 0.8</td>\n",
       "      <td>0.97</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.96</td>\n",
       "      <td>0.98</td>\n",
       "      <td>1.760307e-10</td>\n",
       "      <td>158513</td>\n",
       "      <td>1678021</td>\n",
       "      <td>77832</td>\n",
       "      <td>800150</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>37</td>\n",
       "      <td>Consequence</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>0.8 - 0.99</td>\n",
       "      <td>0.97</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.96</td>\n",
       "      <td>0.99</td>\n",
       "      <td>5.061799e-03</td>\n",
       "      <td>35098</td>\n",
       "      <td>1801436</td>\n",
       "      <td>17132</td>\n",
       "      <td>856392</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>87</td>\n",
       "      <td>Status</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>0.2 - 0.4</td>\n",
       "      <td>0.54</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.59</td>\n",
       "      <td>6.384516e-67</td>\n",
       "      <td>689</td>\n",
       "      <td>23732</td>\n",
       "      <td>139975</td>\n",
       "      <td>2625364</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>89</td>\n",
       "      <td>Status</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>0.4 - 0.6</td>\n",
       "      <td>0.91</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.88</td>\n",
       "      <td>0.95</td>\n",
       "      <td>6.249360e-06</td>\n",
       "      <td>2693</td>\n",
       "      <td>21728</td>\n",
       "      <td>330748</td>\n",
       "      <td>2434591</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>91</td>\n",
       "      <td>Status</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>0.6 - 0.8</td>\n",
       "      <td>1.26</td>\n",
       "      <td>0.02</td>\n",
       "      <td>1.21</td>\n",
       "      <td>1.31</td>\n",
       "      <td>4.207859e-27</td>\n",
       "      <td>2571</td>\n",
       "      <td>21850</td>\n",
       "      <td>235839</td>\n",
       "      <td>2529500</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>93</td>\n",
       "      <td>Status</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>0.8 - 0.99</td>\n",
       "      <td>0.84</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.76</td>\n",
       "      <td>0.93</td>\n",
       "      <td>6.603780e-04</td>\n",
       "      <td>389</td>\n",
       "      <td>24032</td>\n",
       "      <td>52170</td>\n",
       "      <td>2713169</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>95</td>\n",
       "      <td>Status</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.09</td>\n",
       "      <td>0.01</td>\n",
       "      <td>1.06</td>\n",
       "      <td>1.12</td>\n",
       "      <td>1.273585e-09</td>\n",
       "      <td>17767</td>\n",
       "      <td>6654</td>\n",
       "      <td>1963132</td>\n",
       "      <td>802207</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>126 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Unnamed: 0        group               Cat       ratio    OR    SE  IC_low  \\\n",
       "13          13  Consequence  missense variant     0 - 0.2  1.01  0.01    0.99   \n",
       "19          19  Consequence  missense variant   0.2 - 0.4  0.98  0.01    0.97   \n",
       "25          25  Consequence  missense variant   0.4 - 0.6  0.97  0.00    0.96   \n",
       "31          31  Consequence  missense variant   0.6 - 0.8  0.97  0.00    0.96   \n",
       "37          37  Consequence  missense variant  0.8 - 0.99  0.97  0.01    0.96   \n",
       "..         ...          ...               ...         ...   ...   ...     ...   \n",
       "87          87       Status        Pathogenic   0.2 - 0.4  0.54  0.04    0.50   \n",
       "89          89       Status        Pathogenic   0.4 - 0.6  0.91  0.02    0.88   \n",
       "91          91       Status        Pathogenic   0.6 - 0.8  1.26  0.02    1.21   \n",
       "93          93       Status        Pathogenic  0.8 - 0.99  0.84  0.05    0.76   \n",
       "95          95       Status        Pathogenic        1.00  1.09  0.01    1.06   \n",
       "\n",
       "    IC_up        pvalue  Const OK   Alt OK  Const_NOT OK  Alt_NOT OK  id  \n",
       "13   1.03  4.080406e-01     29559  1806975         13942      859678   2  \n",
       "19   0.99  2.264742e-03     94338  1742196         45795      830742   2  \n",
       "25   0.98  9.953412e-14    222004  1614530        109259      771567   2  \n",
       "31   0.98  1.760307e-10    158513  1678021         77832      800150   2  \n",
       "37   0.99  5.061799e-03     35098  1801436         17132      856392   2  \n",
       "..    ...           ...       ...      ...           ...         ...  ..  \n",
       "87   0.59  6.384516e-67       689    23732        139975     2625364   2  \n",
       "89   0.95  6.249360e-06      2693    21728        330748     2434591   2  \n",
       "91   1.31  4.207859e-27      2571    21850        235839     2529500   2  \n",
       "93   0.93  6.603780e-04       389    24032         52170     2713169   2  \n",
       "95   1.12  1.273585e-09     17767     6654       1963132      802207   2  \n",
       "\n",
       "[126 rows x 14 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp = pd.read_excel(yaml['3_EXONS_PROPERTIES']['Figures_data']['odds_ratio_const_alt_gradient'])\n",
    "tmp = tmp.sort_values(by=['group', 'Cat', 'ratio.1'])\n",
    "tmp[['OR', 'SE', 'IC_low', 'IC_up']] = tmp[['OR', 'SE', 'IC_low', 'IC_up']].round(2)\n",
    "tmp"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.7.6 64-bit ('ExoCarto': conda)",
   "language": "python",
   "name": "python37664bitexocartoconda47cfe11889aa4008ba405d9d7b40c893"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
