{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from tqdm.notebook import tqdm\n",
    "import multiprocessing\n",
    "import parmap\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>HGNC ID</th>\n",
       "      <th>HGNC symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000210049</td>\n",
       "      <td>7481</td>\n",
       "      <td>MT-TF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000211459</td>\n",
       "      <td>7470</td>\n",
       "      <td>MT-RNR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000210077</td>\n",
       "      <td>7500</td>\n",
       "      <td>MT-TV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000210082</td>\n",
       "      <td>7471</td>\n",
       "      <td>MT-RNR2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000209082</td>\n",
       "      <td>7490</td>\n",
       "      <td>MT-TL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67150</th>\n",
       "      <td>ENSG00000288200</td>\n",
       "      <td>49825</td>\n",
       "      <td>SNRPCP10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67152</th>\n",
       "      <td>ENSG00000288476</td>\n",
       "      <td>35781</td>\n",
       "      <td>RPS3AP15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67155</th>\n",
       "      <td>ENSG00000284997</td>\n",
       "      <td>14531</td>\n",
       "      <td>TUBAP2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67156</th>\n",
       "      <td>ENSG00000284805</td>\n",
       "      <td>47786</td>\n",
       "      <td>RNU6-823P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67157</th>\n",
       "      <td>ENSG00000284869</td>\n",
       "      <td>24614</td>\n",
       "      <td>EEFSEC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>42848 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Gene stable ID HGNC ID HGNC symbol\n",
       "0      ENSG00000210049    7481       MT-TF\n",
       "1      ENSG00000211459    7470     MT-RNR1\n",
       "2      ENSG00000210077    7500       MT-TV\n",
       "3      ENSG00000210082    7471     MT-RNR2\n",
       "4      ENSG00000209082    7490      MT-TL1\n",
       "...                ...     ...         ...\n",
       "67150  ENSG00000288200   49825    SNRPCP10\n",
       "67152  ENSG00000288476   35781    RPS3AP15\n",
       "67155  ENSG00000284997   14531      TUBAP2\n",
       "67156  ENSG00000284805   47786   RNU6-823P\n",
       "67157  ENSG00000284869   24614      EEFSEC\n",
       "\n",
       "[42848 rows x 3 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biomart_ensembl = pd.read_csv('/home/weber/PycharmProjects/ExoCarto/data/0_external/BIOMART_HGNC_ENSEMBL.txt.gz', sep='\\t', compression='gzip')\n",
    "biomart_ensembl = biomart_ensembl.dropna(subset=['HGNC ID'])\n",
    "biomart_ensembl['HGNC ID'] = biomart_ensembl['HGNC ID'].astype(str)\n",
    "biomart_ensembl['HGNC ID'] = biomart_ensembl['HGNC ID'].str.replace('HGNC:', '')\n",
    "biomart_ensembl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ptex_file = \"/gstock/biolo_datasets/variation/gnomAD/latest/gtex/gtex_gnomad.parquet\"\n",
    "rawdf_ptex = pd.read_parquet(ptex_file)\n",
    "rawdf_ptex = pd.merge(rawdf_ptex, biomart_ensembl, right_on='HGNC symbol', left_on='symbol')\n",
    "rawdf_ptex = rawdf_ptex[['HGNC ID', 'symbol', 'ensg', 'chrom', 'pos'] + list(rawdf_ptex.columns)[4:-3]]\n",
    "rawdf_ptex['HGNC ID'] = rawdf_ptex['HGNC ID'].astype(int)\n",
    "rawdf_ptex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "original_pext = pd.read_csv('/gstock/biolo_datasets/variation/gnomAD/latest/gtex/all.baselevel.021620.tsv.bgz', compression='gzip', sep='\\t', low_memory=False)\n",
    "pext_modified = original_pext.dropna(subset=[list(original_pext.columns)[-1]])\n",
    "pext_modified[['chrom', 'pos']] = pext_modified['locus'].str.split(':', expand=True)\n",
    "pext_modified = pd.merge(pext_modified, biomart_ensembl, right_on='HGNC symbol', left_on='symbol', how='left')\n",
    "pext_modified = pext_modified[['HGNC ID', 'symbol', 'ensg', 'chrom', 'pos'] + list(pext_modified.columns)[4:-3]]\n",
    "# original_pext['HGNC ID'] = original_pext['HGNC ID'].astype(int)\n",
    "pext_modified\n",
    "pext_modified.loc[:,~pext_modified.columns.duplicated()].to_parquet('/gstock/biolo_datasets/variation/gnomAD/latest/gtex/gtex_gnomad_corrected.parquet')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HGNC ID</th>\n",
       "      <th>symbol</th>\n",
       "      <th>ensg</th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>Prostate</th>\n",
       "      <th>Spleen</th>\n",
       "      <th>Brain_FrontalCortex_BA9_</th>\n",
       "      <th>SmallIntestine_TerminalIleum</th>\n",
       "      <th>MinorSalivaryGland</th>\n",
       "      <th>...</th>\n",
       "      <th>Brain_Anteriorcingulatecortex_BA24_</th>\n",
       "      <th>Brain_Putamen_basalganglia_</th>\n",
       "      <th>Brain_Caudate_basalganglia_</th>\n",
       "      <th>Colon_Sigmoid</th>\n",
       "      <th>Cervix_Endocervix</th>\n",
       "      <th>Ovary</th>\n",
       "      <th>Esophagus_GastroesophagealJunction</th>\n",
       "      <th>Testis</th>\n",
       "      <th>Brain_Amygdala</th>\n",
       "      <th>mean_proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "      <td>AL627309.1</td>\n",
       "      <td>ENSG00000237683</td>\n",
       "      <td>1</td>\n",
       "      <td>138530</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>None</td>\n",
       "      <td>AL627309.1</td>\n",
       "      <td>ENSG00000237683</td>\n",
       "      <td>1</td>\n",
       "      <td>138531</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>None</td>\n",
       "      <td>AL627309.1</td>\n",
       "      <td>ENSG00000237683</td>\n",
       "      <td>1</td>\n",
       "      <td>138532</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>None</td>\n",
       "      <td>AL627309.1</td>\n",
       "      <td>ENSG00000237683</td>\n",
       "      <td>1</td>\n",
       "      <td>138533</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>None</td>\n",
       "      <td>AL627309.1</td>\n",
       "      <td>ENSG00000237683</td>\n",
       "      <td>1</td>\n",
       "      <td>138534</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38837009</th>\n",
       "      <td>1809</td>\n",
       "      <td>CDY1</td>\n",
       "      <td>ENSG00000172288</td>\n",
       "      <td>Y</td>\n",
       "      <td>27770670</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.26316</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.13158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38837010</th>\n",
       "      <td>1809</td>\n",
       "      <td>CDY1</td>\n",
       "      <td>ENSG00000172288</td>\n",
       "      <td>Y</td>\n",
       "      <td>27770671</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.26316</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.13158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38837011</th>\n",
       "      <td>1809</td>\n",
       "      <td>CDY1</td>\n",
       "      <td>ENSG00000172288</td>\n",
       "      <td>Y</td>\n",
       "      <td>27770672</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.26316</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.13158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38837012</th>\n",
       "      <td>1809</td>\n",
       "      <td>CDY1</td>\n",
       "      <td>ENSG00000172288</td>\n",
       "      <td>Y</td>\n",
       "      <td>27770673</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.26316</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.13158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38837013</th>\n",
       "      <td>1809</td>\n",
       "      <td>CDY1</td>\n",
       "      <td>ENSG00000172288</td>\n",
       "      <td>Y</td>\n",
       "      <td>27770674</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.26316</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.13158</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>38837014 rows × 58 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         HGNC ID      symbol             ensg chrom       pos  Prostate  \\\n",
       "0           None  AL627309.1  ENSG00000237683     1    138530       1.0   \n",
       "1           None  AL627309.1  ENSG00000237683     1    138531       1.0   \n",
       "2           None  AL627309.1  ENSG00000237683     1    138532       1.0   \n",
       "3           None  AL627309.1  ENSG00000237683     1    138533       1.0   \n",
       "4           None  AL627309.1  ENSG00000237683     1    138534       1.0   \n",
       "...          ...         ...              ...   ...       ...       ...   \n",
       "38837009    1809        CDY1  ENSG00000172288     Y  27770670       NaN   \n",
       "38837010    1809        CDY1  ENSG00000172288     Y  27770671       NaN   \n",
       "38837011    1809        CDY1  ENSG00000172288     Y  27770672       NaN   \n",
       "38837012    1809        CDY1  ENSG00000172288     Y  27770673       NaN   \n",
       "38837013    1809        CDY1  ENSG00000172288     Y  27770674       NaN   \n",
       "\n",
       "          Spleen  Brain_FrontalCortex_BA9_  SmallIntestine_TerminalIleum  \\\n",
       "0            1.0                       1.0                           1.0   \n",
       "1            1.0                       1.0                           1.0   \n",
       "2            1.0                       1.0                           1.0   \n",
       "3            1.0                       1.0                           1.0   \n",
       "4            1.0                       1.0                           1.0   \n",
       "...          ...                       ...                           ...   \n",
       "38837009     NaN                       NaN                           NaN   \n",
       "38837010     NaN                       NaN                           NaN   \n",
       "38837011     NaN                       NaN                           NaN   \n",
       "38837012     NaN                       NaN                           NaN   \n",
       "38837013     NaN                       NaN                           NaN   \n",
       "\n",
       "          MinorSalivaryGland  ...  Brain_Anteriorcingulatecortex_BA24_  \\\n",
       "0                        1.0  ...                                  1.0   \n",
       "1                        1.0  ...                                  1.0   \n",
       "2                        1.0  ...                                  1.0   \n",
       "3                        1.0  ...                                  1.0   \n",
       "4                        1.0  ...                                  1.0   \n",
       "...                      ...  ...                                  ...   \n",
       "38837009                 NaN  ...                                  NaN   \n",
       "38837010                 NaN  ...                                  NaN   \n",
       "38837011                 NaN  ...                                  NaN   \n",
       "38837012                 NaN  ...                                  NaN   \n",
       "38837013                 NaN  ...                                  NaN   \n",
       "\n",
       "          Brain_Putamen_basalganglia_  Brain_Caudate_basalganglia_  \\\n",
       "0                                 1.0                          1.0   \n",
       "1                                 1.0                          1.0   \n",
       "2                                 1.0                          1.0   \n",
       "3                                 1.0                          1.0   \n",
       "4                                 1.0                          1.0   \n",
       "...                               ...                          ...   \n",
       "38837009                          NaN                          NaN   \n",
       "38837010                          NaN                          NaN   \n",
       "38837011                          NaN                          NaN   \n",
       "38837012                          NaN                          NaN   \n",
       "38837013                          NaN                          NaN   \n",
       "\n",
       "          Colon_Sigmoid  Cervix_Endocervix  Ovary  \\\n",
       "0                   1.0                1.0    1.0   \n",
       "1                   1.0                1.0    1.0   \n",
       "2                   1.0                1.0    1.0   \n",
       "3                   1.0                1.0    1.0   \n",
       "4                   1.0                1.0    1.0   \n",
       "...                 ...                ...    ...   \n",
       "38837009            NaN                NaN    NaN   \n",
       "38837010            NaN                NaN    NaN   \n",
       "38837011            NaN                NaN    NaN   \n",
       "38837012            NaN                NaN    NaN   \n",
       "38837013            NaN                NaN    NaN   \n",
       "\n",
       "          Esophagus_GastroesophagealJunction   Testis  Brain_Amygdala  \\\n",
       "0                                        1.0  1.00000             1.0   \n",
       "1                                        1.0  1.00000             1.0   \n",
       "2                                        1.0  1.00000             1.0   \n",
       "3                                        1.0  1.00000             1.0   \n",
       "4                                        1.0  1.00000             1.0   \n",
       "...                                      ...      ...             ...   \n",
       "38837009                                 NaN  0.26316             NaN   \n",
       "38837010                                 NaN  0.26316             NaN   \n",
       "38837011                                 NaN  0.26316             NaN   \n",
       "38837012                                 NaN  0.26316             NaN   \n",
       "38837013                                 NaN  0.26316             NaN   \n",
       "\n",
       "          mean_proportion  \n",
       "0                 1.00000  \n",
       "1                 1.00000  \n",
       "2                 1.00000  \n",
       "3                 1.00000  \n",
       "4                 1.00000  \n",
       "...                   ...  \n",
       "38837009          0.13158  \n",
       "38837010          0.13158  \n",
       "38837011          0.13158  \n",
       "38837012          0.13158  \n",
       "38837013          0.13158  \n",
       "\n",
       "[38837014 rows x 58 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pext_modified = pd.read_parquet('/gstock/biolo_datasets/variation/gnomAD/latest/gtex/gtex_gnomad_corrected.parquet')\n",
    "pext_modified"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['HGNC ID', 'symbol', 'ensg', 'chrom', 'pos', 'Prostate', 'Spleen',\n",
       "       'Brain_FrontalCortex_BA9_', 'SmallIntestine_TerminalIleum',\n",
       "       'MinorSalivaryGland', 'Artery_Coronary', 'Skin_SunExposed_Lowerleg_',\n",
       "       'Cells_EBV_transformedlymphocytes', 'Brain_Hippocampus',\n",
       "       'Esophagus_Muscularis', 'Brain_Nucleusaccumbens_basalganglia_',\n",
       "       'Artery_Tibial', 'Brain_Hypothalamus', 'Adipose_Visceral_Omentum_',\n",
       "       'Cervix_Ectocervix', 'Brain_Spinalcord_cervicalc_1_',\n",
       "       'Brain_CerebellarHemisphere', 'Nerve_Tibial', 'Breast_MammaryTissue',\n",
       "       'Liver', 'Skin_NotSunExposed_Suprapubic_', 'AdrenalGland', 'Vagina',\n",
       "       'Pancreas', 'Lung', 'FallopianTube', 'Pituitary', 'Muscle_Skeletal',\n",
       "       'Colon_Transverse', 'Artery_Aorta', 'Heart_AtrialAppendage',\n",
       "       'Adipose_Subcutaneous', 'Esophagus_Mucosa', 'Heart_LeftVentricle',\n",
       "       'Brain_Cerebellum', 'Brain_Cortex', 'Thyroid', 'Brain_Substantianigra',\n",
       "       'Kidney_Cortex', 'Uterus', 'Stomach', 'WholeBlood', 'Bladder',\n",
       "       'Brain_Anteriorcingulatecortex_BA24_', 'Brain_Putamen_basalganglia_',\n",
       "       'Brain_Caudate_basalganglia_', 'Colon_Sigmoid', 'Cervix_Endocervix',\n",
       "       'Ovary', 'Esophagus_GastroesophagealJunction', 'Testis',\n",
       "       'Brain_Amygdala', 'mean_proportion'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pext_modified.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "pext_modified = pext_modified.dropna(subset=['HGNC ID'])\n",
    "pext_modified['HGNC ID'] = pext_modified['HGNC ID'].astype(int)\n",
    "pext_modified['pos'] = pext_modified['pos'].astype(int)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HGNC ID</th>\n",
       "      <th>symbol</th>\n",
       "      <th>ensg</th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>Prostate</th>\n",
       "      <th>Spleen</th>\n",
       "      <th>Brain_FrontalCortex_BA9_</th>\n",
       "      <th>SmallIntestine_TerminalIleum</th>\n",
       "      <th>MinorSalivaryGland</th>\n",
       "      <th>...</th>\n",
       "      <th>Brain_Anteriorcingulatecortex_BA24_</th>\n",
       "      <th>Brain_Putamen_basalganglia_</th>\n",
       "      <th>Brain_Caudate_basalganglia_</th>\n",
       "      <th>Colon_Sigmoid</th>\n",
       "      <th>Cervix_Endocervix</th>\n",
       "      <th>Ovary</th>\n",
       "      <th>Esophagus_GastroesophagealJunction</th>\n",
       "      <th>Testis</th>\n",
       "      <th>Brain_Amygdala</th>\n",
       "      <th>mean_proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>780</th>\n",
       "      <td>31275</td>\n",
       "      <td>OR4F29</td>\n",
       "      <td>ENSG00000235249</td>\n",
       "      <td>1</td>\n",
       "      <td>367659</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>781</th>\n",
       "      <td>31275</td>\n",
       "      <td>OR4F29</td>\n",
       "      <td>ENSG00000235249</td>\n",
       "      <td>1</td>\n",
       "      <td>367660</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>782</th>\n",
       "      <td>31275</td>\n",
       "      <td>OR4F29</td>\n",
       "      <td>ENSG00000235249</td>\n",
       "      <td>1</td>\n",
       "      <td>367661</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>783</th>\n",
       "      <td>31275</td>\n",
       "      <td>OR4F29</td>\n",
       "      <td>ENSG00000235249</td>\n",
       "      <td>1</td>\n",
       "      <td>367662</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>784</th>\n",
       "      <td>31275</td>\n",
       "      <td>OR4F29</td>\n",
       "      <td>ENSG00000235249</td>\n",
       "      <td>1</td>\n",
       "      <td>367663</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38837009</th>\n",
       "      <td>1809</td>\n",
       "      <td>CDY1</td>\n",
       "      <td>ENSG00000172288</td>\n",
       "      <td>Y</td>\n",
       "      <td>27770670</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.26316</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.13158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38837010</th>\n",
       "      <td>1809</td>\n",
       "      <td>CDY1</td>\n",
       "      <td>ENSG00000172288</td>\n",
       "      <td>Y</td>\n",
       "      <td>27770671</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.26316</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.13158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38837011</th>\n",
       "      <td>1809</td>\n",
       "      <td>CDY1</td>\n",
       "      <td>ENSG00000172288</td>\n",
       "      <td>Y</td>\n",
       "      <td>27770672</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.26316</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.13158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38837012</th>\n",
       "      <td>1809</td>\n",
       "      <td>CDY1</td>\n",
       "      <td>ENSG00000172288</td>\n",
       "      <td>Y</td>\n",
       "      <td>27770673</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.26316</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.13158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38837013</th>\n",
       "      <td>1809</td>\n",
       "      <td>CDY1</td>\n",
       "      <td>ENSG00000172288</td>\n",
       "      <td>Y</td>\n",
       "      <td>27770674</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.26316</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.13158</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>36705587 rows × 58 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          HGNC ID  symbol             ensg chrom       pos  Prostate  Spleen  \\\n",
       "780         31275  OR4F29  ENSG00000235249     1    367659       1.0     1.0   \n",
       "781         31275  OR4F29  ENSG00000235249     1    367660       1.0     1.0   \n",
       "782         31275  OR4F29  ENSG00000235249     1    367661       1.0     1.0   \n",
       "783         31275  OR4F29  ENSG00000235249     1    367662       1.0     1.0   \n",
       "784         31275  OR4F29  ENSG00000235249     1    367663       1.0     1.0   \n",
       "...           ...     ...              ...   ...       ...       ...     ...   \n",
       "38837009     1809    CDY1  ENSG00000172288     Y  27770670       NaN     NaN   \n",
       "38837010     1809    CDY1  ENSG00000172288     Y  27770671       NaN     NaN   \n",
       "38837011     1809    CDY1  ENSG00000172288     Y  27770672       NaN     NaN   \n",
       "38837012     1809    CDY1  ENSG00000172288     Y  27770673       NaN     NaN   \n",
       "38837013     1809    CDY1  ENSG00000172288     Y  27770674       NaN     NaN   \n",
       "\n",
       "          Brain_FrontalCortex_BA9_  SmallIntestine_TerminalIleum  \\\n",
       "780                            1.0                           1.0   \n",
       "781                            1.0                           1.0   \n",
       "782                            1.0                           1.0   \n",
       "783                            1.0                           1.0   \n",
       "784                            1.0                           1.0   \n",
       "...                            ...                           ...   \n",
       "38837009                       NaN                           NaN   \n",
       "38837010                       NaN                           NaN   \n",
       "38837011                       NaN                           NaN   \n",
       "38837012                       NaN                           NaN   \n",
       "38837013                       NaN                           NaN   \n",
       "\n",
       "          MinorSalivaryGland  ...  Brain_Anteriorcingulatecortex_BA24_  \\\n",
       "780                      NaN  ...                                  NaN   \n",
       "781                      NaN  ...                                  NaN   \n",
       "782                      NaN  ...                                  NaN   \n",
       "783                      NaN  ...                                  NaN   \n",
       "784                      NaN  ...                                  NaN   \n",
       "...                      ...  ...                                  ...   \n",
       "38837009                 NaN  ...                                  NaN   \n",
       "38837010                 NaN  ...                                  NaN   \n",
       "38837011                 NaN  ...                                  NaN   \n",
       "38837012                 NaN  ...                                  NaN   \n",
       "38837013                 NaN  ...                                  NaN   \n",
       "\n",
       "          Brain_Putamen_basalganglia_  Brain_Caudate_basalganglia_  \\\n",
       "780                               1.0                          1.0   \n",
       "781                               1.0                          1.0   \n",
       "782                               1.0                          1.0   \n",
       "783                               1.0                          1.0   \n",
       "784                               1.0                          1.0   \n",
       "...                               ...                          ...   \n",
       "38837009                          NaN                          NaN   \n",
       "38837010                          NaN                          NaN   \n",
       "38837011                          NaN                          NaN   \n",
       "38837012                          NaN                          NaN   \n",
       "38837013                          NaN                          NaN   \n",
       "\n",
       "          Colon_Sigmoid  Cervix_Endocervix  Ovary  \\\n",
       "780                 NaN                1.0    1.0   \n",
       "781                 NaN                1.0    1.0   \n",
       "782                 NaN                1.0    1.0   \n",
       "783                 NaN                1.0    1.0   \n",
       "784                 NaN                1.0    1.0   \n",
       "...                 ...                ...    ...   \n",
       "38837009            NaN                NaN    NaN   \n",
       "38837010            NaN                NaN    NaN   \n",
       "38837011            NaN                NaN    NaN   \n",
       "38837012            NaN                NaN    NaN   \n",
       "38837013            NaN                NaN    NaN   \n",
       "\n",
       "          Esophagus_GastroesophagealJunction   Testis  Brain_Amygdala  \\\n",
       "780                                      NaN  1.00000             NaN   \n",
       "781                                      NaN  1.00000             NaN   \n",
       "782                                      NaN  1.00000             NaN   \n",
       "783                                      NaN  1.00000             NaN   \n",
       "784                                      NaN  1.00000             NaN   \n",
       "...                                      ...      ...             ...   \n",
       "38837009                                 NaN  0.26316             NaN   \n",
       "38837010                                 NaN  0.26316             NaN   \n",
       "38837011                                 NaN  0.26316             NaN   \n",
       "38837012                                 NaN  0.26316             NaN   \n",
       "38837013                                 NaN  0.26316             NaN   \n",
       "\n",
       "          mean_proportion  \n",
       "780               1.00000  \n",
       "781               1.00000  \n",
       "782               1.00000  \n",
       "783               1.00000  \n",
       "784               1.00000  \n",
       "...                   ...  \n",
       "38837009          0.13158  \n",
       "38837010          0.13158  \n",
       "38837011          0.13158  \n",
       "38837012          0.13158  \n",
       "38837013          0.13158  \n",
       "\n",
       "[36705587 rows x 58 columns]"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pext_modified"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/weber/.conda/envs/ExoCarto/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3146: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Chrom</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>ranges</th>\n",
       "      <th>Length</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>HGNC</th>\n",
       "      <th>CDS_representation</th>\n",
       "      <th>Ratio</th>\n",
       "      <th>Sharing_status</th>\n",
       "      <th>Sources</th>\n",
       "      <th>Count_CDS_alternative</th>\n",
       "      <th>Count_CDS_constitutive</th>\n",
       "      <th>mRNA_IDS</th>\n",
       "      <th>mRNA_nb</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>110056</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>19</td>\n",
       "      <td>58858388</td>\n",
       "      <td>58858395</td>\n",
       "      <td>58858388-58858395</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>Unique</td>\n",
       "      <td>1.0/1</td>\n",
       "      <td>False</td>\n",
       "      <td>[(58858388, 58858395, '1.0/1')]</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>['NM_130786.4']</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110057</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>19</td>\n",
       "      <td>58858719</td>\n",
       "      <td>58859006</td>\n",
       "      <td>58858719-58859006</td>\n",
       "      <td>287</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>Unique</td>\n",
       "      <td>1.0/1</td>\n",
       "      <td>False</td>\n",
       "      <td>[(58858719, 58859006, '1.0/1')]</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>['NM_130786.4']</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110058</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>19</td>\n",
       "      <td>58861736</td>\n",
       "      <td>58862017</td>\n",
       "      <td>58861736-58862017</td>\n",
       "      <td>281</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>Unique</td>\n",
       "      <td>1.0/1</td>\n",
       "      <td>False</td>\n",
       "      <td>[(58861736, 58862017, '1.0/1')]</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>['NM_130786.4']</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110059</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>19</td>\n",
       "      <td>58862757</td>\n",
       "      <td>58863053</td>\n",
       "      <td>58862757-58863053</td>\n",
       "      <td>296</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>Unique</td>\n",
       "      <td>1.0/1</td>\n",
       "      <td>False</td>\n",
       "      <td>[(58862757, 58863053, '1.0/1')]</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>['NM_130786.4']</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110060</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>19</td>\n",
       "      <td>58863649</td>\n",
       "      <td>58863921</td>\n",
       "      <td>58863649-58863921</td>\n",
       "      <td>272</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>Unique</td>\n",
       "      <td>1.0/1</td>\n",
       "      <td>False</td>\n",
       "      <td>[(58863649, 58863921, '1.0/1')]</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>['NM_130786.4']</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79640</th>\n",
       "      <td>DERPC</td>\n",
       "      <td>16</td>\n",
       "      <td>69152757</td>\n",
       "      <td>69154010</td>\n",
       "      <td>69152757-69154010</td>\n",
       "      <td>1253</td>\n",
       "      <td>113455421</td>\n",
       "      <td>54084</td>\n",
       "      <td>Variable_region</td>\n",
       "      <td>4.0/7</td>\n",
       "      <td>True</td>\n",
       "      <td>[(69152757, 69154331, '4.0/7')]</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>['NM_001366605.1', 'NM_001366603.1', 'NM_00104...</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79641</th>\n",
       "      <td>DERPC</td>\n",
       "      <td>16</td>\n",
       "      <td>69154010</td>\n",
       "      <td>69154331</td>\n",
       "      <td>69154010-69154331</td>\n",
       "      <td>321</td>\n",
       "      <td>113455421</td>\n",
       "      <td>54084</td>\n",
       "      <td>Variable_region</td>\n",
       "      <td>7.0/7</td>\n",
       "      <td>True</td>\n",
       "      <td>[(69152757, 69154331, '4.0/7'), (69154010, 691...</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>['NM_001366605.1', 'NM_001366603.1', 'NM_00104...</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79642</th>\n",
       "      <td>DERPC</td>\n",
       "      <td>16</td>\n",
       "      <td>69154010</td>\n",
       "      <td>69154331</td>\n",
       "      <td>69154010-69154331</td>\n",
       "      <td>321</td>\n",
       "      <td>113455421</td>\n",
       "      <td>54084</td>\n",
       "      <td>Variable_region</td>\n",
       "      <td>7.0/7</td>\n",
       "      <td>True</td>\n",
       "      <td>[(69152757, 69154331, '4.0/7'), (69154010, 691...</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>['NM_001366605.1', 'NM_001366603.1', 'NM_00104...</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9453</th>\n",
       "      <td>LINC02801</td>\n",
       "      <td>1</td>\n",
       "      <td>87709298</td>\n",
       "      <td>87709320</td>\n",
       "      <td>87709298-87709320</td>\n",
       "      <td>22</td>\n",
       "      <td>101927844</td>\n",
       "      <td>54325</td>\n",
       "      <td>Unique</td>\n",
       "      <td>1.0/1</td>\n",
       "      <td>False</td>\n",
       "      <td>[(87709298, 87709320, '1.0/1')]</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>['NM_001290053.1']</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9454</th>\n",
       "      <td>LINC02801</td>\n",
       "      <td>1</td>\n",
       "      <td>87716184</td>\n",
       "      <td>87716508</td>\n",
       "      <td>87716184-87716508</td>\n",
       "      <td>324</td>\n",
       "      <td>101927844</td>\n",
       "      <td>54325</td>\n",
       "      <td>Unique</td>\n",
       "      <td>1.0/1</td>\n",
       "      <td>False</td>\n",
       "      <td>[(87716184, 87716508, '1.0/1')]</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>['NM_001290053.1']</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>238140 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Gene Chrom     Start       End             ranges  Length  \\\n",
       "110056       A1BG    19  58858388  58858395  58858388-58858395       7   \n",
       "110057       A1BG    19  58858719  58859006  58858719-58859006     287   \n",
       "110058       A1BG    19  58861736  58862017  58861736-58862017     281   \n",
       "110059       A1BG    19  58862757  58863053  58862757-58863053     296   \n",
       "110060       A1BG    19  58863649  58863921  58863649-58863921     272   \n",
       "...           ...   ...       ...       ...                ...     ...   \n",
       "79640       DERPC    16  69152757  69154010  69152757-69154010    1253   \n",
       "79641       DERPC    16  69154010  69154331  69154010-69154331     321   \n",
       "79642       DERPC    16  69154010  69154331  69154010-69154331     321   \n",
       "9453    LINC02801     1  87709298  87709320  87709298-87709320      22   \n",
       "9454    LINC02801     1  87716184  87716508  87716184-87716508     324   \n",
       "\n",
       "           GeneID   HGNC CDS_representation  Ratio  Sharing_status  \\\n",
       "110056          1      5             Unique  1.0/1           False   \n",
       "110057          1      5             Unique  1.0/1           False   \n",
       "110058          1      5             Unique  1.0/1           False   \n",
       "110059          1      5             Unique  1.0/1           False   \n",
       "110060          1      5             Unique  1.0/1           False   \n",
       "...           ...    ...                ...    ...             ...   \n",
       "79640   113455421  54084    Variable_region  4.0/7            True   \n",
       "79641   113455421  54084    Variable_region  7.0/7            True   \n",
       "79642   113455421  54084    Variable_region  7.0/7            True   \n",
       "9453    101927844  54325             Unique  1.0/1           False   \n",
       "9454    101927844  54325             Unique  1.0/1           False   \n",
       "\n",
       "                                                  Sources  \\\n",
       "110056                    [(58858388, 58858395, '1.0/1')]   \n",
       "110057                    [(58858719, 58859006, '1.0/1')]   \n",
       "110058                    [(58861736, 58862017, '1.0/1')]   \n",
       "110059                    [(58862757, 58863053, '1.0/1')]   \n",
       "110060                    [(58863649, 58863921, '1.0/1')]   \n",
       "...                                                   ...   \n",
       "79640                     [(69152757, 69154331, '4.0/7')]   \n",
       "79641   [(69152757, 69154331, '4.0/7'), (69154010, 691...   \n",
       "79642   [(69152757, 69154331, '4.0/7'), (69154010, 691...   \n",
       "9453                      [(87709298, 87709320, '1.0/1')]   \n",
       "9454                      [(87716184, 87716508, '1.0/1')]   \n",
       "\n",
       "        Count_CDS_alternative  Count_CDS_constitutive  \\\n",
       "110056                      0                       8   \n",
       "110057                      0                       8   \n",
       "110058                      0                       8   \n",
       "110059                      0                       8   \n",
       "110060                      0                       8   \n",
       "...                       ...                     ...   \n",
       "79640                       2                       1   \n",
       "79641                       2                       1   \n",
       "79642                       2                       1   \n",
       "9453                        0                       2   \n",
       "9454                        0                       2   \n",
       "\n",
       "                                                 mRNA_IDS  mRNA_nb  \n",
       "110056                                    ['NM_130786.4']        1  \n",
       "110057                                    ['NM_130786.4']        1  \n",
       "110058                                    ['NM_130786.4']        1  \n",
       "110059                                    ['NM_130786.4']        1  \n",
       "110060                                    ['NM_130786.4']        1  \n",
       "...                                                   ...      ...  \n",
       "79640   ['NM_001366605.1', 'NM_001366603.1', 'NM_00104...        7  \n",
       "79641   ['NM_001366605.1', 'NM_001366603.1', 'NM_00104...        7  \n",
       "79642   ['NM_001366605.1', 'NM_001366603.1', 'NM_00104...        7  \n",
       "9453                                   ['NM_001290053.1']        1  \n",
       "9454                                   ['NM_001290053.1']        1  \n",
       "\n",
       "[238140 rows x 16 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "refseq = pd.read_csv('/home/weber/PycharmProjects/ExoCarto/data/2_processed/DF_transformed.csv.gz', compression='gzip', sep='\\t').dropna(subset=['HGNC']).sort_values(by=['HGNC', 'Chrom', 'Start'], ascending=True)\n",
    "refseq['HGNC'] = refseq['HGNC'].astype(int)\n",
    "refseq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Chrom</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>ranges</th>\n",
       "      <th>Length</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>HGNC</th>\n",
       "      <th>CDS_representation</th>\n",
       "      <th>Ratio</th>\n",
       "      <th>Sharing_status</th>\n",
       "      <th>Sources</th>\n",
       "      <th>Count_CDS_alternative</th>\n",
       "      <th>Count_CDS_constitutive</th>\n",
       "      <th>mRNA_IDS</th>\n",
       "      <th>mRNA_nb</th>\n",
       "      <th>Ratio_num</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>231896</th>\n",
       "      <td>DMD</td>\n",
       "      <td>X</td>\n",
       "      <td>31139950</td>\n",
       "      <td>31140036</td>\n",
       "      <td>31139950-31140036</td>\n",
       "      <td>86</td>\n",
       "      <td>1756</td>\n",
       "      <td>2928</td>\n",
       "      <td>Variable_region</td>\n",
       "      <td>5.0/17</td>\n",
       "      <td>True</td>\n",
       "      <td>[(31139950, 31140047, '5.0/17')]</td>\n",
       "      <td>81</td>\n",
       "      <td>8</td>\n",
       "      <td>['NM_000109.4', 'NM_004011.4', 'NM_004012.4', ...</td>\n",
       "      <td>17</td>\n",
       "      <td>0.294118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231897</th>\n",
       "      <td>DMD</td>\n",
       "      <td>X</td>\n",
       "      <td>31140036</td>\n",
       "      <td>31140047</td>\n",
       "      <td>31140036-31140047</td>\n",
       "      <td>11</td>\n",
       "      <td>1756</td>\n",
       "      <td>2928</td>\n",
       "      <td>Variable_region</td>\n",
       "      <td>16.0/17</td>\n",
       "      <td>True</td>\n",
       "      <td>[(31139950, 31140047, '5.0/17'), (31140036, 31...</td>\n",
       "      <td>81</td>\n",
       "      <td>8</td>\n",
       "      <td>['NM_000109.4', 'NM_004011.4', 'NM_004012.4', ...</td>\n",
       "      <td>17</td>\n",
       "      <td>0.941176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231898</th>\n",
       "      <td>DMD</td>\n",
       "      <td>X</td>\n",
       "      <td>31144759</td>\n",
       "      <td>31144790</td>\n",
       "      <td>31144759-31144790</td>\n",
       "      <td>31</td>\n",
       "      <td>1756</td>\n",
       "      <td>2928</td>\n",
       "      <td>Unique</td>\n",
       "      <td>11.0/17</td>\n",
       "      <td>False</td>\n",
       "      <td>[(31144759, 31144790, '11.0/17')]</td>\n",
       "      <td>81</td>\n",
       "      <td>8</td>\n",
       "      <td>['NM_000109.4', 'NM_004011.4', 'NM_004012.4', ...</td>\n",
       "      <td>17</td>\n",
       "      <td>0.647059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231899</th>\n",
       "      <td>DMD</td>\n",
       "      <td>X</td>\n",
       "      <td>31152219</td>\n",
       "      <td>31152311</td>\n",
       "      <td>31152219-31152311</td>\n",
       "      <td>92</td>\n",
       "      <td>1756</td>\n",
       "      <td>2928</td>\n",
       "      <td>Unique</td>\n",
       "      <td>16.0/17</td>\n",
       "      <td>False</td>\n",
       "      <td>[(31152219, 31152311, '16.0/17')]</td>\n",
       "      <td>81</td>\n",
       "      <td>8</td>\n",
       "      <td>['NM_000109.4', 'NM_004011.4', 'NM_004012.4', ...</td>\n",
       "      <td>17</td>\n",
       "      <td>0.941176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231900</th>\n",
       "      <td>DMD</td>\n",
       "      <td>X</td>\n",
       "      <td>31164408</td>\n",
       "      <td>31164531</td>\n",
       "      <td>31164408-31164531</td>\n",
       "      <td>123</td>\n",
       "      <td>1756</td>\n",
       "      <td>2928</td>\n",
       "      <td>Unique</td>\n",
       "      <td>16.0/17</td>\n",
       "      <td>False</td>\n",
       "      <td>[(31164408, 31164531, '16.0/17')]</td>\n",
       "      <td>81</td>\n",
       "      <td>8</td>\n",
       "      <td>['NM_000109.4', 'NM_004011.4', 'NM_004012.4', ...</td>\n",
       "      <td>17</td>\n",
       "      <td>0.941176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231980</th>\n",
       "      <td>DMD</td>\n",
       "      <td>X</td>\n",
       "      <td>32867845</td>\n",
       "      <td>32867937</td>\n",
       "      <td>32867845-32867937</td>\n",
       "      <td>92</td>\n",
       "      <td>1756</td>\n",
       "      <td>2928</td>\n",
       "      <td>Unique</td>\n",
       "      <td>3.0/17</td>\n",
       "      <td>False</td>\n",
       "      <td>[(32867845, 32867937, '3.0/17')]</td>\n",
       "      <td>81</td>\n",
       "      <td>8</td>\n",
       "      <td>['NM_000109.4', 'NM_004011.4', 'NM_004012.4', ...</td>\n",
       "      <td>17</td>\n",
       "      <td>0.176471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231981</th>\n",
       "      <td>DMD</td>\n",
       "      <td>X</td>\n",
       "      <td>33038256</td>\n",
       "      <td>33038317</td>\n",
       "      <td>33038256-33038317</td>\n",
       "      <td>61</td>\n",
       "      <td>1756</td>\n",
       "      <td>2928</td>\n",
       "      <td>Unique</td>\n",
       "      <td>3.0/17</td>\n",
       "      <td>False</td>\n",
       "      <td>[(33038256, 33038317, '3.0/17')]</td>\n",
       "      <td>81</td>\n",
       "      <td>8</td>\n",
       "      <td>['NM_000109.4', 'NM_004011.4', 'NM_004012.4', ...</td>\n",
       "      <td>17</td>\n",
       "      <td>0.176471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231982</th>\n",
       "      <td>DMD</td>\n",
       "      <td>X</td>\n",
       "      <td>33146264</td>\n",
       "      <td>33146282</td>\n",
       "      <td>33146264-33146282</td>\n",
       "      <td>18</td>\n",
       "      <td>1756</td>\n",
       "      <td>2928</td>\n",
       "      <td>Unique</td>\n",
       "      <td>1.0/17</td>\n",
       "      <td>False</td>\n",
       "      <td>[(33146264, 33146282, '1.0/17')]</td>\n",
       "      <td>81</td>\n",
       "      <td>8</td>\n",
       "      <td>['NM_000109.4', 'NM_004011.4', 'NM_004012.4', ...</td>\n",
       "      <td>17</td>\n",
       "      <td>0.058824</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231983</th>\n",
       "      <td>DMD</td>\n",
       "      <td>X</td>\n",
       "      <td>33229399</td>\n",
       "      <td>33229429</td>\n",
       "      <td>33229399-33229429</td>\n",
       "      <td>30</td>\n",
       "      <td>1756</td>\n",
       "      <td>2928</td>\n",
       "      <td>Unique</td>\n",
       "      <td>1.0/17</td>\n",
       "      <td>False</td>\n",
       "      <td>[(33229399, 33229429, '1.0/17')]</td>\n",
       "      <td>81</td>\n",
       "      <td>8</td>\n",
       "      <td>['NM_000109.4', 'NM_004011.4', 'NM_004012.4', ...</td>\n",
       "      <td>17</td>\n",
       "      <td>0.058824</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231984</th>\n",
       "      <td>DMD</td>\n",
       "      <td>X</td>\n",
       "      <td>33357376</td>\n",
       "      <td>33357382</td>\n",
       "      <td>33357376-33357382</td>\n",
       "      <td>6</td>\n",
       "      <td>1756</td>\n",
       "      <td>2928</td>\n",
       "      <td>Unique</td>\n",
       "      <td>1.0/17</td>\n",
       "      <td>False</td>\n",
       "      <td>[(33357376, 33357382, '1.0/17')]</td>\n",
       "      <td>81</td>\n",
       "      <td>8</td>\n",
       "      <td>['NM_000109.4', 'NM_004011.4', 'NM_004012.4', ...</td>\n",
       "      <td>17</td>\n",
       "      <td>0.058824</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>89 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Gene Chrom     Start       End             ranges  Length  GeneID  \\\n",
       "231896  DMD     X  31139950  31140036  31139950-31140036      86    1756   \n",
       "231897  DMD     X  31140036  31140047  31140036-31140047      11    1756   \n",
       "231898  DMD     X  31144759  31144790  31144759-31144790      31    1756   \n",
       "231899  DMD     X  31152219  31152311  31152219-31152311      92    1756   \n",
       "231900  DMD     X  31164408  31164531  31164408-31164531     123    1756   \n",
       "...     ...   ...       ...       ...                ...     ...     ...   \n",
       "231980  DMD     X  32867845  32867937  32867845-32867937      92    1756   \n",
       "231981  DMD     X  33038256  33038317  33038256-33038317      61    1756   \n",
       "231982  DMD     X  33146264  33146282  33146264-33146282      18    1756   \n",
       "231983  DMD     X  33229399  33229429  33229399-33229429      30    1756   \n",
       "231984  DMD     X  33357376  33357382  33357376-33357382       6    1756   \n",
       "\n",
       "        HGNC CDS_representation    Ratio  Sharing_status  \\\n",
       "231896  2928    Variable_region   5.0/17            True   \n",
       "231897  2928    Variable_region  16.0/17            True   \n",
       "231898  2928             Unique  11.0/17           False   \n",
       "231899  2928             Unique  16.0/17           False   \n",
       "231900  2928             Unique  16.0/17           False   \n",
       "...      ...                ...      ...             ...   \n",
       "231980  2928             Unique   3.0/17           False   \n",
       "231981  2928             Unique   3.0/17           False   \n",
       "231982  2928             Unique   1.0/17           False   \n",
       "231983  2928             Unique   1.0/17           False   \n",
       "231984  2928             Unique   1.0/17           False   \n",
       "\n",
       "                                                  Sources  \\\n",
       "231896                   [(31139950, 31140047, '5.0/17')]   \n",
       "231897  [(31139950, 31140047, '5.0/17'), (31140036, 31...   \n",
       "231898                  [(31144759, 31144790, '11.0/17')]   \n",
       "231899                  [(31152219, 31152311, '16.0/17')]   \n",
       "231900                  [(31164408, 31164531, '16.0/17')]   \n",
       "...                                                   ...   \n",
       "231980                   [(32867845, 32867937, '3.0/17')]   \n",
       "231981                   [(33038256, 33038317, '3.0/17')]   \n",
       "231982                   [(33146264, 33146282, '1.0/17')]   \n",
       "231983                   [(33229399, 33229429, '1.0/17')]   \n",
       "231984                   [(33357376, 33357382, '1.0/17')]   \n",
       "\n",
       "        Count_CDS_alternative  Count_CDS_constitutive  \\\n",
       "231896                     81                       8   \n",
       "231897                     81                       8   \n",
       "231898                     81                       8   \n",
       "231899                     81                       8   \n",
       "231900                     81                       8   \n",
       "...                       ...                     ...   \n",
       "231980                     81                       8   \n",
       "231981                     81                       8   \n",
       "231982                     81                       8   \n",
       "231983                     81                       8   \n",
       "231984                     81                       8   \n",
       "\n",
       "                                                 mRNA_IDS  mRNA_nb  Ratio_num  \n",
       "231896  ['NM_000109.4', 'NM_004011.4', 'NM_004012.4', ...       17   0.294118  \n",
       "231897  ['NM_000109.4', 'NM_004011.4', 'NM_004012.4', ...       17   0.941176  \n",
       "231898  ['NM_000109.4', 'NM_004011.4', 'NM_004012.4', ...       17   0.647059  \n",
       "231899  ['NM_000109.4', 'NM_004011.4', 'NM_004012.4', ...       17   0.941176  \n",
       "231900  ['NM_000109.4', 'NM_004011.4', 'NM_004012.4', ...       17   0.941176  \n",
       "...                                                   ...      ...        ...  \n",
       "231980  ['NM_000109.4', 'NM_004011.4', 'NM_004012.4', ...       17   0.176471  \n",
       "231981  ['NM_000109.4', 'NM_004011.4', 'NM_004012.4', ...       17   0.176471  \n",
       "231982  ['NM_000109.4', 'NM_004011.4', 'NM_004012.4', ...       17   0.058824  \n",
       "231983  ['NM_000109.4', 'NM_004011.4', 'NM_004012.4', ...       17   0.058824  \n",
       "231984  ['NM_000109.4', 'NM_004011.4', 'NM_004012.4', ...       17   0.058824  \n",
       "\n",
       "[89 rows x 17 columns]"
      ]
     },
     "execution_count": 251,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "refseq.loc[refseq['Gene'] == 'DMD']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([    7,    18,    19, ..., 53651, 53757, 54084])"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "multi_mrna_genes = refseq.loc[refseq['mRNA_nb'] > 1, 'HGNC'].unique()\n",
    "multi_mrna_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "19200it [05:34, 57.41it/s]                            \n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "\n",
    "def map_refseq_pext(gene, l):\n",
    "    tmp_l = list()\n",
    "    refseq_gene = refseq.loc[refseq['HGNC'] == gene].reset_index(drop=True)\n",
    "    pext_gene = pext_modified.loc[pext_modified['HGNC ID'] == gene].reset_index(drop=True)\n",
    "    for r in refseq_gene[['Start', 'End']].values:\n",
    "        start = r[0]\n",
    "        end = r[1]\n",
    "        pext_exon_values = pext_gene.loc[pext_gene['pos'].between(start, end)].groupby(['HGNC ID', 'symbol', 'ensg']).median().drop(['pos'], axis=1)\n",
    "        pext_exon_values['Exon'] = str(start) + '-' + str(end)\n",
    "        tmp_l.append(pext_exon_values)\n",
    "    if tmp_l:\n",
    "        concat_df_gene = pd.concat(tmp_l).reset_index()\n",
    "        concat_df_gene = concat_df_gene[list(concat_df_gene.columns[:3]) + ['Exon'] + sorted(list(concat_df_gene.columns[3:-2])) + ['mean_proportion']]\n",
    "        l.append(concat_df_gene)\n",
    "    \n",
    "    \n",
    "m = multiprocessing.Manager()\n",
    "l = m.list()\n",
    "parmap.starmap(map_refseq_pext, list(zip(sorted(refseq.HGNC.unique()))), l, pm_pbar=True, pm_processes=20)\n",
    "    \n",
    "output_df = pd.concat(list(l))\n",
    "output_df.to_parquet('/home/weber/PycharmProjects/ExoCarto/data/clean/2_expression/RefSeq_PEXT.parquet', index=False)\n",
    "output_df\n",
    "\n",
    "\n",
    "\n",
    "    # transposed = test.T\n",
    "    # percentile_df = list()\n",
    "    # for col in transposed:\n",
    "    #     percentile_df.append(transposed[col].rank(pct=True))\n",
    "    # pd.concat(percentile_df, axis=1).sort_index().T\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>start_coordinate</th>\n",
       "      <th>end_coordinate</th>\n",
       "      <th>tag</th>\n",
       "      <th>HGNC ID</th>\n",
       "      <th>HGNC symbol</th>\n",
       "      <th>Range</th>\n",
       "      <th>max_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>1</td>\n",
       "      <td>861321</td>\n",
       "      <td>861392</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>28706</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>861321-861392</td>\n",
       "      <td>334.9858</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>1</td>\n",
       "      <td>865534</td>\n",
       "      <td>865715</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>28706</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>865534-865715</td>\n",
       "      <td>576.4071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>1</td>\n",
       "      <td>866418</td>\n",
       "      <td>866468</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>28706</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>866418-866468</td>\n",
       "      <td>146.7116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>1</td>\n",
       "      <td>871151</td>\n",
       "      <td>871275</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>28706</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>871151-871275</td>\n",
       "      <td>473.3769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>1</td>\n",
       "      <td>874419</td>\n",
       "      <td>874508</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>28706</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>874419-874508</td>\n",
       "      <td>112.1655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167795</th>\n",
       "      <td>ENSG00000205916</td>\n",
       "      <td>Y</td>\n",
       "      <td>27042651</td>\n",
       "      <td>27042685</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>15966</td>\n",
       "      <td>DAZ4</td>\n",
       "      <td>27042651-27042685</td>\n",
       "      <td>55.4482</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167796</th>\n",
       "      <td>ENSG00000185894</td>\n",
       "      <td>Y</td>\n",
       "      <td>27184955</td>\n",
       "      <td>27185060</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>18225</td>\n",
       "      <td>BPY2C</td>\n",
       "      <td>27184955-27185060</td>\n",
       "      <td>5.1943</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167797</th>\n",
       "      <td>ENSG00000185894</td>\n",
       "      <td>Y</td>\n",
       "      <td>27187915</td>\n",
       "      <td>27188032</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>18225</td>\n",
       "      <td>BPY2C</td>\n",
       "      <td>27187915-27188032</td>\n",
       "      <td>-14.0672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167798</th>\n",
       "      <td>ENSG00000185894</td>\n",
       "      <td>Y</td>\n",
       "      <td>27190092</td>\n",
       "      <td>27190169</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>18225</td>\n",
       "      <td>BPY2C</td>\n",
       "      <td>27190092-27190169</td>\n",
       "      <td>11.2080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167799</th>\n",
       "      <td>ENSG00000172288</td>\n",
       "      <td>Y</td>\n",
       "      <td>27768589</td>\n",
       "      <td>27770180</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1809</td>\n",
       "      <td>CDY1</td>\n",
       "      <td>27768589-27770180</td>\n",
       "      <td>81.3163</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>167800 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                gene_id chromosome_name  start_coordinate  end_coordinate  \\\n",
       "0       ENSG00000187634               1            861321          861392   \n",
       "1       ENSG00000187634               1            865534          865715   \n",
       "2       ENSG00000187634               1            866418          866468   \n",
       "3       ENSG00000187634               1            871151          871275   \n",
       "4       ENSG00000187634               1            874419          874508   \n",
       "...                 ...             ...               ...             ...   \n",
       "167795  ENSG00000205916               Y          27042651        27042685   \n",
       "167796  ENSG00000185894               Y          27184955        27185060   \n",
       "167797  ENSG00000185894               Y          27187915        27188032   \n",
       "167798  ENSG00000185894               Y          27190092        27190169   \n",
       "167799  ENSG00000172288               Y          27768589        27770180   \n",
       "\n",
       "         tag HGNC ID HGNC symbol              Range  max_score  \n",
       "0       CCDS   28706      SAMD11      861321-861392   334.9858  \n",
       "1       CCDS   28706      SAMD11      865534-865715   576.4071  \n",
       "2       CCDS   28706      SAMD11      866418-866468   146.7116  \n",
       "3       CCDS   28706      SAMD11      871151-871275   473.3769  \n",
       "4       CCDS   28706      SAMD11      874419-874508   112.1655  \n",
       "...      ...     ...         ...                ...        ...  \n",
       "167795  CCDS   15966        DAZ4  27042651-27042685    55.4482  \n",
       "167796  CCDS   18225       BPY2C  27184955-27185060     5.1943  \n",
       "167797  CCDS   18225       BPY2C  27187915-27188032   -14.0672  \n",
       "167798  CCDS   18225       BPY2C  27190092-27190169    11.2080  \n",
       "167799  CCDS    1809        CDY1  27768589-27770180    81.3163  \n",
       "\n",
       "[167800 rows x 9 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phylocsf_file = \"/gstock/biolo_datasets/variation/gnomAD/latest/gtex/phylocsf_data.tsv.bgz\"\n",
    "phylocsf_df = pd.read_csv(phylocsf_file, compression='gzip', sep='\\t')\n",
    "phylocsf_df = phylocsf_df[['gene_id', 'chromosome_name', 'start_coordinate', 'end_coordinate', 'tag', 'max_score']]\n",
    "phylocsf_df = phylocsf_df.loc[phylocsf_df['tag'] == 'CCDS']\n",
    "phylocsf_df['chromosome_name'] = phylocsf_df['chromosome_name'].str.replace('chr', '')\n",
    "phylocsf_df = phylocsf_df.sort_values(by=['chromosome_name', 'start_coordinate'], ascending=[True, True]).drop_duplicates()\n",
    "phylocsf_df['range'] = phylocsf_df['start_coordinate'].astype(str) + '-' + phylocsf_df['end_coordinate'].astype(str)\n",
    "phylocsf_df['gene_id'] = phylocsf_df['gene_id'].str.split('.').apply(lambda r: r[0])\n",
    "phylocsf_df = pd.merge(phylocsf_df, biomart_ensembl, right_on='Gene stable ID', left_on='gene_id', how='left')\n",
    "phylocsf_df = phylocsf_df[list(phylocsf_df.columns)[:5] + ['HGNC ID', 'HGNC symbol', 'range', 'max_score']]\n",
    "phylocsf_df.rename({'range': 'Range'}, axis=1, inplace=True)\n",
    "# phylocsf_df['HGNC ID'] = phylocsf_df['HGNC ID'].astype(int)\n",
    "phylocsf_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>symbol</th>\n",
       "      <th>hi_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100652477</td>\n",
       "      <td>100715409</td>\n",
       "      <td>DBT</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>103342022</td>\n",
       "      <td>103574052</td>\n",
       "      <td>COL11A1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10535002</td>\n",
       "      <td>10690815</td>\n",
       "      <td>PEX14</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10696660</td>\n",
       "      <td>10856733</td>\n",
       "      <td>CASZ1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>109419602</td>\n",
       "      <td>109476957</td>\n",
       "      <td>GPSM2</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1397</th>\n",
       "      <td>chrY</td>\n",
       "      <td>25130409</td>\n",
       "      <td>25151612</td>\n",
       "      <td>BPY2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1398</th>\n",
       "      <td>chrY</td>\n",
       "      <td>25275501</td>\n",
       "      <td>25345239</td>\n",
       "      <td>DAZ1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1399</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26356113</td>\n",
       "      <td>26360978</td>\n",
       "      <td>GOLGA2P2Y</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1400</th>\n",
       "      <td>chrY</td>\n",
       "      <td>2654895</td>\n",
       "      <td>2655782</td>\n",
       "      <td>SRY</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1401</th>\n",
       "      <td>chrY</td>\n",
       "      <td>27768263</td>\n",
       "      <td>27771049</td>\n",
       "      <td>CDY1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1402 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom      start        end     symbol  hi_level\n",
       "0     chr1  100652477  100715409        DBT        30\n",
       "1     chr1  103342022  103574052    COL11A1         1\n",
       "2     chr1   10535002   10690815      PEX14        30\n",
       "3     chr1   10696660   10856733      CASZ1         1\n",
       "4     chr1  109419602  109476957      GPSM2        30\n",
       "...    ...        ...        ...        ...       ...\n",
       "1397  chrY   25130409   25151612       BPY2         0\n",
       "1398  chrY   25275501   25345239       DAZ1         0\n",
       "1399  chrY   26356113   26360978  GOLGA2P2Y         0\n",
       "1400  chrY    2654895    2655782        SRY         3\n",
       "1401  chrY   27768263   27771049       CDY1         0\n",
       "\n",
       "[1402 rows x 5 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hi_clingen_file = \"/gstock/biolo_datasets/variation/benchmark/Databases/ClinGen/ClinGen_haploinsufficiency_gene_GRCh37.bed\"\n",
    "hi_clingen = pd.read_csv(hi_clingen_file, sep='\\t', skiprows=1, names=['chrom', 'start', 'end', 'symbol', 'hi_level'])\n",
    "hi_clingen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phylocsf_df.loc[phylocsf_df['HGNC symbol'] == 'CACNB2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CCRS_Gene</th>\n",
       "      <th>RefSeq_CDS_representation</th>\n",
       "      <th>RefSeq_Chrom</th>\n",
       "      <th>RefSeq_Count_CDS_alternative</th>\n",
       "      <th>RefSeq_Count_CDS_constitutive</th>\n",
       "      <th>RefSeq_End</th>\n",
       "      <th>RefSeq_GeneID</th>\n",
       "      <th>RefSeq_HGNC</th>\n",
       "      <th>RefSeq_Length</th>\n",
       "      <th>RefSeq_Ratio</th>\n",
       "      <th>...</th>\n",
       "      <th>RefSeq_mRNA_IDS</th>\n",
       "      <th>RefSeq_mRNA_nb</th>\n",
       "      <th>RefSeq_ranges</th>\n",
       "      <th>MC</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>Source</th>\n",
       "      <th>VAR_ID</th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>HGNC ID</th>\n",
       "      <th>HGNC symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A2M</td>\n",
       "      <td>Unique</td>\n",
       "      <td>12</td>\n",
       "      <td>5</td>\n",
       "      <td>33</td>\n",
       "      <td>9266139</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>183</td>\n",
       "      <td>0.5</td>\n",
       "      <td>...</td>\n",
       "      <td>['NM_001347423.2', 'NM_001347425.1', 'NM_00001...</td>\n",
       "      <td>4</td>\n",
       "      <td>9265956-9266139</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>7</td>\n",
       "      <td>A2M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A2M</td>\n",
       "      <td>Unique</td>\n",
       "      <td>12</td>\n",
       "      <td>5</td>\n",
       "      <td>33</td>\n",
       "      <td>9243078</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>126</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>['NM_001347423.2', 'NM_001347425.1', 'NM_00001...</td>\n",
       "      <td>4</td>\n",
       "      <td>9242952-9243078</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>7</td>\n",
       "      <td>A2M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A2M</td>\n",
       "      <td>Unique</td>\n",
       "      <td>12</td>\n",
       "      <td>5</td>\n",
       "      <td>33</td>\n",
       "      <td>9244025</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>228</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>['NM_001347423.2', 'NM_001347425.1', 'NM_00001...</td>\n",
       "      <td>4</td>\n",
       "      <td>9243797-9244025</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>7</td>\n",
       "      <td>A2M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A2M</td>\n",
       "      <td>Unique</td>\n",
       "      <td>12</td>\n",
       "      <td>5</td>\n",
       "      <td>33</td>\n",
       "      <td>9246175</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>114</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>['NM_001347423.2', 'NM_001347425.1', 'NM_00001...</td>\n",
       "      <td>4</td>\n",
       "      <td>9246061-9246175</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>7</td>\n",
       "      <td>A2M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A2M</td>\n",
       "      <td>Unique</td>\n",
       "      <td>12</td>\n",
       "      <td>5</td>\n",
       "      <td>33</td>\n",
       "      <td>9247680</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>111</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>['NM_001347423.2', 'NM_001347425.1', 'NM_00001...</td>\n",
       "      <td>4</td>\n",
       "      <td>9247569-9247680</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>7</td>\n",
       "      <td>A2M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137431</th>\n",
       "      <td>ZPBP</td>\n",
       "      <td>Unique</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>50070906</td>\n",
       "      <td>11055</td>\n",
       "      <td>15662</td>\n",
       "      <td>218</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>['NM_001159878.2', 'NM_007009.3']</td>\n",
       "      <td>2</td>\n",
       "      <td>50070688-50070906</td>\n",
       "      <td>missense</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>HGMD</td>\n",
       "      <td>50070744_T_A</td>\n",
       "      <td>ENSG00000042813</td>\n",
       "      <td>15662</td>\n",
       "      <td>ZPBP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137432</th>\n",
       "      <td>ZPBP</td>\n",
       "      <td>Unique</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>50121495</td>\n",
       "      <td>11055</td>\n",
       "      <td>15662</td>\n",
       "      <td>125</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>['NM_001159878.2', 'NM_007009.3']</td>\n",
       "      <td>2</td>\n",
       "      <td>50121370-50121495</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000042813</td>\n",
       "      <td>15662</td>\n",
       "      <td>ZPBP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137433</th>\n",
       "      <td>ZPBP</td>\n",
       "      <td>Variable_region</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>50097737</td>\n",
       "      <td>11055</td>\n",
       "      <td>15662</td>\n",
       "      <td>3</td>\n",
       "      <td>0.5</td>\n",
       "      <td>...</td>\n",
       "      <td>['NM_001159878.2', 'NM_007009.3']</td>\n",
       "      <td>2</td>\n",
       "      <td>50097734-50097737</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000042813</td>\n",
       "      <td>15662</td>\n",
       "      <td>ZPBP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137434</th>\n",
       "      <td>ZPBP</td>\n",
       "      <td>Unique</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>50070906</td>\n",
       "      <td>11055</td>\n",
       "      <td>15662</td>\n",
       "      <td>218</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>['NM_001159878.2', 'NM_007009.3']</td>\n",
       "      <td>2</td>\n",
       "      <td>50070688-50070906</td>\n",
       "      <td>missense</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>HGMD</td>\n",
       "      <td>50070844_T_C</td>\n",
       "      <td>ENSG00000042813</td>\n",
       "      <td>15662</td>\n",
       "      <td>ZPBP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137435</th>\n",
       "      <td>ZPBP</td>\n",
       "      <td>Unique</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>50132790</td>\n",
       "      <td>11055</td>\n",
       "      <td>15662</td>\n",
       "      <td>126</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>['NM_001159878.2', 'NM_007009.3']</td>\n",
       "      <td>2</td>\n",
       "      <td>50132664-50132790</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000042813</td>\n",
       "      <td>15662</td>\n",
       "      <td>ZPBP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>137436 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       CCRS_Gene RefSeq_CDS_representation RefSeq_Chrom  \\\n",
       "0            A2M                    Unique           12   \n",
       "1            A2M                    Unique           12   \n",
       "2            A2M                    Unique           12   \n",
       "3            A2M                    Unique           12   \n",
       "4            A2M                    Unique           12   \n",
       "...          ...                       ...          ...   \n",
       "137431      ZPBP                    Unique            7   \n",
       "137432      ZPBP                    Unique            7   \n",
       "137433      ZPBP           Variable_region            7   \n",
       "137434      ZPBP                    Unique            7   \n",
       "137435      ZPBP                    Unique            7   \n",
       "\n",
       "        RefSeq_Count_CDS_alternative  RefSeq_Count_CDS_constitutive  \\\n",
       "0                                  5                             33   \n",
       "1                                  5                             33   \n",
       "2                                  5                             33   \n",
       "3                                  5                             33   \n",
       "4                                  5                             33   \n",
       "...                              ...                            ...   \n",
       "137431                             1                              8   \n",
       "137432                             1                              8   \n",
       "137433                             1                              8   \n",
       "137434                             1                              8   \n",
       "137435                             1                              8   \n",
       "\n",
       "        RefSeq_End  RefSeq_GeneID  RefSeq_HGNC  RefSeq_Length  RefSeq_Ratio  \\\n",
       "0          9266139              2            7            183           0.5   \n",
       "1          9243078              2            7            126           1.0   \n",
       "2          9244025              2            7            228           1.0   \n",
       "3          9246175              2            7            114           1.0   \n",
       "4          9247680              2            7            111           1.0   \n",
       "...            ...            ...          ...            ...           ...   \n",
       "137431    50070906          11055        15662            218           1.0   \n",
       "137432    50121495          11055        15662            125           1.0   \n",
       "137433    50097737          11055        15662              3           0.5   \n",
       "137434    50070906          11055        15662            218           1.0   \n",
       "137435    50132790          11055        15662            126           1.0   \n",
       "\n",
       "        ...                                    RefSeq_mRNA_IDS RefSeq_mRNA_nb  \\\n",
       "0       ...  ['NM_001347423.2', 'NM_001347425.1', 'NM_00001...              4   \n",
       "1       ...  ['NM_001347423.2', 'NM_001347425.1', 'NM_00001...              4   \n",
       "2       ...  ['NM_001347423.2', 'NM_001347425.1', 'NM_00001...              4   \n",
       "3       ...  ['NM_001347423.2', 'NM_001347425.1', 'NM_00001...              4   \n",
       "4       ...  ['NM_001347423.2', 'NM_001347425.1', 'NM_00001...              4   \n",
       "...     ...                                                ...            ...   \n",
       "137431  ...                  ['NM_001159878.2', 'NM_007009.3']              2   \n",
       "137432  ...                  ['NM_001159878.2', 'NM_007009.3']              2   \n",
       "137433  ...                  ['NM_001159878.2', 'NM_007009.3']              2   \n",
       "137434  ...                  ['NM_001159878.2', 'NM_007009.3']              2   \n",
       "137435  ...                  ['NM_001159878.2', 'NM_007009.3']              2   \n",
       "\n",
       "            RefSeq_ranges        MC      STATUS Source        VAR_ID  \\\n",
       "0         9265956-9266139      None        None   None          None   \n",
       "1         9242952-9243078      None        None   None          None   \n",
       "2         9243797-9244025      None        None   None          None   \n",
       "3         9246061-9246175      None        None   None          None   \n",
       "4         9247569-9247680      None        None   None          None   \n",
       "...                   ...       ...         ...    ...           ...   \n",
       "137431  50070688-50070906  missense  Pathogenic   HGMD  50070744_T_A   \n",
       "137432  50121370-50121495      None        None   None          None   \n",
       "137433  50097734-50097737      None        None   None          None   \n",
       "137434  50070688-50070906  missense  Pathogenic   HGMD  50070844_T_C   \n",
       "137435  50132664-50132790      None        None   None          None   \n",
       "\n",
       "         Gene stable ID HGNC ID HGNC symbol  \n",
       "0       ENSG00000175899       7         A2M  \n",
       "1       ENSG00000175899       7         A2M  \n",
       "2       ENSG00000175899       7         A2M  \n",
       "3       ENSG00000175899       7         A2M  \n",
       "4       ENSG00000175899       7         A2M  \n",
       "...                 ...     ...         ...  \n",
       "137431  ENSG00000042813   15662        ZPBP  \n",
       "137432  ENSG00000042813   15662        ZPBP  \n",
       "137433  ENSG00000042813   15662        ZPBP  \n",
       "137434  ENSG00000042813   15662        ZPBP  \n",
       "137435  ENSG00000042813   15662        ZPBP  \n",
       "\n",
       "[137436 rows x 23 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vcf_df = pd.read_parquet('/home/weber/PycharmProjects/ExoCarto/data/2_processed/RefSeq_X_VCF_NO_CCRS_pathogenic_only.parquet')\n",
    "biomart_ensembl['HGNC ID'] = biomart_ensembl['HGNC ID'].astype(int)\n",
    "vcf_df = pd.merge(vcf_df, biomart_ensembl, right_on='HGNC ID', left_on='RefSeq_HGNC', how='left')\n",
    "vcf_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CCRS_Gene</th>\n",
       "      <th>Chrom</th>\n",
       "      <th>RefSeq_Start</th>\n",
       "      <th>RefSeq_End</th>\n",
       "      <th>RefSeq_ranges</th>\n",
       "      <th>RefSeq_Length</th>\n",
       "      <th>RefSeq_GeneID</th>\n",
       "      <th>RefSeq_HGNC</th>\n",
       "      <th>CDS_representation</th>\n",
       "      <th>RefSeq_Ratio</th>\n",
       "      <th>RefSeq_Sharing_status</th>\n",
       "      <th>RefSeq_Sources</th>\n",
       "      <th>RefSeq_Count_CDS_alternative</th>\n",
       "      <th>RefSeq_Count_CDS_constitutive</th>\n",
       "      <th>RefSeq_mRNA_IDS</th>\n",
       "      <th>RefSeq_mRNA_nb</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>1</td>\n",
       "      <td>901912</td>\n",
       "      <td>901994</td>\n",
       "      <td>901912-901994</td>\n",
       "      <td>82</td>\n",
       "      <td>84069</td>\n",
       "      <td>25284.0</td>\n",
       "      <td>Unique</td>\n",
       "      <td>3.0/3</td>\n",
       "      <td>False</td>\n",
       "      <td>[(901912, 901994, '3.0/3')]</td>\n",
       "      <td>2</td>\n",
       "      <td>15</td>\n",
       "      <td>['NM_001367552.1', 'NM_001160184.2', 'NM_03212...</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>1</td>\n",
       "      <td>902084</td>\n",
       "      <td>902183</td>\n",
       "      <td>902084-902183</td>\n",
       "      <td>99</td>\n",
       "      <td>84069</td>\n",
       "      <td>25284.0</td>\n",
       "      <td>Unique</td>\n",
       "      <td>3.0/3</td>\n",
       "      <td>False</td>\n",
       "      <td>[(902084, 902183, '3.0/3')]</td>\n",
       "      <td>2</td>\n",
       "      <td>15</td>\n",
       "      <td>['NM_001367552.1', 'NM_001160184.2', 'NM_03212...</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>1</td>\n",
       "      <td>905657</td>\n",
       "      <td>905803</td>\n",
       "      <td>905657-905803</td>\n",
       "      <td>146</td>\n",
       "      <td>84069</td>\n",
       "      <td>25284.0</td>\n",
       "      <td>Unique</td>\n",
       "      <td>3.0/3</td>\n",
       "      <td>False</td>\n",
       "      <td>[(905657, 905803, '3.0/3')]</td>\n",
       "      <td>2</td>\n",
       "      <td>15</td>\n",
       "      <td>['NM_001367552.1', 'NM_001160184.2', 'NM_03212...</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>1</td>\n",
       "      <td>905901</td>\n",
       "      <td>905981</td>\n",
       "      <td>905901-905981</td>\n",
       "      <td>80</td>\n",
       "      <td>84069</td>\n",
       "      <td>25284.0</td>\n",
       "      <td>Unique</td>\n",
       "      <td>3.0/3</td>\n",
       "      <td>False</td>\n",
       "      <td>[(905901, 905981, '3.0/3')]</td>\n",
       "      <td>2</td>\n",
       "      <td>15</td>\n",
       "      <td>['NM_001367552.1', 'NM_001160184.2', 'NM_03212...</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>1</td>\n",
       "      <td>906066</td>\n",
       "      <td>906138</td>\n",
       "      <td>906066-906138</td>\n",
       "      <td>72</td>\n",
       "      <td>84069</td>\n",
       "      <td>25284.0</td>\n",
       "      <td>Unique</td>\n",
       "      <td>3.0/3</td>\n",
       "      <td>False</td>\n",
       "      <td>[(906066, 906138, '3.0/3')]</td>\n",
       "      <td>2</td>\n",
       "      <td>15</td>\n",
       "      <td>['NM_001367552.1', 'NM_001160184.2', 'NM_03212...</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160740</th>\n",
       "      <td>CDY1</td>\n",
       "      <td>Y</td>\n",
       "      <td>27768590</td>\n",
       "      <td>27770181</td>\n",
       "      <td>27768590-27770181</td>\n",
       "      <td>1591</td>\n",
       "      <td>9085</td>\n",
       "      <td>1809.0</td>\n",
       "      <td>Variable_region</td>\n",
       "      <td>2.0/2</td>\n",
       "      <td>True</td>\n",
       "      <td>[(27768590, 27770181, '1.0/2'), (27768590, 277...</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>['NM_004680.2', 'NM_170723.1']</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160741</th>\n",
       "      <td>CDY1</td>\n",
       "      <td>Y</td>\n",
       "      <td>27770181</td>\n",
       "      <td>27770212</td>\n",
       "      <td>27770181-27770212</td>\n",
       "      <td>31</td>\n",
       "      <td>9085</td>\n",
       "      <td>1809.0</td>\n",
       "      <td>Variable_region</td>\n",
       "      <td>1.0/2</td>\n",
       "      <td>True</td>\n",
       "      <td>[(27768590, 27770212, '1.0/2')]</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>['NM_004680.2', 'NM_170723.1']</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160742</th>\n",
       "      <td>CDY1</td>\n",
       "      <td>Y</td>\n",
       "      <td>27770181</td>\n",
       "      <td>27770212</td>\n",
       "      <td>27770181-27770212</td>\n",
       "      <td>31</td>\n",
       "      <td>9085</td>\n",
       "      <td>1809.0</td>\n",
       "      <td>Variable_region</td>\n",
       "      <td>1.0/2</td>\n",
       "      <td>True</td>\n",
       "      <td>[(27768590, 27770212, '1.0/2')]</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>['NM_004680.2', 'NM_170723.1']</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160743</th>\n",
       "      <td>CDY1</td>\n",
       "      <td>Y</td>\n",
       "      <td>27770602</td>\n",
       "      <td>27770674</td>\n",
       "      <td>27770602-27770674</td>\n",
       "      <td>72</td>\n",
       "      <td>9085</td>\n",
       "      <td>1809.0</td>\n",
       "      <td>Unique</td>\n",
       "      <td>1.0/2</td>\n",
       "      <td>False</td>\n",
       "      <td>[(27770602, 27770674, '1.0/2')]</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>['NM_004680.2', 'NM_170723.1']</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160744</th>\n",
       "      <td>CDY1</td>\n",
       "      <td>Y</td>\n",
       "      <td>27770602</td>\n",
       "      <td>27770674</td>\n",
       "      <td>27770602-27770674</td>\n",
       "      <td>72</td>\n",
       "      <td>9085</td>\n",
       "      <td>1809.0</td>\n",
       "      <td>Unique</td>\n",
       "      <td>1.0/2</td>\n",
       "      <td>False</td>\n",
       "      <td>[(27770602, 27770674, '1.0/2')]</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>['NM_004680.2', 'NM_170723.1']</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>160745 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       CCRS_Gene Chrom  RefSeq_Start  RefSeq_End      RefSeq_ranges  \\\n",
       "0        PLEKHN1     1        901912      901994      901912-901994   \n",
       "1        PLEKHN1     1        902084      902183      902084-902183   \n",
       "2        PLEKHN1     1        905657      905803      905657-905803   \n",
       "3        PLEKHN1     1        905901      905981      905901-905981   \n",
       "4        PLEKHN1     1        906066      906138      906066-906138   \n",
       "...          ...   ...           ...         ...                ...   \n",
       "160740      CDY1     Y      27768590    27770181  27768590-27770181   \n",
       "160741      CDY1     Y      27770181    27770212  27770181-27770212   \n",
       "160742      CDY1     Y      27770181    27770212  27770181-27770212   \n",
       "160743      CDY1     Y      27770602    27770674  27770602-27770674   \n",
       "160744      CDY1     Y      27770602    27770674  27770602-27770674   \n",
       "\n",
       "        RefSeq_Length  RefSeq_GeneID  RefSeq_HGNC CDS_representation  \\\n",
       "0                  82          84069      25284.0             Unique   \n",
       "1                  99          84069      25284.0             Unique   \n",
       "2                 146          84069      25284.0             Unique   \n",
       "3                  80          84069      25284.0             Unique   \n",
       "4                  72          84069      25284.0             Unique   \n",
       "...               ...            ...          ...                ...   \n",
       "160740           1591           9085       1809.0    Variable_region   \n",
       "160741             31           9085       1809.0    Variable_region   \n",
       "160742             31           9085       1809.0    Variable_region   \n",
       "160743             72           9085       1809.0             Unique   \n",
       "160744             72           9085       1809.0             Unique   \n",
       "\n",
       "       RefSeq_Ratio  RefSeq_Sharing_status  \\\n",
       "0             3.0/3                  False   \n",
       "1             3.0/3                  False   \n",
       "2             3.0/3                  False   \n",
       "3             3.0/3                  False   \n",
       "4             3.0/3                  False   \n",
       "...             ...                    ...   \n",
       "160740        2.0/2                   True   \n",
       "160741        1.0/2                   True   \n",
       "160742        1.0/2                   True   \n",
       "160743        1.0/2                  False   \n",
       "160744        1.0/2                  False   \n",
       "\n",
       "                                           RefSeq_Sources  \\\n",
       "0                             [(901912, 901994, '3.0/3')]   \n",
       "1                             [(902084, 902183, '3.0/3')]   \n",
       "2                             [(905657, 905803, '3.0/3')]   \n",
       "3                             [(905901, 905981, '3.0/3')]   \n",
       "4                             [(906066, 906138, '3.0/3')]   \n",
       "...                                                   ...   \n",
       "160740  [(27768590, 27770181, '1.0/2'), (27768590, 277...   \n",
       "160741                    [(27768590, 27770212, '1.0/2')]   \n",
       "160742                    [(27768590, 27770212, '1.0/2')]   \n",
       "160743                    [(27770602, 27770674, '1.0/2')]   \n",
       "160744                    [(27770602, 27770674, '1.0/2')]   \n",
       "\n",
       "        RefSeq_Count_CDS_alternative  RefSeq_Count_CDS_constitutive  \\\n",
       "0                                  2                             15   \n",
       "1                                  2                             15   \n",
       "2                                  2                             15   \n",
       "3                                  2                             15   \n",
       "4                                  2                             15   \n",
       "...                              ...                            ...   \n",
       "160740                             2                              1   \n",
       "160741                             2                              1   \n",
       "160742                             2                              1   \n",
       "160743                             2                              1   \n",
       "160744                             2                              1   \n",
       "\n",
       "                                          RefSeq_mRNA_IDS  RefSeq_mRNA_nb  \n",
       "0       ['NM_001367552.1', 'NM_001160184.2', 'NM_03212...               3  \n",
       "1       ['NM_001367552.1', 'NM_001160184.2', 'NM_03212...               3  \n",
       "2       ['NM_001367552.1', 'NM_001160184.2', 'NM_03212...               3  \n",
       "3       ['NM_001367552.1', 'NM_001160184.2', 'NM_03212...               3  \n",
       "4       ['NM_001367552.1', 'NM_001160184.2', 'NM_03212...               3  \n",
       "...                                                   ...             ...  \n",
       "160740                     ['NM_004680.2', 'NM_170723.1']               2  \n",
       "160741                     ['NM_004680.2', 'NM_170723.1']               2  \n",
       "160742                     ['NM_004680.2', 'NM_170723.1']               2  \n",
       "160743                     ['NM_004680.2', 'NM_170723.1']               2  \n",
       "160744                     ['NM_004680.2', 'NM_170723.1']               2  \n",
       "\n",
       "[160745 rows x 16 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vcf_df = pd.read_csv('/home/weber/PycharmProjects/ExoCarto/' + 'data/2_processed/DF_transformed.csv.gz', compression='gzip', sep='\\t')\n",
    "vcf_df = vcf_df.rename(\n",
    "    {\n",
    "        'Gene' : 'CCRS_Gene',\n",
    "        'Count_CDS_alternative' : 'RefSeq_Count_CDS_alternative',\n",
    "        'Count_CDS_constitutive' : 'RefSeq_Count_CDS_constitutive',\n",
    "        'End' : 'RefSeq_End',\n",
    "        'GeneID' : 'RefSeq_GeneID',\n",
    "        'HGNC' : 'RefSeq_HGNC',\n",
    "        'Length' : 'RefSeq_Length',\n",
    "        'Ratio' : 'RefSeq_Ratio',\n",
    "        'Sharing_status' : 'RefSeq_Sharing_status',\n",
    "        'Sources' : 'RefSeq_Sources',\n",
    "        'Start' : 'RefSeq_Start',\n",
    "        'mRNA_IDS' : 'RefSeq_mRNA_IDS',\n",
    "        'mRNA_nb' : 'RefSeq_mRNA_nb',\n",
    "        'ranges' : 'RefSeq_ranges',\n",
    "    },\n",
    "    axis=1,\n",
    ")\n",
    "vcf_df = vcf_df.loc[(vcf_df['RefSeq_mRNA_nb'] > 1) & (vcf_df['RefSeq_HGNC'].isna() == False)].reset_index(drop=True)\n",
    "vcf_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "919"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test = pd.merge(vcf_df, hi_clingen, left_on='CCRS_Gene', right_on='symbol')\n",
    "len(test.CCRS_Gene.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = vcf_df['RefSeq_HGNC'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_df['RefSeq_HGNC'] = vcf_df['RefSeq_HGNC'].astype(int)\n",
    "\n",
    "pext_modified = pext_modified.dropna(subset=['HGNC ID'])\n",
    "pext_modified['HGNC ID'] = pext_modified['HGNC ID'].astype(int)\n",
    "\n",
    "phylocsf_df = phylocsf_df.dropna(subset=['HGNC ID'])\n",
    "phylocsf_df['HGNC ID'] = phylocsf_df['HGNC ID'].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>start_coordinate</th>\n",
       "      <th>end_coordinate</th>\n",
       "      <th>tag</th>\n",
       "      <th>HGNC ID</th>\n",
       "      <th>HGNC symbol</th>\n",
       "      <th>Range</th>\n",
       "      <th>max_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>1</td>\n",
       "      <td>861321</td>\n",
       "      <td>861392</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>28706</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>861321-861392</td>\n",
       "      <td>334.9858</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>1</td>\n",
       "      <td>865534</td>\n",
       "      <td>865715</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>28706</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>865534-865715</td>\n",
       "      <td>576.4071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>1</td>\n",
       "      <td>866418</td>\n",
       "      <td>866468</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>28706</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>866418-866468</td>\n",
       "      <td>146.7116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>1</td>\n",
       "      <td>871151</td>\n",
       "      <td>871275</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>28706</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>871151-871275</td>\n",
       "      <td>473.3769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>1</td>\n",
       "      <td>874419</td>\n",
       "      <td>874508</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>28706</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>874419-874508</td>\n",
       "      <td>112.1655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167795</th>\n",
       "      <td>ENSG00000205916</td>\n",
       "      <td>Y</td>\n",
       "      <td>27042651</td>\n",
       "      <td>27042685</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>15966</td>\n",
       "      <td>DAZ4</td>\n",
       "      <td>27042651-27042685</td>\n",
       "      <td>55.4482</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167796</th>\n",
       "      <td>ENSG00000185894</td>\n",
       "      <td>Y</td>\n",
       "      <td>27184955</td>\n",
       "      <td>27185060</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>18225</td>\n",
       "      <td>BPY2C</td>\n",
       "      <td>27184955-27185060</td>\n",
       "      <td>5.1943</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167797</th>\n",
       "      <td>ENSG00000185894</td>\n",
       "      <td>Y</td>\n",
       "      <td>27187915</td>\n",
       "      <td>27188032</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>18225</td>\n",
       "      <td>BPY2C</td>\n",
       "      <td>27187915-27188032</td>\n",
       "      <td>-14.0672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167798</th>\n",
       "      <td>ENSG00000185894</td>\n",
       "      <td>Y</td>\n",
       "      <td>27190092</td>\n",
       "      <td>27190169</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>18225</td>\n",
       "      <td>BPY2C</td>\n",
       "      <td>27190092-27190169</td>\n",
       "      <td>11.2080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167799</th>\n",
       "      <td>ENSG00000172288</td>\n",
       "      <td>Y</td>\n",
       "      <td>27768589</td>\n",
       "      <td>27770180</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>1809</td>\n",
       "      <td>CDY1</td>\n",
       "      <td>27768589-27770180</td>\n",
       "      <td>81.3163</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>166046 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                gene_id chromosome_name  start_coordinate  end_coordinate  \\\n",
       "0       ENSG00000187634               1            861321          861392   \n",
       "1       ENSG00000187634               1            865534          865715   \n",
       "2       ENSG00000187634               1            866418          866468   \n",
       "3       ENSG00000187634               1            871151          871275   \n",
       "4       ENSG00000187634               1            874419          874508   \n",
       "...                 ...             ...               ...             ...   \n",
       "167795  ENSG00000205916               Y          27042651        27042685   \n",
       "167796  ENSG00000185894               Y          27184955        27185060   \n",
       "167797  ENSG00000185894               Y          27187915        27188032   \n",
       "167798  ENSG00000185894               Y          27190092        27190169   \n",
       "167799  ENSG00000172288               Y          27768589        27770180   \n",
       "\n",
       "         tag  HGNC ID HGNC symbol              Range  max_score  \n",
       "0       CCDS    28706      SAMD11      861321-861392   334.9858  \n",
       "1       CCDS    28706      SAMD11      865534-865715   576.4071  \n",
       "2       CCDS    28706      SAMD11      866418-866468   146.7116  \n",
       "3       CCDS    28706      SAMD11      871151-871275   473.3769  \n",
       "4       CCDS    28706      SAMD11      874419-874508   112.1655  \n",
       "...      ...      ...         ...                ...        ...  \n",
       "167795  CCDS    15966        DAZ4  27042651-27042685    55.4482  \n",
       "167796  CCDS    18225       BPY2C  27184955-27185060     5.1943  \n",
       "167797  CCDS    18225       BPY2C  27187915-27188032   -14.0672  \n",
       "167798  CCDS    18225       BPY2C  27190092-27190169    11.2080  \n",
       "167799  CCDS     1809        CDY1  27768589-27770180    81.3163  \n",
       "\n",
       "[166046 rows x 9 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phylocsf_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ptex.loc[ptex['symbol'] == 'A2M', ['Range', 'PTEX_mean_value']].sort_values('Range')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_df[['RefSeq_HGNC', 'RefSeq_ranges', 'phyloCSF_max_score', 'PTEX_mean_value']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_rows', 100)\n",
    "tmp_ptex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phylocsf_df.loc[phylocsf_df['HGNC symbol'] == 'TP53']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 160745 entries, 0 to 160744\n",
      "Data columns (total 16 columns):\n",
      "CCRS_Gene                        160745 non-null object\n",
      "Chrom                            160745 non-null object\n",
      "RefSeq_Start                     160745 non-null int64\n",
      "RefSeq_End                       160745 non-null int64\n",
      "RefSeq_ranges                    160745 non-null object\n",
      "RefSeq_Length                    160745 non-null int64\n",
      "RefSeq_GeneID                    160745 non-null int64\n",
      "RefSeq_HGNC                      160745 non-null int64\n",
      "CDS_representation               160745 non-null object\n",
      "RefSeq_Ratio                     160745 non-null object\n",
      "RefSeq_Sharing_status            160745 non-null bool\n",
      "RefSeq_Sources                   160745 non-null object\n",
      "RefSeq_Count_CDS_alternative     160745 non-null int64\n",
      "RefSeq_Count_CDS_constitutive    160745 non-null int64\n",
      "RefSeq_mRNA_IDS                  160745 non-null object\n",
      "RefSeq_mRNA_nb                   160745 non-null int64\n",
      "dtypes: bool(1), int64(8), object(7)\n",
      "memory usage: 18.5+ MB\n"
     ]
    }
   ],
   "source": [
    "vcf_df.info(verbose=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 36705587 entries, 780 to 38837013\n",
      "Data columns (total 58 columns):\n",
      "HGNC ID                                 int64\n",
      "symbol                                  object\n",
      "ensg                                    object\n",
      "chrom                                   object\n",
      "pos                                     int64\n",
      "Prostate                                float64\n",
      "Spleen                                  float64\n",
      "Brain_FrontalCortex_BA9_                float64\n",
      "SmallIntestine_TerminalIleum            float64\n",
      "MinorSalivaryGland                      float64\n",
      "Artery_Coronary                         float64\n",
      "Skin_SunExposed_Lowerleg_               float64\n",
      "Cells_EBV_transformedlymphocytes        float64\n",
      "Brain_Hippocampus                       float64\n",
      "Esophagus_Muscularis                    float64\n",
      "Brain_Nucleusaccumbens_basalganglia_    float64\n",
      "Artery_Tibial                           float64\n",
      "Brain_Hypothalamus                      float64\n",
      "Adipose_Visceral_Omentum_               float64\n",
      "Cervix_Ectocervix                       float64\n",
      "Brain_Spinalcord_cervicalc_1_           float64\n",
      "Brain_CerebellarHemisphere              float64\n",
      "Nerve_Tibial                            float64\n",
      "Breast_MammaryTissue                    float64\n",
      "Liver                                   float64\n",
      "Skin_NotSunExposed_Suprapubic_          float64\n",
      "AdrenalGland                            float64\n",
      "Vagina                                  float64\n",
      "Pancreas                                float64\n",
      "Lung                                    float64\n",
      "FallopianTube                           float64\n",
      "Pituitary                               float64\n",
      "Muscle_Skeletal                         float64\n",
      "Colon_Transverse                        float64\n",
      "Artery_Aorta                            float64\n",
      "Heart_AtrialAppendage                   float64\n",
      "Adipose_Subcutaneous                    float64\n",
      "Esophagus_Mucosa                        float64\n",
      "Heart_LeftVentricle                     float64\n",
      "Brain_Cerebellum                        float64\n",
      "Brain_Cortex                            float64\n",
      "Thyroid                                 float64\n",
      "Brain_Substantianigra                   float64\n",
      "Kidney_Cortex                           float64\n",
      "Uterus                                  float64\n",
      "Stomach                                 float64\n",
      "WholeBlood                              float64\n",
      "Bladder                                 float64\n",
      "Brain_Anteriorcingulatecortex_BA24_     float64\n",
      "Brain_Putamen_basalganglia_             float64\n",
      "Brain_Caudate_basalganglia_             float64\n",
      "Colon_Sigmoid                           float64\n",
      "Cervix_Endocervix                       float64\n",
      "Ovary                                   float64\n",
      "Esophagus_GastroesophagealJunction      float64\n",
      "Testis                                  float64\n",
      "Brain_Amygdala                          float64\n",
      "mean_proportion                         float64\n",
      "dtypes: float64(53), int64(2), object(3)\n",
      "memory usage: 16.1+ GB\n"
     ]
    }
   ],
   "source": [
    "pext_modified['pos'] = pext_modified['pos'].astype(int)\n",
    "pext_modified.info(verbose=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phylocsf_df.loc[(phylocsf_df['HGNC symbol'] == 'TP53') & (phylocsf_df['start_coordinate'] <= 7579908) & (phylocsf_df['end_coordinate'] >= 7579909)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "11008it [02:09, 84.78it/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CCRS_Gene</th>\n",
       "      <th>Chrom</th>\n",
       "      <th>RefSeq_Start</th>\n",
       "      <th>RefSeq_End</th>\n",
       "      <th>RefSeq_ranges</th>\n",
       "      <th>RefSeq_Length</th>\n",
       "      <th>RefSeq_GeneID</th>\n",
       "      <th>RefSeq_HGNC</th>\n",
       "      <th>CDS_representation</th>\n",
       "      <th>RefSeq_Ratio</th>\n",
       "      <th>...</th>\n",
       "      <th>Brain_Putamen_basalganglia_</th>\n",
       "      <th>Brain_Caudate_basalganglia_</th>\n",
       "      <th>Colon_Sigmoid</th>\n",
       "      <th>Cervix_Endocervix</th>\n",
       "      <th>Ovary</th>\n",
       "      <th>Esophagus_GastroesophagealJunction</th>\n",
       "      <th>Testis</th>\n",
       "      <th>Brain_Amygdala</th>\n",
       "      <th>mean_proportion</th>\n",
       "      <th>phyloCSF_max_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PIGC</td>\n",
       "      <td>1</td>\n",
       "      <td>172410869</td>\n",
       "      <td>172411762</td>\n",
       "      <td>172410869-172411762</td>\n",
       "      <td>893</td>\n",
       "      <td>5279</td>\n",
       "      <td>8960</td>\n",
       "      <td>Unique</td>\n",
       "      <td>2.0/2</td>\n",
       "      <td>...</td>\n",
       "      <td>0.86556</td>\n",
       "      <td>0.8819</td>\n",
       "      <td>0.90581</td>\n",
       "      <td>0.90592</td>\n",
       "      <td>0.89855</td>\n",
       "      <td>0.92001</td>\n",
       "      <td>0.92957</td>\n",
       "      <td>0.83829</td>\n",
       "      <td>0.90539</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BORCS7</td>\n",
       "      <td>10</td>\n",
       "      <td>104614044</td>\n",
       "      <td>104614184</td>\n",
       "      <td>104614044-104614184</td>\n",
       "      <td>140</td>\n",
       "      <td>119032</td>\n",
       "      <td>23516</td>\n",
       "      <td>Unique</td>\n",
       "      <td>2.0/2</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>628.8381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BORCS7</td>\n",
       "      <td>10</td>\n",
       "      <td>104620089</td>\n",
       "      <td>104620151</td>\n",
       "      <td>104620089-104620151</td>\n",
       "      <td>62</td>\n",
       "      <td>119032</td>\n",
       "      <td>23516</td>\n",
       "      <td>Unique</td>\n",
       "      <td>2.0/2</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>289.6790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BORCS7</td>\n",
       "      <td>10</td>\n",
       "      <td>104620255</td>\n",
       "      <td>104620298</td>\n",
       "      <td>104620255-104620298</td>\n",
       "      <td>43</td>\n",
       "      <td>119032</td>\n",
       "      <td>23516</td>\n",
       "      <td>Unique</td>\n",
       "      <td>2.0/2</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>220.8366</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BORCS7</td>\n",
       "      <td>10</td>\n",
       "      <td>104621917</td>\n",
       "      <td>104621937</td>\n",
       "      <td>104621917-104621937</td>\n",
       "      <td>20</td>\n",
       "      <td>119032</td>\n",
       "      <td>23516</td>\n",
       "      <td>Unique</td>\n",
       "      <td>2.0/2</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.6227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>AGPAT1</td>\n",
       "      <td>CHR_HSCHR6_MHC_QBL_CTG1</td>\n",
       "      <td>32144287</td>\n",
       "      <td>32144289</td>\n",
       "      <td>32144287-32144289</td>\n",
       "      <td>2</td>\n",
       "      <td>10554</td>\n",
       "      <td>324</td>\n",
       "      <td>Unique</td>\n",
       "      <td>1.0/5</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>AGPAT1</td>\n",
       "      <td>CHR_HSCHR6_MHC_APD_CTG1</td>\n",
       "      <td>32144287</td>\n",
       "      <td>32144289</td>\n",
       "      <td>32144287-32144289</td>\n",
       "      <td>2</td>\n",
       "      <td>10554</td>\n",
       "      <td>324</td>\n",
       "      <td>Unique</td>\n",
       "      <td>1.0/5</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>AGPAT1</td>\n",
       "      <td>CHR_HSCHR6_MHC_MANN_CTG1</td>\n",
       "      <td>32144287</td>\n",
       "      <td>32144289</td>\n",
       "      <td>32144287-32144289</td>\n",
       "      <td>2</td>\n",
       "      <td>10554</td>\n",
       "      <td>324</td>\n",
       "      <td>Unique</td>\n",
       "      <td>1.0/5</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>AGPAT1</td>\n",
       "      <td>6</td>\n",
       "      <td>32144287</td>\n",
       "      <td>32144289</td>\n",
       "      <td>32144287-32144289</td>\n",
       "      <td>2</td>\n",
       "      <td>10554</td>\n",
       "      <td>324</td>\n",
       "      <td>Unique</td>\n",
       "      <td>1.0/5</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>AGPAT1</td>\n",
       "      <td>CHR_HSCHR6_MHC_SSTO_CTG1</td>\n",
       "      <td>32144287</td>\n",
       "      <td>32144289</td>\n",
       "      <td>32144287-32144289</td>\n",
       "      <td>2</td>\n",
       "      <td>10554</td>\n",
       "      <td>324</td>\n",
       "      <td>Unique</td>\n",
       "      <td>1.0/5</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>160745 rows × 70 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   CCRS_Gene                     Chrom  RefSeq_Start  RefSeq_End  \\\n",
       "0       PIGC                         1     172410869   172411762   \n",
       "0     BORCS7                        10     104614044   104614184   \n",
       "1     BORCS7                        10     104620089   104620151   \n",
       "2     BORCS7                        10     104620255   104620298   \n",
       "3     BORCS7                        10     104621917   104621937   \n",
       "..       ...                       ...           ...         ...   \n",
       "51    AGPAT1   CHR_HSCHR6_MHC_QBL_CTG1      32144287    32144289   \n",
       "52    AGPAT1   CHR_HSCHR6_MHC_APD_CTG1      32144287    32144289   \n",
       "53    AGPAT1  CHR_HSCHR6_MHC_MANN_CTG1      32144287    32144289   \n",
       "54    AGPAT1                         6      32144287    32144289   \n",
       "55    AGPAT1  CHR_HSCHR6_MHC_SSTO_CTG1      32144287    32144289   \n",
       "\n",
       "          RefSeq_ranges  RefSeq_Length  RefSeq_GeneID  RefSeq_HGNC  \\\n",
       "0   172410869-172411762            893           5279         8960   \n",
       "0   104614044-104614184            140         119032        23516   \n",
       "1   104620089-104620151             62         119032        23516   \n",
       "2   104620255-104620298             43         119032        23516   \n",
       "3   104621917-104621937             20         119032        23516   \n",
       "..                  ...            ...            ...          ...   \n",
       "51    32144287-32144289              2          10554          324   \n",
       "52    32144287-32144289              2          10554          324   \n",
       "53    32144287-32144289              2          10554          324   \n",
       "54    32144287-32144289              2          10554          324   \n",
       "55    32144287-32144289              2          10554          324   \n",
       "\n",
       "   CDS_representation RefSeq_Ratio  ...  Brain_Putamen_basalganglia_  \\\n",
       "0              Unique        2.0/2  ...                      0.86556   \n",
       "0              Unique        2.0/2  ...                          NaN   \n",
       "1              Unique        2.0/2  ...                          NaN   \n",
       "2              Unique        2.0/2  ...                          NaN   \n",
       "3              Unique        2.0/2  ...                          NaN   \n",
       "..                ...          ...  ...                          ...   \n",
       "51             Unique        1.0/5  ...                          NaN   \n",
       "52             Unique        1.0/5  ...                          NaN   \n",
       "53             Unique        1.0/5  ...                          NaN   \n",
       "54             Unique        1.0/5  ...                          NaN   \n",
       "55             Unique        1.0/5  ...                          NaN   \n",
       "\n",
       "   Brain_Caudate_basalganglia_  Colon_Sigmoid  Cervix_Endocervix    Ovary  \\\n",
       "0                       0.8819        0.90581            0.90592  0.89855   \n",
       "0                          NaN            NaN                NaN      NaN   \n",
       "1                          NaN            NaN                NaN      NaN   \n",
       "2                          NaN            NaN                NaN      NaN   \n",
       "3                          NaN            NaN                NaN      NaN   \n",
       "..                         ...            ...                ...      ...   \n",
       "51                         NaN            NaN                NaN      NaN   \n",
       "52                         NaN            NaN                NaN      NaN   \n",
       "53                         NaN            NaN                NaN      NaN   \n",
       "54                         NaN            NaN                NaN      NaN   \n",
       "55                         NaN            NaN                NaN      NaN   \n",
       "\n",
       "    Esophagus_GastroesophagealJunction   Testis  Brain_Amygdala  \\\n",
       "0                              0.92001  0.92957         0.83829   \n",
       "0                                  NaN      NaN             NaN   \n",
       "1                                  NaN      NaN             NaN   \n",
       "2                                  NaN      NaN             NaN   \n",
       "3                                  NaN      NaN             NaN   \n",
       "..                                 ...      ...             ...   \n",
       "51                                 NaN      NaN             NaN   \n",
       "52                                 NaN      NaN             NaN   \n",
       "53                                 NaN      NaN             NaN   \n",
       "54                                 NaN      NaN             NaN   \n",
       "55                                 NaN      NaN             NaN   \n",
       "\n",
       "    mean_proportion  phyloCSF_max_score  \n",
       "0           0.90539                 NaN  \n",
       "0               NaN            628.8381  \n",
       "1               NaN            289.6790  \n",
       "2               NaN            220.8366  \n",
       "3               NaN              7.6227  \n",
       "..              ...                 ...  \n",
       "51              NaN                 NaN  \n",
       "52              NaN                 NaN  \n",
       "53              NaN                 NaN  \n",
       "54              NaN                 NaN  \n",
       "55              NaN                 NaN  \n",
       "\n",
       "[160745 rows x 70 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def mapping(gene, l):\n",
    "    tmp_ptex = pext_modified.loc[pext_modified['HGNC ID'] == int(gene)].reset_index(drop=True)\n",
    "    tmp_phylocsf_df = phylocsf_df.loc[phylocsf_df['HGNC ID'] == int(gene)].reset_index(drop=True)\n",
    "    tmp_vcf = vcf_df.loc[vcf_df['RefSeq_HGNC'] == int(gene)].sort_values(by='RefSeq_ranges').reset_index(drop=True)\n",
    "\n",
    "    l_ptex = list()\n",
    "    l_phylocsf = list()\n",
    "\n",
    "    for r in tmp_vcf['RefSeq_ranges'].values:\n",
    "        ptex_values = tmp_ptex.loc[(tmp_ptex['pos'] >= int(r.split('-')[0])) & (tmp_ptex['pos'] <= int(r.split('-')[1])), list(tmp_ptex.columns)[5:]].mean().to_dict()\n",
    "        l_ptex.append(ptex_values)\n",
    "\n",
    "        phylocsf_score = tmp_phylocsf_df.loc[(tmp_phylocsf_df['start_coordinate']-1 <= int(r.split('-')[0])) & (int(r.split('-')[1]) <= tmp_phylocsf_df['end_coordinate']+1), 'max_score'].values\n",
    "        if len(phylocsf_score) > 0:\n",
    "            l_phylocsf.append({'phyloCSF_max_score' : np.mean(phylocsf_score)})\n",
    "        else:\n",
    "            l_phylocsf.append({'phyloCSF_max_score' : np.nan})\n",
    "\n",
    "    l.append(pd.concat([tmp_vcf, pd.DataFrame(l_ptex), pd.DataFrame(l_phylocsf)], axis=1))\n",
    "\n",
    "m = multiprocessing.Manager()\n",
    "mp_l = m.list()\n",
    "# parmap.starmap(mapping, list(zip([11998])), mp_l, pm_pbar=True)\n",
    "parmap.starmap(mapping, list(zip(genes)), mp_l, pm_pbar=True)\n",
    "output_df = pd.concat(list(mp_l))\n",
    "output_df['Chrom'] = output_df['Chrom'].astype(str)\n",
    "output_df = output_df.loc[(~output_df.Chrom.str.contains('CHR')) & output_df['mean_proportion'].isna() == False]\n",
    "output_df\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_df.loc[output_df['CCRS_Gene'] == 'TP53'].sort_values(by='RefSeq_ranges')[['RefSeq_ranges', 'mean_proportion', 'phyloCSF_max_score']].drop_duplicates().values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CCRS_Gene</th>\n",
       "      <th>Chrom</th>\n",
       "      <th>RefSeq_Start</th>\n",
       "      <th>RefSeq_End</th>\n",
       "      <th>RefSeq_ranges</th>\n",
       "      <th>RefSeq_Length</th>\n",
       "      <th>RefSeq_GeneID</th>\n",
       "      <th>RefSeq_HGNC</th>\n",
       "      <th>CDS_representation</th>\n",
       "      <th>RefSeq_Ratio</th>\n",
       "      <th>...</th>\n",
       "      <th>Brain_Putamen_basalganglia_</th>\n",
       "      <th>Brain_Caudate_basalganglia_</th>\n",
       "      <th>Colon_Sigmoid</th>\n",
       "      <th>Cervix_Endocervix</th>\n",
       "      <th>Ovary</th>\n",
       "      <th>Esophagus_GastroesophagealJunction</th>\n",
       "      <th>Testis</th>\n",
       "      <th>Brain_Amygdala</th>\n",
       "      <th>mean_proportion</th>\n",
       "      <th>phyloCSF_max_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PIGC</td>\n",
       "      <td>1</td>\n",
       "      <td>172410869</td>\n",
       "      <td>172411762</td>\n",
       "      <td>172410869-172411762</td>\n",
       "      <td>893</td>\n",
       "      <td>5279</td>\n",
       "      <td>8960</td>\n",
       "      <td>Unique</td>\n",
       "      <td>2.0/2</td>\n",
       "      <td>...</td>\n",
       "      <td>0.865560</td>\n",
       "      <td>0.881900</td>\n",
       "      <td>0.905810</td>\n",
       "      <td>0.905920</td>\n",
       "      <td>0.898550</td>\n",
       "      <td>0.920010</td>\n",
       "      <td>0.929570</td>\n",
       "      <td>0.838290</td>\n",
       "      <td>0.905390</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ZNF669</td>\n",
       "      <td>1</td>\n",
       "      <td>247263676</td>\n",
       "      <td>247264621</td>\n",
       "      <td>247263676-247264621</td>\n",
       "      <td>945</td>\n",
       "      <td>79862</td>\n",
       "      <td>25736</td>\n",
       "      <td>Unique</td>\n",
       "      <td>2.0/2</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.908027</td>\n",
       "      <td>0.964345</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.996064</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ZNF669</td>\n",
       "      <td>1</td>\n",
       "      <td>247265033</td>\n",
       "      <td>247265093</td>\n",
       "      <td>247265033-247265093</td>\n",
       "      <td>60</td>\n",
       "      <td>79862</td>\n",
       "      <td>25736</td>\n",
       "      <td>Unique</td>\n",
       "      <td>2.0/2</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.898830</td>\n",
       "      <td>0.960780</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.995670</td>\n",
       "      <td>-8.1007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ZNF669</td>\n",
       "      <td>1</td>\n",
       "      <td>247265289</td>\n",
       "      <td>247265415</td>\n",
       "      <td>247265289-247265415</td>\n",
       "      <td>126</td>\n",
       "      <td>79862</td>\n",
       "      <td>25736</td>\n",
       "      <td>Unique</td>\n",
       "      <td>2.0/2</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>14.8000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ZNF669</td>\n",
       "      <td>1</td>\n",
       "      <td>247267241</td>\n",
       "      <td>247267499</td>\n",
       "      <td>247267241-247267499</td>\n",
       "      <td>258</td>\n",
       "      <td>79862</td>\n",
       "      <td>25736</td>\n",
       "      <td>Variable_region</td>\n",
       "      <td>1.0/2</td>\n",
       "      <td>...</td>\n",
       "      <td>0.003861</td>\n",
       "      <td>0.003861</td>\n",
       "      <td>0.003861</td>\n",
       "      <td>0.104640</td>\n",
       "      <td>0.077106</td>\n",
       "      <td>0.003861</td>\n",
       "      <td>0.065503</td>\n",
       "      <td>0.003861</td>\n",
       "      <td>0.039514</td>\n",
       "      <td>143.2459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>AGPAT1</td>\n",
       "      <td>6</td>\n",
       "      <td>32137942</td>\n",
       "      <td>32138037</td>\n",
       "      <td>32137942-32138037</td>\n",
       "      <td>95</td>\n",
       "      <td>10554</td>\n",
       "      <td>324</td>\n",
       "      <td>Unique</td>\n",
       "      <td>5.0/5</td>\n",
       "      <td>...</td>\n",
       "      <td>0.980250</td>\n",
       "      <td>0.984710</td>\n",
       "      <td>0.982810</td>\n",
       "      <td>0.988680</td>\n",
       "      <td>0.983010</td>\n",
       "      <td>0.978470</td>\n",
       "      <td>0.982300</td>\n",
       "      <td>0.983850</td>\n",
       "      <td>0.984970</td>\n",
       "      <td>391.7055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>AGPAT1</td>\n",
       "      <td>6</td>\n",
       "      <td>32138202</td>\n",
       "      <td>32138377</td>\n",
       "      <td>32138202-32138377</td>\n",
       "      <td>175</td>\n",
       "      <td>10554</td>\n",
       "      <td>324</td>\n",
       "      <td>Unique</td>\n",
       "      <td>5.0/5</td>\n",
       "      <td>...</td>\n",
       "      <td>0.980250</td>\n",
       "      <td>0.984710</td>\n",
       "      <td>0.982760</td>\n",
       "      <td>0.988603</td>\n",
       "      <td>0.982960</td>\n",
       "      <td>0.978418</td>\n",
       "      <td>0.982263</td>\n",
       "      <td>0.983829</td>\n",
       "      <td>0.984935</td>\n",
       "      <td>754.5904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>AGPAT1</td>\n",
       "      <td>6</td>\n",
       "      <td>32138714</td>\n",
       "      <td>32138847</td>\n",
       "      <td>32138714-32138847</td>\n",
       "      <td>133</td>\n",
       "      <td>10554</td>\n",
       "      <td>324</td>\n",
       "      <td>Unique</td>\n",
       "      <td>5.0/5</td>\n",
       "      <td>...</td>\n",
       "      <td>0.980250</td>\n",
       "      <td>0.984710</td>\n",
       "      <td>0.978390</td>\n",
       "      <td>0.981860</td>\n",
       "      <td>0.978610</td>\n",
       "      <td>0.973890</td>\n",
       "      <td>0.979080</td>\n",
       "      <td>0.982010</td>\n",
       "      <td>0.981910</td>\n",
       "      <td>534.3331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>AGPAT1</td>\n",
       "      <td>6</td>\n",
       "      <td>32139074</td>\n",
       "      <td>32139273</td>\n",
       "      <td>32139074-32139273</td>\n",
       "      <td>199</td>\n",
       "      <td>10554</td>\n",
       "      <td>324</td>\n",
       "      <td>Variable_region</td>\n",
       "      <td>5.0/5</td>\n",
       "      <td>...</td>\n",
       "      <td>0.980250</td>\n",
       "      <td>0.984710</td>\n",
       "      <td>0.978390</td>\n",
       "      <td>0.981860</td>\n",
       "      <td>0.978610</td>\n",
       "      <td>0.973890</td>\n",
       "      <td>0.979080</td>\n",
       "      <td>0.982010</td>\n",
       "      <td>0.981910</td>\n",
       "      <td>475.7562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>AGPAT1</td>\n",
       "      <td>6</td>\n",
       "      <td>32139273</td>\n",
       "      <td>32139282</td>\n",
       "      <td>32139273-32139282</td>\n",
       "      <td>9</td>\n",
       "      <td>10554</td>\n",
       "      <td>324</td>\n",
       "      <td>Variable_region</td>\n",
       "      <td>1.0/5</td>\n",
       "      <td>...</td>\n",
       "      <td>0.980250</td>\n",
       "      <td>0.984710</td>\n",
       "      <td>0.978390</td>\n",
       "      <td>0.981860</td>\n",
       "      <td>0.978610</td>\n",
       "      <td>0.973890</td>\n",
       "      <td>0.979080</td>\n",
       "      <td>0.982010</td>\n",
       "      <td>0.981910</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>135251 rows × 70 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   CCRS_Gene Chrom  RefSeq_Start  RefSeq_End        RefSeq_ranges  \\\n",
       "0       PIGC     1     172410869   172411762  172410869-172411762   \n",
       "0     ZNF669     1     247263676   247264621  247263676-247264621   \n",
       "1     ZNF669     1     247265033   247265093  247265033-247265093   \n",
       "2     ZNF669     1     247265289   247265415  247265289-247265415   \n",
       "3     ZNF669     1     247267241   247267499  247267241-247267499   \n",
       "..       ...   ...           ...         ...                  ...   \n",
       "20    AGPAT1     6      32137942    32138037    32137942-32138037   \n",
       "27    AGPAT1     6      32138202    32138377    32138202-32138377   \n",
       "28    AGPAT1     6      32138714    32138847    32138714-32138847   \n",
       "36    AGPAT1     6      32139074    32139273    32139074-32139273   \n",
       "42    AGPAT1     6      32139273    32139282    32139273-32139282   \n",
       "\n",
       "    RefSeq_Length  RefSeq_GeneID  RefSeq_HGNC CDS_representation RefSeq_Ratio  \\\n",
       "0             893           5279         8960             Unique        2.0/2   \n",
       "0             945          79862        25736             Unique        2.0/2   \n",
       "1              60          79862        25736             Unique        2.0/2   \n",
       "2             126          79862        25736             Unique        2.0/2   \n",
       "3             258          79862        25736    Variable_region        1.0/2   \n",
       "..            ...            ...          ...                ...          ...   \n",
       "20             95          10554          324             Unique        5.0/5   \n",
       "27            175          10554          324             Unique        5.0/5   \n",
       "28            133          10554          324             Unique        5.0/5   \n",
       "36            199          10554          324    Variable_region        5.0/5   \n",
       "42              9          10554          324    Variable_region        1.0/5   \n",
       "\n",
       "    ...  Brain_Putamen_basalganglia_ Brain_Caudate_basalganglia_  \\\n",
       "0   ...                     0.865560                    0.881900   \n",
       "0   ...                     1.000000                    1.000000   \n",
       "1   ...                     1.000000                    1.000000   \n",
       "2   ...                     1.000000                    1.000000   \n",
       "3   ...                     0.003861                    0.003861   \n",
       "..  ...                          ...                         ...   \n",
       "20  ...                     0.980250                    0.984710   \n",
       "27  ...                     0.980250                    0.984710   \n",
       "28  ...                     0.980250                    0.984710   \n",
       "36  ...                     0.980250                    0.984710   \n",
       "42  ...                     0.980250                    0.984710   \n",
       "\n",
       "    Colon_Sigmoid  Cervix_Endocervix     Ovary  \\\n",
       "0        0.905810           0.905920  0.898550   \n",
       "0        1.000000           0.908027  0.964345   \n",
       "1        1.000000           0.898830  0.960780   \n",
       "2        1.000000           1.000000  1.000000   \n",
       "3        0.003861           0.104640  0.077106   \n",
       "..            ...                ...       ...   \n",
       "20       0.982810           0.988680  0.983010   \n",
       "27       0.982760           0.988603  0.982960   \n",
       "28       0.978390           0.981860  0.978610   \n",
       "36       0.978390           0.981860  0.978610   \n",
       "42       0.978390           0.981860  0.978610   \n",
       "\n",
       "    Esophagus_GastroesophagealJunction    Testis  Brain_Amygdala  \\\n",
       "0                             0.920010  0.929570        0.838290   \n",
       "0                             1.000000  1.000000        1.000000   \n",
       "1                             1.000000  1.000000        1.000000   \n",
       "2                             1.000000  1.000000        1.000000   \n",
       "3                             0.003861  0.065503        0.003861   \n",
       "..                                 ...       ...             ...   \n",
       "20                            0.978470  0.982300        0.983850   \n",
       "27                            0.978418  0.982263        0.983829   \n",
       "28                            0.973890  0.979080        0.982010   \n",
       "36                            0.973890  0.979080        0.982010   \n",
       "42                            0.973890  0.979080        0.982010   \n",
       "\n",
       "    mean_proportion  phyloCSF_max_score  \n",
       "0          0.905390                 NaN  \n",
       "0          0.996064                 NaN  \n",
       "1          0.995670             -8.1007  \n",
       "2          1.000000             14.8000  \n",
       "3          0.039514            143.2459  \n",
       "..              ...                 ...  \n",
       "20         0.984970            391.7055  \n",
       "27         0.984935            754.5904  \n",
       "28         0.981910            534.3331  \n",
       "36         0.981910            475.7562  \n",
       "42         0.981910                 NaN  \n",
       "\n",
       "[135251 rows x 70 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "output_df['Chrom'] = output_df['Chrom'].astype(str)\n",
    "output_df = output_df[~output_df.Chrom.str.contains('CHR')]\n",
    "output_df = output_df.loc[output_df['mean_proportion'].isna() == False]\n",
    "output_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_df.to_parquet('/home/weber/PycharmProjects/ExoCarto/data/2_processed/refseq_multi_iso_pext_phylocsf.parquet')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "tp53_df = pd.read_csv('/home/weber/PycharmProjects/ExoCarto-Augusta/RefSeq_X_VCF_TP53-BRCA1.csv.gz', sep='\\t', compression='gzip')\n",
    "tp53_df = tp53_df.loc[tp53_df['RefSeq_HGNC'] == 11998].sort_values(by='CCRS_ranges')\n",
    "tp53_df.CCRS_ranges.unique()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.7.6 64-bit ('ExoCarto': conda)",
   "language": "python",
   "name": "python37664bitexocartoconda47cfe11889aa4008ba405d9d7b40c893"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
